{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "palestinian-motion",
   "metadata": {},
   "source": [
    "# Data preparation for [Compass](https://github.com/YosefLab/Compass)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "elegant-salvation",
   "metadata": {},
   "source": [
    "## Preparing novarino processed data for Compass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "identified-editing",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.io import mmread\n",
    "\n",
    "from scipy.sparse import issparse\n",
    "def normalize_tpm(adata, scale = 1e6):\n",
    "    X = adata.X.toarray().T / adata.var.Length.values[:, None] * 1000\n",
    "    factors = X.sum(axis = 0) / scale\n",
    "    return X.T / factors[:, None]\n",
    "\n",
    "def write_expression(adataobj, file, layer = None, normalize = True):\n",
    "    adataobj = adataobj.copy()\n",
    "    if layer:\n",
    "        adataobj.X = adataobj.layers[layer]\n",
    "    \n",
    "    if normalize:\n",
    "        X = normalize_tpm(adataobj)\n",
    "    \n",
    "    else:\n",
    "        X = adataobj.X\n",
    "    \n",
    "    if issparse(X):\n",
    "        X = X.toarray()\n",
    "    \n",
    "    df = pd.DataFrame(\n",
    "        X.T,\n",
    "        index = adataobj.var.index.to_list(),\n",
    "        columns = adataobj.obs.index.to_list()\n",
    "    )\n",
    "    df.to_csv(\n",
    "        file,\n",
    "        sep = '\\t'\n",
    "    )\n",
    "\n",
    "def data2h5ad(prefix):\n",
    "    X = mmread(prefix + '.mtx').tocsr()\n",
    "    obs = pd.read_csv(\n",
    "        prefix + '.metadata.tsv',\n",
    "        sep = '\\t'\n",
    "    )\n",
    "    with open(prefix + f'.rownames.txt', 'r') as f:\n",
    "        genes = [r.rstrip() for r in f]\n",
    "    \n",
    "    var = pd.DataFrame(\n",
    "        index = genes\n",
    "    )\n",
    "    \n",
    "    return ad.AnnData(\n",
    "        X = X.T,\n",
    "        obs = obs,\n",
    "        var = var\n",
    "    )\n",
    "\n",
    "def splith5ad(adataobj, split_on):\n",
    "    subsets = adataobj.obs[split_on].unique()\n",
    "    return {subset: adataobj[adataobj.obs[split_on] == subset, :].copy() for subset in subsets}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "different-stadium",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 203 × 26249\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = data2h5ad('../processed/novarino_scRNA')\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "closing-glance",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = sc.read_h5ad('../processed/scRNAseq_featureCounts.h5ad')\n",
    "genes = tmp.var.copy()\n",
    "adata.var['Length'] = adata.var.merge(\n",
    "    tmp.var,\n",
    "    left_index = True,\n",
    "    right_index = True\n",
    ").Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "existing-evaluation",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata[:, ~adata.var.Length.isna()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "champion-criterion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>seq_folder</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>log10GenesPerUMI</th>\n",
       "      <th>mitoRatio</th>\n",
       "      <th>cells</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_mouseID</th>\n",
       "      <th>sample_litter</th>\n",
       "      <th>sample_treatment</th>\n",
       "      <th>sample_sex</th>\n",
       "      <th>RNA_snn_res.0.9</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>wt_162322</th>\n",
       "      <td>wt</td>\n",
       "      <td>62570</td>\n",
       "      <td>2514</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>62723</td>\n",
       "      <td>2537</td>\n",
       "      <td>0.709614</td>\n",
       "      <td>0.065861</td>\n",
       "      <td>wt_162322</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>olfactory_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wt_162323</th>\n",
       "      <td>wt</td>\n",
       "      <td>99844</td>\n",
       "      <td>3289</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>99977</td>\n",
       "      <td>3315</td>\n",
       "      <td>0.704111</td>\n",
       "      <td>0.043490</td>\n",
       "      <td>wt_162323</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>oligodendrytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wt_162324</th>\n",
       "      <td>wt</td>\n",
       "      <td>107544</td>\n",
       "      <td>3215</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>107931</td>\n",
       "      <td>3279</td>\n",
       "      <td>0.698518</td>\n",
       "      <td>0.121318</td>\n",
       "      <td>wt_162324</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>olfactory_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wt_162325</th>\n",
       "      <td>wt</td>\n",
       "      <td>136321</td>\n",
       "      <td>3141</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>136624</td>\n",
       "      <td>3172</td>\n",
       "      <td>0.681787</td>\n",
       "      <td>0.037987</td>\n",
       "      <td>wt_162325</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wt_162326</th>\n",
       "      <td>wt</td>\n",
       "      <td>109397</td>\n",
       "      <td>3513</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>109533</td>\n",
       "      <td>3537</td>\n",
       "      <td>0.704158</td>\n",
       "      <td>0.073549</td>\n",
       "      <td>wt_162326</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ko_162553</th>\n",
       "      <td>ko</td>\n",
       "      <td>331737</td>\n",
       "      <td>4999</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>332022</td>\n",
       "      <td>5055</td>\n",
       "      <td>0.670822</td>\n",
       "      <td>0.071040</td>\n",
       "      <td>ko_162553</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ko_162554</th>\n",
       "      <td>ko</td>\n",
       "      <td>435996</td>\n",
       "      <td>5670</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>436357</td>\n",
       "      <td>5742</td>\n",
       "      <td>0.666519</td>\n",
       "      <td>0.042850</td>\n",
       "      <td>ko_162554</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ko_162555</th>\n",
       "      <td>ko</td>\n",
       "      <td>260654</td>\n",
       "      <td>4243</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>260946</td>\n",
       "      <td>4306</td>\n",
       "      <td>0.670920</td>\n",
       "      <td>0.062971</td>\n",
       "      <td>ko_162555</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>olfactory_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ko_162556</th>\n",
       "      <td>ko</td>\n",
       "      <td>211388</td>\n",
       "      <td>3149</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>211590</td>\n",
       "      <td>3177</td>\n",
       "      <td>0.657595</td>\n",
       "      <td>0.034033</td>\n",
       "      <td>ko_162556</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>olfactory_cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ko_162557</th>\n",
       "      <td>ko</td>\n",
       "      <td>415384</td>\n",
       "      <td>5334</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>415574</td>\n",
       "      <td>5380</td>\n",
       "      <td>0.664000</td>\n",
       "      <td>0.020302</td>\n",
       "      <td>ko_162557</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>olfactory_cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>203 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          orig.ident  nCount_RNA  nFeature_RNA     seq_folder    nUMI  nGene  \\\n",
       "wt_162322         wt       62570          2514  SeuratProject   62723   2537   \n",
       "wt_162323         wt       99844          3289  SeuratProject   99977   3315   \n",
       "wt_162324         wt      107544          3215  SeuratProject  107931   3279   \n",
       "wt_162325         wt      136321          3141  SeuratProject  136624   3172   \n",
       "wt_162326         wt      109397          3513  SeuratProject  109533   3537   \n",
       "...              ...         ...           ...            ...     ...    ...   \n",
       "ko_162553         ko      331737          4999  SeuratProject  332022   5055   \n",
       "ko_162554         ko      435996          5670  SeuratProject  436357   5742   \n",
       "ko_162555         ko      260654          4243  SeuratProject  260946   4306   \n",
       "ko_162556         ko      211388          3149  SeuratProject  211590   3177   \n",
       "ko_162557         ko      415384          5334  SeuratProject  415574   5380   \n",
       "\n",
       "           log10GenesPerUMI  mitoRatio      cells sample sample_mouseID  \\\n",
       "wt_162322          0.709614   0.065861  wt_162322     wt       GNF1/464   \n",
       "wt_162323          0.704111   0.043490  wt_162323     wt       GNF1/464   \n",
       "wt_162324          0.698518   0.121318  wt_162324     wt       GNF1/464   \n",
       "wt_162325          0.681787   0.037987  wt_162325     wt       GNF1/464   \n",
       "wt_162326          0.704158   0.073549  wt_162326     wt       GNF1/464   \n",
       "...                     ...        ...        ...    ...            ...   \n",
       "ko_162553          0.670822   0.071040  ko_162553     ko       GN21/475   \n",
       "ko_162554          0.666519   0.042850  ko_162554     ko       GN21/475   \n",
       "ko_162555          0.670920   0.062971  ko_162555     ko       GN21/475   \n",
       "ko_162556          0.657595   0.034033  ko_162556     ko       GN21/475   \n",
       "ko_162557          0.664000   0.020302  ko_162557     ko       GN21/475   \n",
       "\n",
       "          sample_litter sample_treatment sample_sex  RNA_snn_res.0.9  \\\n",
       "wt_162322       G1/59 A           prep 1          f                0   \n",
       "wt_162323       G1/59 A           prep 1          f                1   \n",
       "wt_162324       G1/59 A           prep 1          f                0   \n",
       "wt_162325       G1/59 A           prep 1          f                3   \n",
       "wt_162326       G1/59 A           prep 1          f                3   \n",
       "...                 ...              ...        ...              ...   \n",
       "ko_162553       G1/59 B           prep 2          f                3   \n",
       "ko_162554       G1/59 B           prep 2          f                3   \n",
       "ko_162555       G1/59 B           prep 2          f                0   \n",
       "ko_162556       G1/59 B           prep 2          f                0   \n",
       "ko_162557       G1/59 B           prep 2          f                0   \n",
       "\n",
       "           seurat_clusters        cell_type  \n",
       "wt_162322                0  olfactory_cells  \n",
       "wt_162323                1   oligodendrytes  \n",
       "wt_162324                0  olfactory_cells  \n",
       "wt_162325                3          neurons  \n",
       "wt_162326                3          neurons  \n",
       "...                    ...              ...  \n",
       "ko_162553                3          neurons  \n",
       "ko_162554                3          neurons  \n",
       "ko_162555                0  olfactory_cells  \n",
       "ko_162556                0  olfactory_cells  \n",
       "ko_162557                0  olfactory_cells  \n",
       "\n",
       "[203 rows x 17 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation = {\n",
    "    0: 'olfactory_cells',\n",
    "    1: 'oligodendrytes',\n",
    "    2: 'astrocytes',\n",
    "    3: 'neurons'\n",
    "}\n",
    "adata.obs['cell_type'] = adata.obs.seurat_clusters.apply(lambda x: annotation[x])\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dangerous-electricity",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'olfactory_cells_wt': View of AnnData object with n_obs × n_vars = 35 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'olfactory_cells_ko': View of AnnData object with n_obs × n_vars = 20 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'oligodendrytes_wt': View of AnnData object with n_obs × n_vars = 33 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'oligodendrytes_ko': View of AnnData object with n_obs × n_vars = 19 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'astrocytes_wt': View of AnnData object with n_obs × n_vars = 23 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'astrocytes_ko': View of AnnData object with n_obs × n_vars = 27 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'neurons_wt': View of AnnData object with n_obs × n_vars = 15 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length',\n",
       " 'neurons_ko': View of AnnData object with n_obs × n_vars = 31 × 21907\n",
       "     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "     var: 'Length'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = {}\n",
    "for celltype in annotation.values():\n",
    "    for sample in ['wt', 'ko']:\n",
    "        tmp = adata[(adata.obs.cell_type == celltype) & (adata.obs['sample'] == sample), :].copy()\n",
    "        data[celltype + '_' + sample] = tmp[:, np.array((adata.X > 0).sum(axis = 0)).flatten() > 3]\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "convinced-burlington",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in data.keys():\n",
    "    write_expression(\n",
    "        data[key], \n",
    "        f'../processed/novarino/{key}.expression.tsv',\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "provincial-wagner",
   "metadata": {},
   "source": [
    "## Preparing featureCounts and Salmon quantification for Compass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "assured-isolation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'featureCounts': {'olfactory_cells_wt': AnnData object with n_obs × n_vars = 32 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'olfactory_cells_ko': AnnData object with n_obs × n_vars = 20 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'oligodendrytes_wt': AnnData object with n_obs × n_vars = 32 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'oligodendrytes_ko': AnnData object with n_obs × n_vars = 15 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'astrocytes_wt': AnnData object with n_obs × n_vars = 22 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'astrocytes_ko': AnnData object with n_obs × n_vars = 27 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'neurons_wt': AnnData object with n_obs × n_vars = 15 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length',\n",
       "  'neurons_ko': AnnData object with n_obs × n_vars = 31 × 23897\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length'},\n",
       " 'salmon': {'olfactory_cells_wt': AnnData object with n_obs × n_vars = 36 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'olfactory_cells_ko': AnnData object with n_obs × n_vars = 22 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'oligodendrytes_wt': AnnData object with n_obs × n_vars = 27 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'oligodendrytes_ko': AnnData object with n_obs × n_vars = 14 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'astrocytes_wt': AnnData object with n_obs × n_vars = 23 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'astrocytes_ko': AnnData object with n_obs × n_vars = 27 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'neurons_wt': AnnData object with n_obs × n_vars = 15 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm',\n",
       "  'neurons_ko': AnnData object with n_obs × n_vars = 30 × 24500\n",
       "      obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID', 'nCount_RNA', 'nFeature_RNA', 'percent_mt', 'percent_ribo', 'leiden_r0.25', 'leiden_r0.5', 'leiden_r0.75', 'leiden_r1', 'cell_type'\n",
       "      var: 'Length', 'gencode_id'\n",
       "      layers: 'counts', 'tpm'}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fdata_obs = sc.read_h5ad('../processed/scRNAseq_featureCounts.clustered.h5ad').obs.copy()\n",
    "fdata = sc.read_h5ad('../processed/scRNAseq_featureCounts.h5ad')[fdata_obs.index, :].copy()\n",
    "fdata.obs = fdata_obs\n",
    "sdata_obs = sc.read_h5ad('../processed/scRNAseq_salmon.clustered.h5ad').obs.copy()\n",
    "sdata = sc.read_h5ad('../processed/scRNAseq_salmon.h5ad')[sdata_obs.index, :].copy()\n",
    "sdata.obs = sdata_obs\n",
    "\n",
    "data = {}\n",
    "for key, adata in zip( \n",
    "    ['featureCounts', 'salmon'],\n",
    "    [fdata, sdata]\n",
    "):\n",
    "    tmp = {}\n",
    "    for celltype in annotation.values():\n",
    "        for sample in ['wt', 'ko']:\n",
    "            tmp_adata = adata[(adata.obs.cell_type == celltype) & (adata.obs['sample_genotype'] == sample), :].copy()\n",
    "            tmp[celltype + '_' + sample] = tmp_adata[:, np.array((adata.X > 0).sum(axis = 0)).flatten() > 3].copy()\n",
    "    \n",
    "    data[key] = tmp\n",
    "\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "violent-physiology",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "for key in data.keys():\n",
    "    path = f'../processed/{key}'\n",
    "    if not os.path.exists(path):\n",
    "        os.mkdir(path)\n",
    "\n",
    "key1 = 'salmon'\n",
    "for key2, adata in data[key1].items():\n",
    "    write_expression(\n",
    "        adata, \n",
    "        f'../processed/{key1}/{key2}.expression.tsv',\n",
    "        normalize = False,\n",
    "        layer = 'tpm'\n",
    "    )\n",
    "    \n",
    "key1 = 'featureCounts'\n",
    "for key2, adata in data[key1].items():\n",
    "    write_expression(\n",
    "        adata, \n",
    "        f'../processed/{key1}/{key2}.expression.tsv',\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hairy-teacher",
   "metadata": {},
   "source": [
    "## Preparing bulk RNAseq for Compass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "weird-sister",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import re\n",
    "def read_feature_counts(\n",
    "    filename, \n",
    "    clean_regex = None,\n",
    "    rename_columns = None\n",
    "):\n",
    "    with gzip.open(filename) as file:\n",
    "        #getting rid of comment\n",
    "        file.readline()\n",
    "        \n",
    "        tab = pd.read_csv(\n",
    "            file,\n",
    "            sep = '\\t'\n",
    "        )\n",
    "        \n",
    "    exclude_cols = ['Chr', 'Start', 'End', 'Strand']\n",
    "    if clean_regex:\n",
    "        new_columns = []\n",
    "        for col in tab.columns:\n",
    "            match = re.findall(clean_regex, col)\n",
    "            if match:\n",
    "                new_columns.append(match[0])\n",
    "                \n",
    "            else:\n",
    "                new_columns.append(col)\n",
    "                \n",
    "        tab.columns = new_columns\n",
    "    \n",
    "    if rename_columns:\n",
    "        tab.rename(\n",
    "            columns = rename_columns,\n",
    "            inplace = True\n",
    "        )\n",
    "    \n",
    "    return tab.loc[:, ~tab.columns.isin(exclude_cols)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "parental-calculator",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>genotype</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>sex</th>\n",
       "      <th>litter</th>\n",
       "      <th>barcode</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>81446</td>\n",
       "      <td>GTGCCA</td>\n",
       "      <td>wt</td>\n",
       "      <td>P1</td>\n",
       "      <td>m</td>\n",
       "      <td>1</td>\n",
       "      <td>7049</td>\n",
       "      <td>A_wt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>81447</td>\n",
       "      <td>TCGAGG</td>\n",
       "      <td>ko</td>\n",
       "      <td>P1</td>\n",
       "      <td>m</td>\n",
       "      <td>1</td>\n",
       "      <td>7050</td>\n",
       "      <td>A_ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>81448</td>\n",
       "      <td>CACTAA</td>\n",
       "      <td>wt</td>\n",
       "      <td>P1</td>\n",
       "      <td>f</td>\n",
       "      <td>1</td>\n",
       "      <td>7051</td>\n",
       "      <td>B_wt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>81449</td>\n",
       "      <td>GGTATA</td>\n",
       "      <td>ko</td>\n",
       "      <td>P1</td>\n",
       "      <td>f</td>\n",
       "      <td>1</td>\n",
       "      <td>7052</td>\n",
       "      <td>B_ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>81450</td>\n",
       "      <td>CGCCTG</td>\n",
       "      <td>wt</td>\n",
       "      <td>P1</td>\n",
       "      <td>m</td>\n",
       "      <td>2</td>\n",
       "      <td>7053</td>\n",
       "      <td>C_wt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>81451</td>\n",
       "      <td>AATGAA</td>\n",
       "      <td>ko</td>\n",
       "      <td>P1</td>\n",
       "      <td>m</td>\n",
       "      <td>2</td>\n",
       "      <td>7054</td>\n",
       "      <td>C_ko</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_ID sequence genotype timepoint sex  litter  barcode  name\n",
       "0      81446   GTGCCA       wt        P1   m       1     7049  A_wt\n",
       "1      81447   TCGAGG       ko        P1   m       1     7050  A_ko\n",
       "2      81448   CACTAA       wt        P1   f       1     7051  B_wt\n",
       "3      81449   GGTATA       ko        P1   f       1     7052  B_ko\n",
       "4      81450   CGCCTG       wt        P1   m       2     7053  C_wt\n",
       "5      81451   AATGAA       ko        P1   m       2     7054  C_ko"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv('../meta/sampleList_Slc7a5_EmxCre_P1_Ctx.csv')\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "controlled-glenn",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'wt': [81446, 81448, 81450], 'ko': [81447, 81449, 81451]}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_type = {\n",
    "    key: meta.loc[meta.genotype == key, 'sample_ID'].tolist() for key in ['wt', 'ko']\n",
    "}\n",
    "sample_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "physical-money",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>81446</th>\n",
       "      <th>81447</th>\n",
       "      <th>81448</th>\n",
       "      <th>81449</th>\n",
       "      <th>81450</th>\n",
       "      <th>81451</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4933401J01Rik</th>\n",
       "      <td>1070</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm26206</th>\n",
       "      <td>110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>6094</td>\n",
       "      <td>29.669137</td>\n",
       "      <td>24.213824</td>\n",
       "      <td>24.332219</td>\n",
       "      <td>25.701282</td>\n",
       "      <td>25.134798</td>\n",
       "      <td>27.160686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm18956</th>\n",
       "      <td>480</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.752581</td>\n",
       "      <td>7.501131</td>\n",
       "      <td>4.198779</td>\n",
       "      <td>16.820856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37180</th>\n",
       "      <td>2819</td>\n",
       "      <td>48.965217</td>\n",
       "      <td>47.437170</td>\n",
       "      <td>40.461849</td>\n",
       "      <td>60.668955</td>\n",
       "      <td>42.181420</td>\n",
       "      <td>61.579011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>519</td>\n",
       "      <td>1771.814280</td>\n",
       "      <td>2016.853483</td>\n",
       "      <td>1898.834837</td>\n",
       "      <td>1706.615662</td>\n",
       "      <td>1809.600956</td>\n",
       "      <td>1804.595863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Te</th>\n",
       "      <td>69</td>\n",
       "      <td>169.054437</td>\n",
       "      <td>133.658553</td>\n",
       "      <td>132.245743</td>\n",
       "      <td>104.363559</td>\n",
       "      <td>116.835591</td>\n",
       "      <td>117.014649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Cytb</th>\n",
       "      <td>1144</td>\n",
       "      <td>1845.560198</td>\n",
       "      <td>1958.962376</td>\n",
       "      <td>2045.936443</td>\n",
       "      <td>1776.666443</td>\n",
       "      <td>1944.944922</td>\n",
       "      <td>1870.290963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tt</th>\n",
       "      <td>67</td>\n",
       "      <td>580.336128</td>\n",
       "      <td>447.357172</td>\n",
       "      <td>442.628477</td>\n",
       "      <td>456.785281</td>\n",
       "      <td>511.373686</td>\n",
       "      <td>512.157401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tp</th>\n",
       "      <td>67</td>\n",
       "      <td>2408.394931</td>\n",
       "      <td>2890.615571</td>\n",
       "      <td>2826.012585</td>\n",
       "      <td>2525.753905</td>\n",
       "      <td>2316.221991</td>\n",
       "      <td>2169.137230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54456 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Length        81446        81447        81448        81449  \\\n",
       "4933401J01Rik    1070     0.000000     0.000000     0.000000     0.000000   \n",
       "Gm26206           110     0.000000     0.000000     0.000000     0.000000   \n",
       "Xkr4             6094    29.669137    24.213824    24.332219    25.701282   \n",
       "Gm18956           480     0.000000     0.000000     4.752581     7.501131   \n",
       "Gm37180          2819    48.965217    47.437170    40.461849    60.668955   \n",
       "...               ...          ...          ...          ...          ...   \n",
       "mt-Nd6            519  1771.814280  2016.853483  1898.834837  1706.615662   \n",
       "mt-Te              69   169.054437   133.658553   132.245743   104.363559   \n",
       "mt-Cytb          1144  1845.560198  1958.962376  2045.936443  1776.666443   \n",
       "mt-Tt              67   580.336128   447.357172   442.628477   456.785281   \n",
       "mt-Tp              67  2408.394931  2890.615571  2826.012585  2525.753905   \n",
       "\n",
       "                     81450        81451  \n",
       "4933401J01Rik     0.000000     0.000000  \n",
       "Gm26206           0.000000     0.000000  \n",
       "Xkr4             25.134798    27.160686  \n",
       "Gm18956           4.198779    16.820856  \n",
       "Gm37180          42.181420    61.579011  \n",
       "...                    ...          ...  \n",
       "mt-Nd6         1809.600956  1804.595863  \n",
       "mt-Te           116.835591   117.014649  \n",
       "mt-Cytb        1944.944922  1870.290963  \n",
       "mt-Tt           511.373686   512.157401  \n",
       "mt-Tp          2316.221991  2169.137230  \n",
       "\n",
       "[54456 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_featureCounts = read_feature_counts(\n",
    "    '../raw/RNA_featureCounts.tsv.gz',\n",
    "    clean_regex = '[0-9]{5}',\n",
    "    rename_columns = {\n",
    "        'Geneid': 'gene_name'\n",
    "    }\n",
    ")\n",
    "\n",
    "X = df_featureCounts.loc[:, df_featureCounts.columns.str.startswith('8')].values / df_featureCounts.Length.values[:, None] * 1000\n",
    "factors = X.sum(axis = 0) / 1e6\n",
    "df_featureCounts.loc[:, df_featureCounts.columns.str.startswith('8')] = X / factors\n",
    "# zero_expression_genes = df_featureCounts[\n",
    "#     (df_featureCounts.loc[:, df_featureCounts.columns[df_featureCounts.columns.str.startswith('8')]] == 0).all(axis = 1)\n",
    "# ]\n",
    "# df_featureCounts.drop(\n",
    "#     columns = 'Length',\n",
    "#     index = zero_expression_genes.index\n",
    "# )\n",
    "df_featureCounts.set_index(\n",
    "    'gene_name',\n",
    "    inplace = True\n",
    ")\n",
    "df_featureCounts.index.name = None\n",
    "df_featureCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "crazy-innocent",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>81446</th>\n",
       "      <th>81447</th>\n",
       "      <th>81448</th>\n",
       "      <th>81449</th>\n",
       "      <th>81450</th>\n",
       "      <th>81451</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>3583</td>\n",
       "      <td>ENSMUSG00000109644.2</td>\n",
       "      <td>0.430413</td>\n",
       "      <td>0.355705</td>\n",
       "      <td>0.200148</td>\n",
       "      <td>0.295800</td>\n",
       "      <td>0.246198</td>\n",
       "      <td>0.364364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610006L08Rik</th>\n",
       "      <td>2128</td>\n",
       "      <td>ENSMUSG00000108652.2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>998</td>\n",
       "      <td>ENSMUSG00000007777.10</td>\n",
       "      <td>6.510933</td>\n",
       "      <td>5.957220</td>\n",
       "      <td>7.262461</td>\n",
       "      <td>10.436020</td>\n",
       "      <td>8.559581</td>\n",
       "      <td>5.558127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>1803</td>\n",
       "      <td>ENSMUSG00000086714.2</td>\n",
       "      <td>1.251554</td>\n",
       "      <td>1.448534</td>\n",
       "      <td>2.135169</td>\n",
       "      <td>3.813021</td>\n",
       "      <td>2.110855</td>\n",
       "      <td>2.137491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>619</td>\n",
       "      <td>ENSMUSG00000043644.5</td>\n",
       "      <td>9.021717</td>\n",
       "      <td>6.975433</td>\n",
       "      <td>7.106079</td>\n",
       "      <td>7.876659</td>\n",
       "      <td>9.057562</td>\n",
       "      <td>7.501791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s88</th>\n",
       "      <td>118</td>\n",
       "      <td>ENSMUSG00000065107.4</td>\n",
       "      <td>9.405298</td>\n",
       "      <td>9.806197</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.858370</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s92</th>\n",
       "      <td>119</td>\n",
       "      <td>ENSMUSG00000119822.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s93</th>\n",
       "      <td>119</td>\n",
       "      <td>ENSMUSG00000119639.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoZ159</th>\n",
       "      <td>87</td>\n",
       "      <td>ENSMUSG00002075734.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoZ196</th>\n",
       "      <td>35</td>\n",
       "      <td>ENSMUSG00002074855.1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54456 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Length                gene_id     81446     81447     81448  \\\n",
       "0610005C13Rik    3583   ENSMUSG00000109644.2  0.430413  0.355705  0.200148   \n",
       "0610006L08Rik    2128   ENSMUSG00000108652.2  0.000000  0.000000  0.000000   \n",
       "0610009B22Rik     998  ENSMUSG00000007777.10  6.510933  5.957220  7.262461   \n",
       "0610009E02Rik    1803   ENSMUSG00000086714.2  1.251554  1.448534  2.135169   \n",
       "0610009L18Rik     619   ENSMUSG00000043644.5  9.021717  6.975433  7.106079   \n",
       "...               ...                    ...       ...       ...       ...   \n",
       "n-R5s88           118   ENSMUSG00000065107.4  9.405298  9.806197  0.000000   \n",
       "n-R5s92           119   ENSMUSG00000119822.1  0.000000  0.000000  0.000000   \n",
       "n-R5s93           119   ENSMUSG00000119639.1  0.000000  0.000000  0.000000   \n",
       "snoZ159            87   ENSMUSG00002075734.1  0.000000  0.000000  0.000000   \n",
       "snoZ196            35   ENSMUSG00002074855.1  0.000000  0.000000  0.000000   \n",
       "\n",
       "                   81449     81450     81451  \n",
       "0610005C13Rik   0.295800  0.246198  0.364364  \n",
       "0610006L08Rik   0.000000  0.000000  0.000000  \n",
       "0610009B22Rik  10.436020  8.559581  5.558127  \n",
       "0610009E02Rik   3.813021  2.110855  2.137491  \n",
       "0610009L18Rik   7.876659  9.057562  7.501791  \n",
       "...                  ...       ...       ...  \n",
       "n-R5s88         7.858370  0.000000  0.000000  \n",
       "n-R5s92         0.000000  0.000000  0.000000  \n",
       "n-R5s93         0.000000  0.000000  0.000000  \n",
       "snoZ159         0.000000  0.000000  0.000000  \n",
       "snoZ196         0.000000  0.000000  0.000000  \n",
       "\n",
       "[54456 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_salmon = pd.read_csv(\n",
    "    '../raw/RNA_salmon_quant_tpm.tsv.gz',\n",
    "    sep = '\\t',\n",
    "    compression = 'gzip'\n",
    ")\n",
    "df_salmon.set_index(\n",
    "    'gene_name',\n",
    "    inplace = True\n",
    ")\n",
    "df_salmon.index.name = None\n",
    "df_salmon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "eleven-radical",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsgAAAFgCAYAAACmDI9oAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAYk0lEQVR4nO3de7SddX3n8c8Xcg/hMoBCtJIRBRG1XkpFWmaqWLRe6gIvpTp2KdYOWHVBx6mjw4xaq9XRKs5QWmu1tF1aFPFSlTpq1RmQqgV1lmLxWkSRm0KAhIQk8Js/np3yS3JyTsBz9j45eb3W2sucZ+/zPN8Tc37nzXOevXe11gIAAAz2mvQAAAAwnwhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZBaMqmpV9aBJzwEwblV1ZFV9rapuq6qXT3oe2N0JZADY/f1+ks+11la11v7nvd1JVX2+qn57Fuea6hj7VtXZVXV1Va2rqu+NPj5ojo/7gqq6ZC6PwcIhkAFg93dYkismPURVLZrh/iVJ/iHJ0UmenGTfJI9L8tMkvzjnA8IuEsjMS1X1yqq6ZvTrwm9V1QlV9YtV9Y9Vtbaqrq2qc0aL7VSff15VnVtVfz86Q/GFqjpkdJbi5qq6sqoe1T3+qNGZk7VVdUVV/fp2+/qTqvrEaJ4vVdXh4/h7AJhJVX02yeOTnDNa746sqreOztBeX1V/VlXLR489oKo+XlU3jtbCj1fV/Uf3vSHJ8d1+zqmqNaPL1xZ1x/vXs8yjs7JfqKq3V9VPk7y2qpbu7PhJfivJA5Kc1Fr7ZmvtrtbaDa2117fWLhrtc7r1eJsz3NufFR7NelpVfWf0+X9Sg6OS/FmSx42+trWjxz+lqr45WtuvqapXzPr/QeyWBDLzTlUdmeSlSY5pra1K8qQkVyW5M8mZSQ7KcMbhhCQvmWZXz0ly1ujxdyT5xyRfGX38wSRvGx1vcZKPJflUkvskeVmS947m2OqUJK9LckCS7yZ5w8/+lQL87FprT0hycZKXttb2SXJakiOSPDLJg5LcL8l/Hz18ryR/meGM8wOSbEhyzmg//7XfT2vtpbs4wmOTfD/JfTOsjW+a5vhPTPLJ1tq6qXa0i+vxTJ6W5Jgkj8jwc+BJrbV/zvD38o+jr23/0WPfneQ/jn7WPCzJZ+/BcVjABDLz0Z1JliZ5aFUtbq1d1Vr7Xmvt8tbaF1trW1prVyV5Z5J/P81+Pjz6nI1JPpxkY2vtr1trdyZ5f5KtZ5CPTbJPkje11ja11j6b5ONJfnO7fX25tbYlyXszLPwA80pVVZLfSXJma+2m1tptSd6Y4T/y01r7aWvtwtba7aP73pDp19Fd8ePW2v8arY8bpzt+kgOTXDvNvnZlPZ7Jm1pra1trVyf5XKZfrzdn+Fmzb2vt5tbaV+7BcVjABDLzTmvtu0nOSPLaJDdU1flVtbqqjhj9OvC6qro1w6I73ZM6ru/+vGGKj/cZ/Xl1kh+21u7q7v9BhrMeW13X/fn27nMB5pODk6xIcvnoEoO1ST452p6qWlFV76yqH4zW0f+bZP+q2vtnOOYPd/X4Ga41PnSafe3KejyTe7JePzPJU5L8oKr+T1U97h4chwVMIDMvtdbe11r75Qy/BmxJ3pzkT5NcmeTBrbV9k7w6Sc3C4X6c5Oeqqv9+eECSa2Zh3wDj9JMMJwCObq3tP7rtN7r0Ikn+U5Ijkzx2tI7+u9H2rWtp225/60f/u6Lbdsh2j+k/Z6bjfybJk6pq5U7mn2k9Xj/DLNPZ/mtLa+2fWmvPyHA5x0eSfOAe7I8FTCAz74yeYPKEqlqa4dd1G5LclWRVkluTrKuqhyQ5fZYO+aUMZxl+v6oWV9WvJHl6kvNnaf8AYzE68/quJG+vqvskSVXdr6qeNHrIqgxr6tqq+jdJXrPdLq5P8sBufzdmiNP/UFV7V9WpSXb6JOVdOP7fZDjjfGFVPaSq9qqqA6vq1VX1lMy8Hn8tycmjM+EPSvKie/DXc32S+9foyd1VtaSqnldV+7XWNmf4+XLXtHtgjyGQmY+WZniSx08y/KrsPkleleQVSZ6b5LYMC/D7Z+NgrbVNGRbgXxsd89wkv9Vau3I29g8wZq/M8GTiL44uo/hMhrPGSXJ2kuUZ1rovZrj8ofeOJM8avcLF1tdTfnGS/5zh8oijk1x6b4/fWrsjwxP1rkzy6QxR+uUMl8t9aRfW47cn2ZQhdv8qw3NCdtVnM7wU3nVV9ZPRtucnuWo052lJnncP9scCVq3t8BsHAADYYzmDDAAAHYEMAAAdgQwAAB2BDAAAnUUzP+SeO+igg9qaNWvmYtcAC8Lll1/+k9bawTM/8t6zFgNMb2dr8ZwE8po1a3LZZZfNxa4BFoSq+sFcH8NaDDC9na3FLrEAAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5AhjHasiW56abkrrsmPQnsRGvJ2rXJxo2TngRgYgQyjEFryRvfmBx4YHLoocnBByd/+qeTngq2c/HFyUMektznPsn++yfPe15y222Tngpg7ObkraaBbb3lLckb3pDcfvvw8U03Ja94RbLPPsnznz/Z2SBJ8q1vJU9+8t3/SJPkwguT669PPvOZyc0FMAHOIMMcay1505u27Y5k+Ph1r5vMTLCDs89O7rhj22133JFcemny7W9PZCSASRHIMMfuuCO55Zap77vmmvHOAjv1zW8md9654/YlS5Krrhr7OACTJJBhji1dmqxePfV9Rx013llgp447bvjHur2NG5Ojjx7/PAATJJBhjlUN1yCvWLHt9uXLh+0wL7z85cM/0r26HwsrViTPfW5yv/tNbi6ACRDIMAannJK8//3Jz/98su++ybHHJhddlJxwwqQng5FDD00uuyw5+eRkv/2SBzwg+YM/SN71rklPBjB2XsUCxuRpTxtuMG898IHJBRdMegqAiXMGGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAIDdwpVXJp/7XHLzzXN7HIEMAMC8duONybHHJo9+dHLSScnq1cnrXjd3xxPIAADMa89+dvKVryQbNiS33JJs3Ji85S3JhRfOzfEEMgAA89aPf5x86UvJ5s3bbl+/Pnnb2+bmmAIZAIB56+abk0WLpr7vxhvn5pgCGQCAeeuII5LFi3fcvmRJ8vSnz80xBTIAAPPW4sXJuecmK1YkVcO2pUuTAw9MXvnKuTnmTk5YAwDA/HDKKcnhhw/XHF99dXLiiclLXzpE8lwQyAAAzHvHHJP87d+O51gusQAAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAIDOokkPwMxaSy67LLnuuuSYY5JDDpn0RAAAC5dAnud+9KPkxBOTH/4w2WuvZNOm5GUvS9785qRq0tMBACw8LrGY504+Ofn2t5N165Jbb002bkzOPTf54AcnPRkAwMIkkOexq65KvvGN5M47t92+fn3yjndMZCQAgAVPIM9jt96aLNrJRTBr1451FACAPYZAnseOOmrqQF66dLj0AgCA2SeQ57HFi5N3vStZsSLZe+9h2/LlyerVyZlnTnY2AICFyqtYzHPPfGZyxBHJOeckV1+dPPnJyamnJqtWTXoyAICFSSDvBh7+8OSd75z0FAAAewaXWAAAQEcgAwBARyADAEBHIAMAQEcgAwBARyADAEBHIAMAQEcgAwBARyADAEBHIAMAQEcgAwBARyADAEBHIAMAQEcgAwBARyADAEBHIAMAQEcgAwBARyADAEBHIAMAQGeXArmqFk+x7aDZHweAnbEWA4zHtIFcVY+vqh8lubaqPlVVa7q7PzWnkwGQxFoMMG4znUH+H0me1Fo7KMmfJ/l0VR07uq/mdDIAtrIWA4zRohnuX9JauyJJWmsfrKp/TvKhqnplkjbn0wGQWIsBxmqmQN5cVYe01q5LktbaFVV1QpKPJzl8zqcDILEWA4zVTJdY/Jck9+03tNZ+lORXkrxpjmYCYFvWYoAxmvYMcmvtMzu5a1WSTbM/DgDbsxYDjNcuvw5yVR1cVS+pqouTfD7bnc0AYO5ZiwHm3rRnkKtqVZKTkzw3yRFJPpTk37bW7j+G2QCItRhg3GZ6kt4NSb6c5Kwkl7TWWlWdNPdjAdCxFgOM0UyXWLwqydIk5yZ5VVV5tjTA+FmLAcZo2kBurZ3dWjs2yTNGmz6SZHVVvbKqjpjr4QCwFgOM2y49Sa+19v3W2htbaw9P8gtJ9k1y0ZxOBsA2rMUA4zFtIFfVg6rql/ptrbVvJPn7JE+ey8EAGFiLAcZrpjPIZye5dYrttyR5+6xPA8BUzo61GGBsZgrk+7bWvr79xtG2NXMyEQDbsxYDjNFMgbz/NPctn8U5ANi5/ae5z1oMMMtmCuTLqurF22+sqt9OcvncjATAdqzFAGM00xuFnJHkw1X1vNy9CP9CkiVJvEg9wHicEWsxwNhMG8itteuTHFdVj0/ysNHmT7TWPjvnkwGQxFoMMG7TBnJVLUtyWpIHJfl6kne31raMYzAABtZigPGa6Rrkv8rwa7yvJ/m1JG+d84kA2J61GGCMZroG+aGjd2xKVb07yZfnfiQAtmMtBhijmc4gb976B7/OA5gYazHAGM10Bvnnq2rruzdVkuWjjytJa63tO6fTAZBYiwHGaqZXsdh7XIMAMDVr8fjcckvy53+efPKTyc/9XPLylyePfvSkpwLGbaYzyACwR7jppiGGb7gh2bAh2Wuv5IILkve8J/mN35j0dMA4zXQNMgDsEf74j5PrrhviOEnuuiu5/fbktNOSzZun/1xgYRHIAJDkox9N7rhjx+1btiRXXDH+eYDJEcgAkOSAA6bevm5d8upXJ2vXjnUcYIIEMgAkOfPMZOXKqe/7h39Inva08c4DTI5ABoAkJ500RPLeU7xmyKZNyVe/mnzjG+OfCxg/gQwASaqS178+Of74qe9ftCj5/vfHOxMwGQIZADpPeEKybNmO2zdtSh7xiPHPA4yfQAaAzmmnJfvsM7wO8lYrViTPfGayZs3ExgLGSCADQOfgg5PLLhveHOSAA4Z31HvNa5Lzzpv0ZMC4eCc9ANjOYYcl73vfpKcAJsUZZAAA6MyfM8jf/W7yl3+Z/PSnw4tNPuUp214ABgAAYzA/Avn9709e+MLhze63bEne+97kuOOST3xieF0dAAAYk8mfol2/PnnRi5ING4Y4Tob39fzCF5IPfGCyswEAsMeZfCBfcsnUb1u0fv1wJhkAAMZo8oG8bFnS2tT3rVgx3lkAANjjTT6Qf+mXkqVLd9y+cmXy4hePfx4AAPZokw/kRYuSj3882W+/ZNWq4azxsmXJ6acnJ5446ekAANjDzI+XiHjsY5Nrr00uuihZuzY54QTv5wkAwETMj0BOkuXLhze6BwCACZr8JRYAADCPCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkgHujteEGwIIjkAHuiWuvTU46KVmyZLj9+q8nP/7xpKcCYBbNnzcKAZjvNm1KHve45Jprki1bhm0XXTS8G+h3v5ssXTrZ+QCYFc4gA+yqj30suemmu+M4Se68M1m7NvnIRyY1FQCzTCAD7KpvfStZv37H7evWJVdeOf55AJgTAhlgVx19dLJy5Y7bV60a7gNgQRDIALvqqU9NDjkkWbz47m2LFycHH5w84xmTmwuAWSWQAXbVokXJpZcmp5ySrFiRLF+ePOc5yRe/uG00A7Bb8yoWAPfEQQclf/3Xww2ABckZZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoCGQAAOgIZAAA6AhkAADoLJr0AAAAcI/cckvyF3+RXHJJcuSRyemnJ4cdNmu7F8gAAOw+rr02ecxjkrVrkw0bkiVLknPOST71qeS442blEC6xAABg93HWWcmNNw5xnCSbNiXr1yennjprhxDIMIX165OLLko+/enh+w4AmCc+9rFky5Ydt//LvwzhPAtcYgHb+cAHkhe+MFk0+u7Ya6/k7/4uOf74yc4FACRZuXLnIbxs2awcwhlk6Hz/+8kLXpDcfnty663Dbe3a5KlPTdatm/R0AEBe8pJkxYptty1enPzqryarVs3KIQQydP7mb6b+rU1rw1lkAGDCzjwzefrTh7PFq1YNZ5Qf+tDkvPNm7RAusYDOzTcnmzfvuP3OO4dXlAEAJmzRouT885PvfCf52teGl3c75pikavYOMWt7ggXgqU8dXlZx/fod7zvxxPHPAwDsxIMfPNzmgEssoPPEJyYnnDD8tmarlSuHy50OP3xycwEA4+MMMnSqkg99KPnwh5P3vjdZujR50YuGcAYA9gwCGbaz997Js5413ACAPY9LLAAAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5ABAKAjkAEAoCOQAQCgI5CZP266KXn965Pjj0+e//zk8ssnPRHs2b73veR3fzf55V9OzjwzufrqSU8EMBaLJj0AJEluuCF51KOGSN64Mbn00uRDH0rOOy959rMnPR3seb785eQJT0juuCPZsmX4+D3vGb43jz560tMBzClnkJkf/uiPkp/8ZIjjJLnrruT225PTTht+OAPjdfrpyfr1d3//bd6c3HZbcsYZEx0LYBwEMvPDJz6RbNq04/ZNm5Jvf3v888CebMuW5Ktf3XF7a8nFF49/HoAxE8jMDwceOPX2LVuSAw4Y7yywp9t772TZsqnvW7VqvLMATIBAZn74vd9LVq7cdtuiRcljH5sceuhkZoI9VVVy6qk7RvLy5cOT9gAWOIHM/PCsZw3XNi5bluy3X7JiRfLIRyYXXDDpyWDP9Na3JieeePf35LJlyUknJWedNenJAOacV7FgfqhK/vAPh5eS+upXk9Wrk4c+dNJTwZ5r2bLkox9Nrroq+c53kqOOSu5//0lPBTAWApn55cADkyc+cdJTAFutWTPcAPYgLrEAAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCAjkAGAICOQAYAgI5ABgCATrXWZn+nVTcm+cGs7xhg4TistXbwXB7AWgwwoynX4jkJZAAA2F25xAIAADoCGQAAOgIZAAA6ApndUlXdWVVfq6pvVNUFVbVitP2Qqjq/qr5XVZdX1UVVdUT3eWdU1caq2q/bdmBVfa6q1lXVOZP4egB2R9ZiFiqBzO5qQ2vtka21hyXZlOS0qqokH07y+dba4a21xyR5VZL7dp/3m0n+KcnJ3baNSf5bkleMZ3SABcNazIIkkFkILk7yoCSPT7K5tfZnW+9orf2/1trFSVJVhyfZJ8lZGRbnrY9Z31q7JMPiDMC9Yy1mwRDI7NaqalGSX0vy9SQPS3L5NA8/Jcn5GRbxI6vqvtM8FoBdZC1moRHI7K6WV9XXklyW5Ook796Fz/nNJOe31u5KcmGSZ8/deAB7BGsxC9KiSQ8A99KG1toj+w1VdUWSZ0314Kp6eJIHJ/n0cHlcliT5lySeCAJw71mLWZCcQWYh+WySpVX1O1s3VNUjqur4DGcsXttaWzO6rU6yuqoOm9SwAAuUtZjdnreaZrdUVetaa/tMsX11krOTPCbDEz2uSnJGkv+d5CmttSu7x74tyfWttTdX1VVJ9s1wNmNtkhNba9+c0y8CYDdnLWahEsgAANBxiQUAAHQEMgAAdAQyAAB0BDIAAHQEMgAAdAQyAAB0BDIAAHT+PzGrJzwPDYDTAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x360 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "pca = PCA(n_components = 2)\n",
    "\n",
    "for key, df, ax in zip(['salmon', 'featureCounts'], [df_salmon, df_featureCounts], axs):\n",
    "    pca_transform = pca.fit_transform(\n",
    "        df.loc[:, df.columns[df.columns.str.startswith('8')]].values.T\n",
    "    )\n",
    "    ax.scatter(\n",
    "        pca_transform[:, 0],\n",
    "        pca_transform[:, 1],\n",
    "        color = ['b', 'r', 'b', 'r', 'b', 'r']\n",
    "    )\n",
    "    ax.set_title(key)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.set_xlabel('PCA1')\n",
    "    ax.set_ylabel('PCA2')\n",
    "    \n",
    "fig.set_figwidth(10)\n",
    "fig.set_figheight(5)\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "empty-female",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key1, df in zip(\n",
    "    ['salmon', 'featureCounts'],\n",
    "    [df_salmon, df_featureCounts]\n",
    "):\n",
    "    path = f'../processed/{key1}_rna'\n",
    "    if not os.path.exists(path):\n",
    "        os.mkdir(path)\n",
    "        \n",
    "    for key2, sampleids in sample_type.items():\n",
    "        sampleids = [str(i) for i in sampleids]\n",
    "        tmp = df.loc[:, sampleids]\n",
    "        tmp = tmp.loc[(df.loc[:, sampleids] > 0).sum(axis = 1) == 3, :]\n",
    "        tmp.to_csv(\n",
    "            path + f'/{key2}.expression.tsv',\n",
    "            sep = '\\t'\n",
    "        )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-nova]",
   "language": "python",
   "name": "conda-env-.conda-nova-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
