{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "toxic-radar",
   "metadata": {},
   "source": [
    "# Converting read quantifications to Anndata objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "effective-france",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from scipy.sparse import csr_matrix\n",
    "import gzip\n",
    "import re\n",
    "\n",
    "def read_feature_counts(\n",
    "    filename, \n",
    "    clean_regex = None,\n",
    "    rename_columns = None\n",
    "):\n",
    "    with gzip.open(filename) as file:\n",
    "        #getting rid of comment\n",
    "        file.readline()\n",
    "        \n",
    "        tab = pd.read_csv(\n",
    "            file,\n",
    "            sep = '\\t'\n",
    "        )\n",
    "        \n",
    "    exclude_cols = ['Chr', 'Start', 'End', 'Strand']\n",
    "    if clean_regex:\n",
    "        new_columns = []\n",
    "        for col in tab.columns:\n",
    "            match = re.findall(clean_regex, col)\n",
    "            if match:\n",
    "                new_columns.append(match[0])\n",
    "                \n",
    "            else:\n",
    "                new_columns.append(col)\n",
    "                \n",
    "        tab.columns = new_columns\n",
    "    \n",
    "    if rename_columns:\n",
    "        tab.rename(\n",
    "            columns = rename_columns,\n",
    "            inplace = True\n",
    "        )\n",
    "    \n",
    "    return tab.loc[:, ~tab.columns.isin(exclude_cols)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "middle-practice",
   "metadata": {},
   "source": [
    "## Gathering metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "informative-beatles",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_description</th>\n",
       "      <th>sample_genotype</th>\n",
       "      <th>sample_treatment</th>\n",
       "      <th>sample_sex</th>\n",
       "      <th>sample_litter</th>\n",
       "      <th>sample_mouseID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162194</th>\n",
       "      <td>plate 1 control (200 unlabeled cells)</td>\n",
       "      <td>mix</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162195</th>\n",
       "      <td>1 b1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162196</th>\n",
       "      <td>1 c1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162197</th>\n",
       "      <td>1 d1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162198</th>\n",
       "      <td>1 e1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162567</th>\n",
       "      <td>4 b12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162568</th>\n",
       "      <td>4 c12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162569</th>\n",
       "      <td>4 d12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162570</th>\n",
       "      <td>4 e12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162571</th>\n",
       "      <td>4 f12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>376 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               sample_description sample_genotype  \\\n",
       "sample_id                                                           \n",
       "162194     plate 1 control (200 unlabeled cells)              mix   \n",
       "162195                                       1 b1              wt   \n",
       "162196                                       1 c1              wt   \n",
       "162197                                       1 d1              wt   \n",
       "162198                                       1 e1              wt   \n",
       "...                                           ...             ...   \n",
       "162567                                      4 b12           empty   \n",
       "162568                                      4 c12           empty   \n",
       "162569                                      4 d12           empty   \n",
       "162570                                      4 e12           empty   \n",
       "162571                                      4 f12           empty   \n",
       "\n",
       "          sample_treatment sample_sex sample_litter sample_mouseID  \n",
       "sample_id                                                           \n",
       "162194              prep 1          m       G1/55 A       GNF1/480  \n",
       "162195              prep 1          m       G1/55 A       GNF1/480  \n",
       "162196              prep 1          m       G1/55 A       GNF1/480  \n",
       "162197              prep 1          m       G1/55 A       GNF1/480  \n",
       "162198              prep 1          m       G1/55 A       GNF1/480  \n",
       "...                    ...        ...           ...            ...  \n",
       "162567              prep 2        NaN           NaN            NaN  \n",
       "162568              prep 2        NaN           NaN            NaN  \n",
       "162569              prep 2        NaN           NaN            NaN  \n",
       "162570              prep 2        NaN           NaN            NaN  \n",
       "162571              prep 2        NaN           NaN            NaN  \n",
       "\n",
       "[376 rows x 6 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(\n",
    "    '../meta/Samples_info_scRNASeq_MADMcells_Knaus_Novarino.csv'\n",
    ")\n",
    "metadata = metadata.loc[:, ['sample_id', 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID']]\n",
    "metadata = metadata.set_index('sample_id')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "retired-designation",
   "metadata": {},
   "source": [
    "## Converting FeatureCounts quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "durable-princess",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>162194</th>\n",
       "      <th>162195</th>\n",
       "      <th>162196</th>\n",
       "      <th>162197</th>\n",
       "      <th>162198</th>\n",
       "      <th>162199</th>\n",
       "      <th>162200</th>\n",
       "      <th>162201</th>\n",
       "      <th>162202</th>\n",
       "      <th>...</th>\n",
       "      <th>162555</th>\n",
       "      <th>162556</th>\n",
       "      <th>162557</th>\n",
       "      <th>162559</th>\n",
       "      <th>162560</th>\n",
       "      <th>162561</th>\n",
       "      <th>162562</th>\n",
       "      <th>162563</th>\n",
       "      <th>162564</th>\n",
       "      <th>162565</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4933401J01Rik</th>\n",
       "      <td>1070</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm26206</th>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>6094</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm18956</th>\n",
       "      <td>480</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37180</th>\n",
       "      <td>2819</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>519</td>\n",
       "      <td>193</td>\n",
       "      <td>29</td>\n",
       "      <td>8</td>\n",
       "      <td>34</td>\n",
       "      <td>61</td>\n",
       "      <td>79</td>\n",
       "      <td>159</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>37</td>\n",
       "      <td>43</td>\n",
       "      <td>53</td>\n",
       "      <td>69</td>\n",
       "      <td>137</td>\n",
       "      <td>97</td>\n",
       "      <td>128</td>\n",
       "      <td>290</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Te</th>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Cytb</th>\n",
       "      <td>1144</td>\n",
       "      <td>3944</td>\n",
       "      <td>870</td>\n",
       "      <td>440</td>\n",
       "      <td>511</td>\n",
       "      <td>629</td>\n",
       "      <td>1047</td>\n",
       "      <td>2217</td>\n",
       "      <td>319</td>\n",
       "      <td>729</td>\n",
       "      <td>...</td>\n",
       "      <td>3684</td>\n",
       "      <td>1684</td>\n",
       "      <td>1203</td>\n",
       "      <td>1517</td>\n",
       "      <td>4216</td>\n",
       "      <td>3993</td>\n",
       "      <td>4811</td>\n",
       "      <td>5539</td>\n",
       "      <td>6595</td>\n",
       "      <td>2376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tt</th>\n",
       "      <td>67</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tp</th>\n",
       "      <td>67</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>31</td>\n",
       "      <td>32</td>\n",
       "      <td>14</td>\n",
       "      <td>63</td>\n",
       "      <td>16</td>\n",
       "      <td>29</td>\n",
       "      <td>37</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54456 rows Ã— 349 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Length  162194  162195  162196  162197  162198  162199  162200  \\\n",
       "gene_name                                                                       \n",
       "4933401J01Rik    1070       0       0       0       0       0       0       0   \n",
       "Gm26206           110       0       0       0       0       0       0       0   \n",
       "Xkr4             6094       2       0       0       0       0       0       0   \n",
       "Gm18956           480       0       0       0       0       0       0       0   \n",
       "Gm37180          2819       0       0       0       0       0       0       0   \n",
       "...               ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "mt-Nd6            519     193      29       8      34      61      79     159   \n",
       "mt-Te              69       1       1       0       0       2       4       0   \n",
       "mt-Cytb          1144    3944     870     440     511     629    1047    2217   \n",
       "mt-Tt              67       1       0       0       0       0       1       0   \n",
       "mt-Tp              67      30       1       2       5       2       5       9   \n",
       "\n",
       "               162201  162202  ...  162555  162556  162557  162559  162560  \\\n",
       "gene_name                      ...                                           \n",
       "4933401J01Rik       0       0  ...       0       0       0       0       0   \n",
       "Gm26206             0       0  ...       0       0       0       0       0   \n",
       "Xkr4                0       0  ...       0       0       0       0       0   \n",
       "Gm18956             0       0  ...       0       0       0       0       0   \n",
       "Gm37180             0       0  ...       0       0       0       0       0   \n",
       "...               ...     ...  ...     ...     ...     ...     ...     ...   \n",
       "mt-Nd6              9       9  ...      66      37      43      53      69   \n",
       "mt-Te               0       0  ...       2       0       1       0       0   \n",
       "mt-Cytb           319     729  ...    3684    1684    1203    1517    4216   \n",
       "mt-Tt               0       0  ...       0       0       0       0       3   \n",
       "mt-Tp               2       3  ...      29      31      32      14      63   \n",
       "\n",
       "               162561  162562  162563  162564  162565  \n",
       "gene_name                                              \n",
       "4933401J01Rik       0       0       0       0       0  \n",
       "Gm26206             0       0       0       0       0  \n",
       "Xkr4                0       0       0       0       0  \n",
       "Gm18956             0       0       0       0       0  \n",
       "Gm37180             0       0       0       0       0  \n",
       "...               ...     ...     ...     ...     ...  \n",
       "mt-Nd6            137      97     128     290      75  \n",
       "mt-Te               0       3       0       1       0  \n",
       "mt-Cytb          3993    4811    5539    6595    2376  \n",
       "mt-Tt               5       0       5       0       1  \n",
       "mt-Tp              16      29      37      27       1  \n",
       "\n",
       "[54456 rows x 349 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sctab = read_feature_counts(\n",
    "    '../raw/scRNA_featureCounts.tsv.gz',\n",
    "    clean_regex = '[0-9]{6}',\n",
    "    rename_columns = {\n",
    "        'Geneid': 'gene_name'\n",
    "    }\n",
    ")\n",
    "sctab = sctab.set_index('gene_name')\n",
    "sctab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "defensive-inside",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_genotype' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_treatment' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_sex' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_litter' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_mouseID' as categorical\n"
     ]
    }
   ],
   "source": [
    "data = csr_matrix(sctab.loc[:, sctab.columns.str.startswith('1')].values.T)\n",
    "obs = metadata.loc[[int(i) for i in sctab.columns[sctab.columns.str.startswith('1')]], :]\n",
    "var = sctab.loc[:, ['Length']]\n",
    "adata = ad.AnnData(\n",
    "    data,\n",
    "    obs = obs,\n",
    "    var = var\n",
    ")\n",
    "adata.write_h5ad(\n",
    "    '../processed/scRNAseq_featureCounts.h5ad'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "improved-metadata",
   "metadata": {},
   "source": [
    "## Converting Salmon quantification\n",
    "After rerunning the nf-core rnaseq pipeline the Salmon quantification continued to have RefSeq IDs as gene names which indicates that there might be some built in Salmon index or annotation or whatever since for the second run we provided Gencode assembly and annotation to the pipeline which did not change the Salmon output. Thus this quantification should either be done manually (see featureCounts below)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "identified-softball",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copying salmon quantifications to raw directory\n",
    "# !cp ../results/star_salmon/salmon.merged.gene_counts.tsv ../raw/scRNA_salmon_quant_counts.tsv\n",
    "# !gzip ../raw/scRNA_salmon_quant_counts.tsv\n",
    "# !cp ../results/star_salmon/salmon.merged.gene_tpm.tsv ../raw/scRNA_salmon_quant_tpm.tsv\n",
    "# !gzip ../raw/scRNA_salmon_quant_tpm.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "beginning-dollar",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1070</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110</td>\n",
       "      <td>Gm26206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6094</td>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>480</td>\n",
       "      <td>Gm18956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2819</td>\n",
       "      <td>Gm37180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54451</th>\n",
       "      <td>519</td>\n",
       "      <td>mt-Nd6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54452</th>\n",
       "      <td>69</td>\n",
       "      <td>mt-Te</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54453</th>\n",
       "      <td>1144</td>\n",
       "      <td>mt-Cytb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54454</th>\n",
       "      <td>67</td>\n",
       "      <td>mt-Tt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54455</th>\n",
       "      <td>67</td>\n",
       "      <td>mt-Tp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54456 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Length      gene_name\n",
       "0        1070  4933401J01Rik\n",
       "1         110        Gm26206\n",
       "2        6094           Xkr4\n",
       "3         480        Gm18956\n",
       "4        2819        Gm37180\n",
       "...       ...            ...\n",
       "54451     519         mt-Nd6\n",
       "54452      69          mt-Te\n",
       "54453    1144        mt-Cytb\n",
       "54454      67          mt-Tt\n",
       "54455      67          mt-Tp\n",
       "\n",
       "[54456 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# getting gene length from featureCounts\n",
    "gene_length = sctab.loc[:, ['Length']].copy()\n",
    "gene_length['gene_name'] = gene_length.index.to_list()\n",
    "gene_length.reset_index(\n",
    "    inplace = True,\n",
    "    drop = True\n",
    ")\n",
    "gene_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "warming-seeker",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'counts':                Length                gene_id  162194  162195  162196  162197  \\\n",
       " gene_name                                                                      \n",
       " 0610005C13Rik    3583   ENSMUSG00000109644.2   0.000     0.0     0.0     0.0   \n",
       " 0610006L08Rik    2128   ENSMUSG00000108652.2   0.000     0.0     0.0     0.0   \n",
       " 0610009B22Rik     998  ENSMUSG00000007777.10  20.581     0.0     0.0     0.0   \n",
       " 0610009E02Rik    1803   ENSMUSG00000086714.2  25.469     0.0     0.0     0.0   \n",
       " 0610009L18Rik     619   ENSMUSG00000043644.5   7.000     0.0    19.0     0.0   \n",
       " ...               ...                    ...     ...     ...     ...     ...   \n",
       " n-R5s88           118   ENSMUSG00000065107.4   0.000     0.0     0.0     0.0   \n",
       " n-R5s92           119   ENSMUSG00000119822.1   0.000     0.0     0.0     0.0   \n",
       " n-R5s93           119   ENSMUSG00000119639.1   0.000     0.0     0.0     0.0   \n",
       " snoZ159            87   ENSMUSG00002075734.1   0.000     0.0     0.0     0.0   \n",
       " snoZ196            35   ENSMUSG00002074855.1   0.000     0.0     0.0     0.0   \n",
       " \n",
       "                162198  162199  162200  162201  ...  162562  162563  162564  \\\n",
       " gene_name                                      ...                           \n",
       " 0610005C13Rik     0.0     0.0     0.0     0.0  ...     0.0    69.0   0.000   \n",
       " 0610006L08Rik     0.0     0.0     0.0     0.0  ...     0.0     0.0   0.000   \n",
       " 0610009B22Rik     0.0     0.0     0.0     2.0  ...   124.0    73.0   7.000   \n",
       " 0610009E02Rik     0.0     0.0     0.0     0.0  ...     6.0     0.0  30.183   \n",
       " 0610009L18Rik     0.0     0.0     0.0     0.0  ...     0.0     0.0   0.000   \n",
       " ...               ...     ...     ...     ...  ...     ...     ...     ...   \n",
       " n-R5s88           0.0     0.0     0.0     1.0  ...     0.0     0.0   0.000   \n",
       " n-R5s92           0.0     0.0     0.0     0.0  ...     0.0     0.0   0.000   \n",
       " n-R5s93           0.0     0.0     0.0     0.0  ...     0.0     0.0   0.000   \n",
       " snoZ159           0.0     0.0     0.0     0.0  ...     0.0     0.0   0.000   \n",
       " snoZ196           0.0     0.0     0.0     0.0  ...     0.0     0.0   0.000   \n",
       " \n",
       "                 162565  162566  162567  162568  162569  162570  162571  \n",
       " gene_name                                                               \n",
       " 0610005C13Rik    0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " 0610006L08Rik    0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " 0610009B22Rik  128.971     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " 0610009E02Rik    0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " 0610009L18Rik    0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " ...                ...     ...     ...     ...     ...     ...     ...  \n",
       " n-R5s88          0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " n-R5s92          0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " n-R5s93          0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " snoZ159          0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " snoZ196          0.000     0.0     0.0     0.0     0.0     0.0     0.0  \n",
       " \n",
       " [54456 rows x 378 columns],\n",
       " 'tpm':                Length                gene_id     162194  162195      162196  \\\n",
       " gene_name                                                                     \n",
       " 0610005C13Rik    3583   ENSMUSG00000109644.2   0.000000     0.0    0.000000   \n",
       " 0610006L08Rik    2128   ENSMUSG00000108652.2   0.000000     0.0    0.000000   \n",
       " 0610009B22Rik     998  ENSMUSG00000007777.10  41.010005     0.0    0.000000   \n",
       " 0610009E02Rik    1803   ENSMUSG00000086714.2  23.786285     0.0    0.000000   \n",
       " 0610009L18Rik     619   ENSMUSG00000043644.5  19.233833     0.0  368.874221   \n",
       " ...               ...                    ...        ...     ...         ...   \n",
       " n-R5s88           118   ENSMUSG00000065107.4   0.000000     0.0    0.000000   \n",
       " n-R5s92           119   ENSMUSG00000119822.1   0.000000     0.0    0.000000   \n",
       " n-R5s93           119   ENSMUSG00000119639.1   0.000000     0.0    0.000000   \n",
       " snoZ159            87   ENSMUSG00002075734.1   0.000000     0.0    0.000000   \n",
       " snoZ196            35   ENSMUSG00002074855.1   0.000000     0.0    0.000000   \n",
       " \n",
       "                162197  162198  162199  162200      162201  ...      162562  \\\n",
       " gene_name                                                  ...               \n",
       " 0610005C13Rik     0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " 0610006L08Rik     0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " 0610009B22Rik     0.0     0.0     0.0     0.0   15.826775  ...  338.213383   \n",
       " 0610009E02Rik     0.0     0.0     0.0     0.0    0.000000  ...    7.519860   \n",
       " 0610009L18Rik     0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " ...               ...     ...     ...     ...         ...  ...         ...   \n",
       " n-R5s88           0.0     0.0     0.0     0.0  103.621361  ...    0.000000   \n",
       " n-R5s92           0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " n-R5s93           0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " snoZ159           0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " snoZ196           0.0     0.0     0.0     0.0    0.000000  ...    0.000000   \n",
       " \n",
       "                    162563     162564      162565  162566  162567  162568  \\\n",
       " gene_name                                                                  \n",
       " 0610005C13Rik  111.901220   0.000000    0.000000     0.0     0.0     0.0   \n",
       " 0610006L08Rik    0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " 0610009B22Rik  152.466141  12.530801  284.259095     0.0     0.0     0.0   \n",
       " 0610009E02Rik    0.000000  24.892144    0.000000     0.0     0.0     0.0   \n",
       " 0610009L18Rik    0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " ...                   ...        ...         ...     ...     ...     ...   \n",
       " n-R5s88          0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " n-R5s92          0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " n-R5s93          0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " snoZ159          0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " snoZ196          0.000000   0.000000    0.000000     0.0     0.0     0.0   \n",
       " \n",
       "                162569  162570  162571  \n",
       " gene_name                              \n",
       " 0610005C13Rik     0.0     0.0     0.0  \n",
       " 0610006L08Rik     0.0     0.0     0.0  \n",
       " 0610009B22Rik     0.0     0.0     0.0  \n",
       " 0610009E02Rik     0.0     0.0     0.0  \n",
       " 0610009L18Rik     0.0     0.0     0.0  \n",
       " ...               ...     ...     ...  \n",
       " n-R5s88           0.0     0.0     0.0  \n",
       " n-R5s92           0.0     0.0     0.0  \n",
       " n-R5s93           0.0     0.0     0.0  \n",
       " snoZ159           0.0     0.0     0.0  \n",
       " snoZ196           0.0     0.0     0.0  \n",
       " \n",
       " [54456 rows x 378 columns]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sctab_dict = {}\n",
    "for key in ['counts', 'tpm']:\n",
    "    sctab = pd.read_csv(\n",
    "        f'../raw/scRNA_salmon_quant_{key}.tsv.gz', \n",
    "        sep = '\\t',\n",
    "        compression = 'gzip'\n",
    "    )\n",
    "    sctab = sctab.merge(\n",
    "        gene_length,\n",
    "        on = 'gene_name'\n",
    "    )\n",
    "    sctab = sctab.loc[:, ['gene_id', 'gene_name', 'Length'] + sctab.columns[sctab.columns.str.startswith('X')].to_list()]\n",
    "\n",
    "    agg_dict = {\n",
    "        'Length': 'max',\n",
    "        'gene_id': lambda x: ','.join(x) if type(x) == list else x\n",
    "    }\n",
    "    agg_dict.update(\n",
    "        {\n",
    "            column[1:]: 'sum' for column in sctab.columns[sctab.columns.str.startswith('X')]\n",
    "        }\n",
    "    )\n",
    "    \n",
    "    sctab.columns = ['gene_id', 'gene_name', 'Length'] + [s[1:] for s in sctab.columns[sctab.columns.str.startswith('X')]]\n",
    "    sctab = sctab.groupby('gene_name') \\\n",
    "        .agg(agg_dict) \\\n",
    "        .reset_index()\n",
    "\n",
    "    # splitting dataset because it also contains bulk RNA samples that were processed together with the scRNA\n",
    "    sctab.loc[:, ~sctab.columns.str.startswith('1')].to_csv(\n",
    "        f'../raw/RNA_salmon_quant_{key}.tsv.gz',\n",
    "        sep = '\\t',\n",
    "        index = False,\n",
    "        compression = 'gzip'\n",
    "    )\n",
    "    sctab.set_index(\n",
    "        'gene_name', \n",
    "        inplace = True\n",
    "    )\n",
    "    sctab_dict[key] = sctab.loc[:, ~sctab.columns.str.startswith('8')]\n",
    "\n",
    "sctab_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "underlying-andrew",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 376 Ã— 54456\n",
       "    obs: 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID'\n",
       "    var: 'Length', 'gencode_id'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = csr_matrix(sctab_dict['tpm'].loc[:, sctab_dict['tpm'].columns.str.startswith('1')].values.T)\n",
    "obs = metadata\n",
    "var = sctab.loc[:, ['Length', 'gene_id']]\n",
    "var.columns = ['Length', 'gencode_id']\n",
    "var.loc[:, 'gencode_id'] = var.gencode_id.astype('string')\n",
    "adata = ad.AnnData(\n",
    "    data,\n",
    "    obs = obs,\n",
    "    var = var\n",
    ")\n",
    "adata.layers['counts'] = csr_matrix(sctab_dict['counts'].loc[:, sctab_dict['counts'].columns.str.startswith('1')].values.T)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "wrapped-placement",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_genotype' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_treatment' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_sex' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_litter' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_mouseID' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'gencode_id' as categorical\n"
     ]
    }
   ],
   "source": [
    "adata.write_h5ad(\n",
    "    '../processed/scRNAseq_salmon.h5ad'\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-nova]",
   "language": "python",
   "name": "conda-env-.conda-nova-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
