{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "single-approval",
   "metadata": {},
   "source": [
    "# Converting read quantifications to Anndata objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "intense-second",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from scipy.sparse import csr_matrix\n",
    "import gzip\n",
    "import re\n",
    "\n",
    "def read_feature_counts(\n",
    "    filename, \n",
    "    clean_regex = None,\n",
    "    rename_columns = None\n",
    "):\n",
    "    with gzip.open(filename) as file:\n",
    "        #getting rid of comment\n",
    "        file.readline()\n",
    "        \n",
    "        tab = pd.read_csv(\n",
    "            file,\n",
    "            sep = '\\t'\n",
    "        )\n",
    "        \n",
    "    exclude_cols = ['Chr', 'Start', 'End', 'Strand']\n",
    "    if clean_regex:\n",
    "        new_columns = []\n",
    "        for col in tab.columns:\n",
    "            match = re.findall(clean_regex, col)\n",
    "            if match:\n",
    "                new_columns.append(match[0])\n",
    "                \n",
    "            else:\n",
    "                new_columns.append(col)\n",
    "                \n",
    "        tab.columns = new_columns\n",
    "    \n",
    "    if rename_columns:\n",
    "        tab.rename(\n",
    "            columns = rename_columns,\n",
    "            inplace = True\n",
    "        )\n",
    "    \n",
    "    return tab.loc[:, ~tab.columns.isin(exclude_cols)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "informative-content",
   "metadata": {},
   "source": [
    "## Gathering metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "chronic-destiny",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>anno</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>x</th>\n",
       "      <th>strand</th>\n",
       "      <th>y</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3143476</td>\n",
       "      <td>3144545</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000102693.2\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3172239</td>\n",
       "      <td>3172348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064842.3\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869204</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064371.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869205</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064371.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869206</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869207</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1869208</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSMUSG00000064372.1\"; transcript_id ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1869209 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr     anno        type    start      end  x strand  y  \\\n",
       "0        chr1   HAVANA        gene  3143476  3144545  .      +  .   \n",
       "1        chr1   HAVANA  transcript  3143476  3144545  .      +  .   \n",
       "2        chr1   HAVANA        exon  3143476  3144545  .      +  .   \n",
       "3        chr1  ENSEMBL        gene  3172239  3172348  .      +  .   \n",
       "4        chr1  ENSEMBL  transcript  3172239  3172348  .      +  .   \n",
       "...       ...      ...         ...      ...      ... ..    ... ..   \n",
       "1869204  chrM  ENSEMBL  transcript    15289    15355  .      +  .   \n",
       "1869205  chrM  ENSEMBL        exon    15289    15355  .      +  .   \n",
       "1869206  chrM  ENSEMBL        gene    15356    15422  .      -  .   \n",
       "1869207  chrM  ENSEMBL  transcript    15356    15422  .      -  .   \n",
       "1869208  chrM  ENSEMBL        exon    15356    15422  .      -  .   \n",
       "\n",
       "                                               description  \n",
       "0        gene_id \"ENSMUSG00000102693.2\"; gene_type \"TEC...  \n",
       "1        gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "2        gene_id \"ENSMUSG00000102693.2\"; transcript_id ...  \n",
       "3        gene_id \"ENSMUSG00000064842.3\"; gene_type \"snR...  \n",
       "4        gene_id \"ENSMUSG00000064842.3\"; transcript_id ...  \n",
       "...                                                    ...  \n",
       "1869204  gene_id \"ENSMUSG00000064371.1\"; transcript_id ...  \n",
       "1869205  gene_id \"ENSMUSG00000064371.1\"; transcript_id ...  \n",
       "1869206  gene_id \"ENSMUSG00000064372.1\"; gene_type \"Mt_...  \n",
       "1869207  gene_id \"ENSMUSG00000064372.1\"; transcript_id ...  \n",
       "1869208  gene_id \"ENSMUSG00000064372.1\"; transcript_id ...  \n",
       "\n",
       "[1869209 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode = pd.read_csv(\n",
    "    '../data/gencode.vM28.annotation.gtf.gz',\n",
    "    sep = '\\t',\n",
    "    comment = '#',\n",
    "    header = None\n",
    ")\n",
    "gencode.columns = ['chr', 'anno', 'type', 'start', 'end', 'x', 'strand', 'y', 'description']\n",
    "gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "internal-drink",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_description</th>\n",
       "      <th>sample_genotype</th>\n",
       "      <th>sample_treatment</th>\n",
       "      <th>sample_sex</th>\n",
       "      <th>sample_litter</th>\n",
       "      <th>sample_mouseID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162194</th>\n",
       "      <td>plate 1 control (200 unlabeled cells)</td>\n",
       "      <td>mix</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162195</th>\n",
       "      <td>1 b1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162196</th>\n",
       "      <td>1 c1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162197</th>\n",
       "      <td>1 d1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162198</th>\n",
       "      <td>1 e1</td>\n",
       "      <td>wt</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>m</td>\n",
       "      <td>G1/55 A</td>\n",
       "      <td>GNF1/480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162567</th>\n",
       "      <td>4 b12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162568</th>\n",
       "      <td>4 c12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162569</th>\n",
       "      <td>4 d12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162570</th>\n",
       "      <td>4 e12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162571</th>\n",
       "      <td>4 f12</td>\n",
       "      <td>empty</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>376 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               sample_description sample_genotype  \\\n",
       "sample_id                                                           \n",
       "162194     plate 1 control (200 unlabeled cells)              mix   \n",
       "162195                                       1 b1              wt   \n",
       "162196                                       1 c1              wt   \n",
       "162197                                       1 d1              wt   \n",
       "162198                                       1 e1              wt   \n",
       "...                                           ...             ...   \n",
       "162567                                      4 b12           empty   \n",
       "162568                                      4 c12           empty   \n",
       "162569                                      4 d12           empty   \n",
       "162570                                      4 e12           empty   \n",
       "162571                                      4 f12           empty   \n",
       "\n",
       "          sample_treatment sample_sex sample_litter sample_mouseID  \n",
       "sample_id                                                           \n",
       "162194              prep 1          m       G1/55 A       GNF1/480  \n",
       "162195              prep 1          m       G1/55 A       GNF1/480  \n",
       "162196              prep 1          m       G1/55 A       GNF1/480  \n",
       "162197              prep 1          m       G1/55 A       GNF1/480  \n",
       "162198              prep 1          m       G1/55 A       GNF1/480  \n",
       "...                    ...        ...           ...            ...  \n",
       "162567              prep 2        NaN           NaN            NaN  \n",
       "162568              prep 2        NaN           NaN            NaN  \n",
       "162569              prep 2        NaN           NaN            NaN  \n",
       "162570              prep 2        NaN           NaN            NaN  \n",
       "162571              prep 2        NaN           NaN            NaN  \n",
       "\n",
       "[376 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(\n",
    "    '../meta/Samples_info_scRNASeq_MADMcells_Knaus_Novarino.csv'\n",
    ")\n",
    "metadata = metadata.loc[:, ['sample_id', 'sample_description', 'sample_genotype', 'sample_treatment', 'sample_sex', 'sample_litter', 'sample_mouseID']]\n",
    "metadata = metadata.set_index('sample_id')\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "suspended-partition",
   "metadata": {},
   "source": [
    "## Converting Salmon quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "standard-winning",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0610005C13Rik</td>\n",
       "      <td>NR_038165</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610005C13Rik</td>\n",
       "      <td>NR_038166</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0610007P14Rik</td>\n",
       "      <td>NM_021446</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0610009B22Rik</td>\n",
       "      <td>NM_025319</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0610009L18Rik</td>\n",
       "      <td>NR_038126</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35114</th>\n",
       "      <td>l7Rn6</td>\n",
       "      <td>NM_001291287</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35115</th>\n",
       "      <td>l7Rn6</td>\n",
       "      <td>NM_001291288</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35116</th>\n",
       "      <td>l7Rn6</td>\n",
       "      <td>NM_001291289</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35117</th>\n",
       "      <td>l7Rn6</td>\n",
       "      <td>NM_026304</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35118</th>\n",
       "      <td>l7Rn6</td>\n",
       "      <td>NR_111916</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35119 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name          Name  gene_id  pid\n",
       "0      0610005C13Rik     NR_038165        1    1\n",
       "1      0610005C13Rik     NR_038166        1    1\n",
       "2      0610007P14Rik     NM_021446        1    1\n",
       "3      0610009B22Rik     NM_025319        1    1\n",
       "4      0610009L18Rik     NR_038126        1    1\n",
       "...              ...           ...      ...  ...\n",
       "35114          l7Rn6  NM_001291287        1    1\n",
       "35115          l7Rn6  NM_001291288        1    1\n",
       "35116          l7Rn6  NM_001291289        1    1\n",
       "35117          l7Rn6     NM_026304        1    1\n",
       "35118          l7Rn6     NR_111916        1    1\n",
       "\n",
       "[35119 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = pd.read_csv(\n",
    "    '/resources/references/igenomes/Mus_musculus/UCSC/mm10/Annotation/Genes/genes.gtf',\n",
    "    sep = '\\t',\n",
    "    header = None\n",
    ")\n",
    "gtf = gtf \\\n",
    "    .loc[:, 8] \\\n",
    "    .str.split(';', expand = True)\n",
    "\n",
    "gtf = gtf.loc[:, range(4)]\n",
    "gtf.columns = ['gene_id', 'gene_name', 'pid', 'Name']\n",
    "\n",
    "for column in gtf.columns:\n",
    "    gtf.loc[:, column] = gtf.loc[:, column].apply(\n",
    "        lambda x: x.split('\"')[1]\n",
    "    )\n",
    "\n",
    "gtf.loc[:, 'Name'] = gtf.apply(\n",
    "    lambda x: x['Name'] if x['Name'].startswith('N') else x['pid'],\n",
    "    axis = 1\n",
    ")\n",
    "gtf = gtf.groupby(['gene_name', 'Name']) \\\n",
    "    .agg('nunique') \\\n",
    "    .reset_index()\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "spiritual-temperature",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>162194</th>\n",
       "      <th>162206</th>\n",
       "      <th>162260</th>\n",
       "      <th>162211</th>\n",
       "      <th>162253</th>\n",
       "      <th>162246</th>\n",
       "      <th>162244</th>\n",
       "      <th>162220</th>\n",
       "      <th>162249</th>\n",
       "      <th>...</th>\n",
       "      <th>162561</th>\n",
       "      <th>162569</th>\n",
       "      <th>162570</th>\n",
       "      <th>162566</th>\n",
       "      <th>162567</th>\n",
       "      <th>162568</th>\n",
       "      <th>162564</th>\n",
       "      <th>162565</th>\n",
       "      <th>162571</th>\n",
       "      <th>162562</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>1122</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>1185</td>\n",
       "      <td>50.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>177.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>795</td>\n",
       "      <td>19.606</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>65.878</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>57.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.000</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>119.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>619</td>\n",
       "      <td>7.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009O20Rik</th>\n",
       "      <td>2404</td>\n",
       "      <td>41.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53.882</td>\n",
       "      <td>92.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zyx</th>\n",
       "      <td>3399</td>\n",
       "      <td>71.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>70.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzef1</th>\n",
       "      <td>11150</td>\n",
       "      <td>53.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzz3</th>\n",
       "      <td>7607</td>\n",
       "      <td>39.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>...</td>\n",
       "      <td>117.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>89.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>692</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>l7Rn6</th>\n",
       "      <td>2579</td>\n",
       "      <td>36.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24411 rows × 377 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Length  162194  162206  162260  162211  162253  162246  162244  \\\n",
       "gene_name                                                                       \n",
       "0610005C13Rik    1122   0.000     0.0     0.0     0.0     0.0     0.0   0.000   \n",
       "0610007P14Rik    1185  50.000     0.0     3.0     0.0    37.0     0.0   6.000   \n",
       "0610009B22Rik     795  19.606     8.0     1.0     9.0     6.0     0.0  65.878   \n",
       "0610009L18Rik     619   7.000     0.0    15.0     0.0     1.0     0.0   0.000   \n",
       "0610009O20Rik    2404  41.000     0.0     3.0    11.0    13.0     0.0   0.000   \n",
       "...               ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "Zyx              3399  71.000     0.0    23.0     0.0     0.0     0.0   0.000   \n",
       "Zzef1           11150  53.000     0.0     3.0     3.0    17.0     0.0   0.000   \n",
       "Zzz3             7607  39.000     0.0     5.0    39.0    19.0     0.0   0.000   \n",
       "a                 692   0.000     0.0     0.0     0.0     8.0     0.0   0.000   \n",
       "l7Rn6            2579  36.000     0.0     8.0     0.0     5.0     0.0   0.000   \n",
       "\n",
       "               162220  162249  ...  162561  162569  162570  162566  162567  \\\n",
       "gene_name                      ...                                           \n",
       "0610005C13Rik     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "0610007P14Rik     0.0     3.0  ...   177.0     0.0     0.0     0.0     0.0   \n",
       "0610009B22Rik     0.0     0.0  ...    57.0     0.0     0.0     0.0     0.0   \n",
       "0610009L18Rik     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "0610009O20Rik     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "...               ...     ...  ...     ...     ...     ...     ...     ...   \n",
       "Zyx               0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "Zzef1             0.0     2.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "Zzz3              0.0     5.0  ...   117.0     0.0     0.0     0.0     0.0   \n",
       "a                 0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0   \n",
       "l7Rn6             0.0     0.0  ...     1.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "               162568  162564  162565  162571  162562  \n",
       "gene_name                                              \n",
       "0610005C13Rik     0.0   0.000     0.0     0.0     0.0  \n",
       "0610007P14Rik     0.0  31.000     6.0     0.0     0.0  \n",
       "0610009B22Rik     0.0   5.000   124.0     0.0   119.0  \n",
       "0610009L18Rik     0.0   0.000     0.0     0.0     0.0  \n",
       "0610009O20Rik     0.0  53.882    92.0     0.0     0.0  \n",
       "...               ...     ...     ...     ...     ...  \n",
       "Zyx               0.0   0.000    39.0     0.0    70.0  \n",
       "Zzef1             0.0   0.000     6.0     0.0     4.0  \n",
       "Zzz3              0.0   0.000    16.0     0.0    89.0  \n",
       "a                 0.0   0.000     0.0     0.0     0.0  \n",
       "l7Rn6             0.0  11.000     0.0     0.0     0.0  \n",
       "\n",
       "[24411 rows x 377 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sctab = pd.read_csv(\n",
    "    '../raw/scRNA_salmon_quant_NumReads.tsv.gz',\n",
    "    sep = '\\t'\n",
    ")\n",
    "sctab = sctab.merge(\n",
    "    gtf,\n",
    "    on = 'Name',\n",
    "    how = 'left'\n",
    ")\n",
    "agg_dict = {\n",
    "    'Length': 'max'\n",
    "}\n",
    "agg_dict.update(\n",
    "    {\n",
    "        column: 'sum' for column in sctab.columns[sctab.columns.str.startswith('1')]\n",
    "    }\n",
    ")\n",
    "\n",
    "sctab = sctab.groupby('gene_name') \\\n",
    "    .agg(agg_dict) \\\n",
    "    .reset_index()\n",
    "sctab = sctab.set_index('gene_name')\n",
    "sctab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "surgical-fluid",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_genotype' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_treatment' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_sex' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_litter' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_mouseID' as categorical\n"
     ]
    }
   ],
   "source": [
    "data = csr_matrix(sctab.loc[:, sctab.columns.str.startswith('1')].values.T)\n",
    "obs = metadata\n",
    "var = sctab.loc[:, ['Length']]\n",
    "adata = ad.AnnData(\n",
    "    data,\n",
    "    obs = obs,\n",
    "    var = var\n",
    ")\n",
    "adata.write_h5ad(\n",
    "    '../processed/scRNAseq_salmon_NumReads.h5ad'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "occupational-boulder",
   "metadata": {},
   "source": [
    "## Converting FeatureCounts quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "popular-vienna",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>162194</th>\n",
       "      <th>162195</th>\n",
       "      <th>162196</th>\n",
       "      <th>162197</th>\n",
       "      <th>162198</th>\n",
       "      <th>162199</th>\n",
       "      <th>162200</th>\n",
       "      <th>162201</th>\n",
       "      <th>162202</th>\n",
       "      <th>...</th>\n",
       "      <th>162555</th>\n",
       "      <th>162556</th>\n",
       "      <th>162557</th>\n",
       "      <th>162559</th>\n",
       "      <th>162560</th>\n",
       "      <th>162561</th>\n",
       "      <th>162562</th>\n",
       "      <th>162563</th>\n",
       "      <th>162564</th>\n",
       "      <th>162565</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4933401J01Rik</th>\n",
       "      <td>1070</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm26206</th>\n",
       "      <td>110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>6094</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm18956</th>\n",
       "      <td>480</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37180</th>\n",
       "      <td>2819</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>519</td>\n",
       "      <td>193</td>\n",
       "      <td>29</td>\n",
       "      <td>8</td>\n",
       "      <td>34</td>\n",
       "      <td>61</td>\n",
       "      <td>79</td>\n",
       "      <td>159</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>66</td>\n",
       "      <td>37</td>\n",
       "      <td>43</td>\n",
       "      <td>53</td>\n",
       "      <td>69</td>\n",
       "      <td>137</td>\n",
       "      <td>97</td>\n",
       "      <td>128</td>\n",
       "      <td>290</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Te</th>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Cytb</th>\n",
       "      <td>1144</td>\n",
       "      <td>3944</td>\n",
       "      <td>870</td>\n",
       "      <td>440</td>\n",
       "      <td>511</td>\n",
       "      <td>629</td>\n",
       "      <td>1047</td>\n",
       "      <td>2217</td>\n",
       "      <td>319</td>\n",
       "      <td>729</td>\n",
       "      <td>...</td>\n",
       "      <td>3684</td>\n",
       "      <td>1684</td>\n",
       "      <td>1203</td>\n",
       "      <td>1517</td>\n",
       "      <td>4216</td>\n",
       "      <td>3993</td>\n",
       "      <td>4811</td>\n",
       "      <td>5539</td>\n",
       "      <td>6595</td>\n",
       "      <td>2376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tt</th>\n",
       "      <td>67</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Tp</th>\n",
       "      <td>67</td>\n",
       "      <td>30</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>29</td>\n",
       "      <td>31</td>\n",
       "      <td>32</td>\n",
       "      <td>14</td>\n",
       "      <td>63</td>\n",
       "      <td>16</td>\n",
       "      <td>29</td>\n",
       "      <td>37</td>\n",
       "      <td>27</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54456 rows × 349 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Length  162194  162195  162196  162197  162198  162199  162200  \\\n",
       "gene_name                                                                       \n",
       "4933401J01Rik    1070       0       0       0       0       0       0       0   \n",
       "Gm26206           110       0       0       0       0       0       0       0   \n",
       "Xkr4             6094       0       0       0       0       0       0       7   \n",
       "Gm18956           480       1       0       0       0       0       0       0   \n",
       "Gm37180          2819       0       0       0       0       0       0       0   \n",
       "...               ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "mt-Nd6            519     193      29       8      34      61      79     159   \n",
       "mt-Te              69       1       1       0       0       2       4       0   \n",
       "mt-Cytb          1144    3944     870     440     511     629    1047    2217   \n",
       "mt-Tt              67       1       0       0       0       0       1       0   \n",
       "mt-Tp              67      30       1       2       5       2       5       9   \n",
       "\n",
       "               162201  162202  ...  162555  162556  162557  162559  162560  \\\n",
       "gene_name                      ...                                           \n",
       "4933401J01Rik       0      21  ...       0       0       0       0       0   \n",
       "Gm26206             0       0  ...       0       0       0       0       0   \n",
       "Xkr4                0       0  ...      20       0       0       0       0   \n",
       "Gm18956             0       0  ...       0       0       0       0       0   \n",
       "Gm37180             0       0  ...       0       0       0       0       0   \n",
       "...               ...     ...  ...     ...     ...     ...     ...     ...   \n",
       "mt-Nd6              9       9  ...      66      37      43      53      69   \n",
       "mt-Te               0       0  ...       2       0       1       0       0   \n",
       "mt-Cytb           319     729  ...    3684    1684    1203    1517    4216   \n",
       "mt-Tt               0       0  ...       0       0       0       0       3   \n",
       "mt-Tp               2       3  ...      29      31      32      14      63   \n",
       "\n",
       "               162561  162562  162563  162564  162565  \n",
       "gene_name                                              \n",
       "4933401J01Rik       0       0       0       0       0  \n",
       "Gm26206             0       0       0       0       0  \n",
       "Xkr4                0       0       0       0       0  \n",
       "Gm18956             0       0       0       0       0  \n",
       "Gm37180             0       0       0       0       0  \n",
       "...               ...     ...     ...     ...     ...  \n",
       "mt-Nd6            137      97     128     290      75  \n",
       "mt-Te               0       3       0       1       0  \n",
       "mt-Cytb          3993    4811    5539    6595    2376  \n",
       "mt-Tt               5       0       5       0       1  \n",
       "mt-Tp              16      29      37      27       1  \n",
       "\n",
       "[54456 rows x 349 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sctab = read_feature_counts(\n",
    "    '../raw/scRNA_featureCounts.tsv.gz',\n",
    "    clean_regex = '[0-9]{6}',\n",
    "    rename_columns = {\n",
    "        'Geneid': 'gene_name'\n",
    "    }\n",
    ")\n",
    "sctab = sctab.set_index('gene_name')\n",
    "sctab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "embedded-conditions",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_genotype' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_treatment' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_sex' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_litter' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sample_mouseID' as categorical\n"
     ]
    }
   ],
   "source": [
    "data = csr_matrix(sctab.loc[:, sctab.columns.str.startswith('1')].values.T)\n",
    "obs = metadata.loc[[int(i) for i in sctab.columns[sctab.columns.str.startswith('1')]], :]\n",
    "var = sctab.loc[:, ['Length']]\n",
    "adata = ad.AnnData(\n",
    "    data,\n",
    "    obs = obs,\n",
    "    var = var\n",
    ")\n",
    "adata.write_h5ad(\n",
    "    '../processed/scRNAseq_featureCounts_NumReads.h5ad'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "antique-empty",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-nova]",
   "language": "python",
   "name": "conda-env-.conda-nova-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
