{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fitting-friendship",
   "metadata": {},
   "source": [
    "# Prepare metabolomics data for ipath ingestion\n",
    "This notebook is dedicated to preparing the VBCF metabolomics facility results for upload to ipath. However, since the data is provided as excel file containing several sheets one needs to do some preprocessing in excel in order to make the data parsable. In particular these preprocessing/reformating steps apply only to the 'Results HILIC' and 'Results RP' sheets since these are the main data sources.\n",
    "\n",
    "1. Reformat the log2 fold change cells to contain numbers with at least 3 decimal places\n",
    "2. Reformat the normalized peak area cells to contain numbers with at least 10 decimal places\n",
    "3. Save each sheet as UTF-8 decoded CSV file\n",
    "\n",
    "The rest will be done by `parse_metabolomics_results` function which will return two `pandas.DataFrame` objects containing all the data you need (one containing everything that has a KEGG ID and one containing everything without a KEGG ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fifth-banner",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define some functions for data ingestion\n",
    "import pandas as pd\n",
    "import csv\n",
    "import requests\n",
    "\n",
    "def get_unambiguous_column_names(column_names):\n",
    "    series = pd.Series(column_names)\n",
    "    series[series.duplicated(keep = 'first')] = series[series.duplicated(keep = 'first')].apply(\n",
    "        lambda x: x + '.1'\n",
    "    )\n",
    "    return series.to_list()\n",
    "\n",
    "def parse_metabolomics_results(filename, delimiter = ',', quotechar = '\"'):\n",
    "    with open(filename) as csvfile:\n",
    "        data = []\n",
    "        entryid = None\n",
    "        csvreader = csv.reader(\n",
    "            csvfile,\n",
    "            delimiter = delimiter,\n",
    "            quotechar = quotechar\n",
    "        )\n",
    "        \n",
    "        # get column names and make them unambiguous\n",
    "        names = []\n",
    "        for i in range(2):\n",
    "            names = csvreader.__next__()\n",
    "        \n",
    "        names = get_unambiguous_column_names(names)\n",
    "        \n",
    "        add_names = [\n",
    "                'Name2', \n",
    "                'Molecular Weight', \n",
    "                'RT [min]2', \n",
    "                'DeltaMass [ppm]', \n",
    "                'DBID', \n",
    "                'Reference List Name', \n",
    "                'mzLogic Score', \n",
    "                'ChemSpider ID', \n",
    "                'KEGG', 'HMDB', \n",
    "                'Mass List Search Results ID'\n",
    "        ]\n",
    "        add_names_set = set(add_names) # faster check\n",
    "        series = None\n",
    "        tmp_names = []\n",
    "        for line in csvreader:\n",
    "            if line[0]:\n",
    "                entryid = line[0]\n",
    "                series = pd.Series(\n",
    "                    {k: v for k, v in zip(names, line)}\n",
    "                )\n",
    "\n",
    "            elif line[1] == 'Name':\n",
    "                tmp_names = [name for name in line if name]\n",
    "            \n",
    "            elif line[1] == series.Name:\n",
    "                for k, v in zip(\n",
    "                    tmp_names,\n",
    "                    line[1: 1 + len(tmp_names)]\n",
    "                ):\n",
    "                    if k in add_names_set:\n",
    "                        k = k + '2' if k in {'Name', 'RT [min]'} else k\n",
    "                        series[k] = v if v else None\n",
    "                    \n",
    "                    elif k == 'KEGG ID':\n",
    "                        series['KEGG'] = v if v else None\n",
    "                    \n",
    "                data.append(series.copy())\n",
    "    \n",
    "    return pd.DataFrame(data, columns = names + add_names)\n",
    "\n",
    "def clean_dataframe(df):\n",
    "    kegg_rows = df.KEGG.isna()\n",
    "    no_kegg = df.loc[kegg_rows, :].copy()\n",
    "    df = df.loc[~kegg_rows, :].copy()\n",
    "    df = df.drop_duplicates(subset = ['ID', 'KEGG'])\n",
    "    return df, no_kegg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "local-butter",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(372,)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Calc. MW</th>\n",
       "      <th>RT [min]</th>\n",
       "      <th>(P40, KO) / (P40, WT)</th>\n",
       "      <th>(P2, KO) / (P2, WT)</th>\n",
       "      <th>(E14_5, KO) / (E14_5, WT)</th>\n",
       "      <th>(P40, WT) / (P2, WT)</th>\n",
       "      <th>(P40, WT) / (E14_5, WT)</th>\n",
       "      <th>(P2, WT) / (E14_5, WT)</th>\n",
       "      <th>...</th>\n",
       "      <th>Molecular Weight</th>\n",
       "      <th>RT [min]2</th>\n",
       "      <th>DeltaMass [ppm]</th>\n",
       "      <th>DBID</th>\n",
       "      <th>Reference List Name</th>\n",
       "      <th>mzLogic Score</th>\n",
       "      <th>ChemSpider ID</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>HMDB</th>\n",
       "      <th>Mass List Search Results ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A001_HILIC</td>\n",
       "      <td>Pyruvic acid</td>\n",
       "      <td>88.0160</td>\n",
       "      <td>3.93</td>\n",
       "      <td>-1.32</td>\n",
       "      <td>-0.31</td>\n",
       "      <td>-0.55</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.27</td>\n",
       "      <td>-0.30</td>\n",
       "      <td>...</td>\n",
       "      <td>88.0160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>B003</td>\n",
       "      <td>iHILIC_neg_2020</td>\n",
       "      <td>68.35</td>\n",
       "      <td>1031</td>\n",
       "      <td>C00022</td>\n",
       "      <td>HMDB0000243</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A002_HILIC</td>\n",
       "      <td>L-(+)-Alanine</td>\n",
       "      <td>89.0477</td>\n",
       "      <td>14.83</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>0.23</td>\n",
       "      <td>-0.34</td>\n",
       "      <td>-0.57</td>\n",
       "      <td>...</td>\n",
       "      <td>89.0477</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.34</td>\n",
       "      <td>A072</td>\n",
       "      <td>iHILIC_neg_2020</td>\n",
       "      <td>74.93</td>\n",
       "      <td>5735</td>\n",
       "      <td>C00041</td>\n",
       "      <td>HMDB0000161</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A003_HILIC</td>\n",
       "      <td>Sarcosine</td>\n",
       "      <td>89.0477</td>\n",
       "      <td>13.63</td>\n",
       "      <td>0.31</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.22</td>\n",
       "      <td>-4.07</td>\n",
       "      <td>-4.17</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>...</td>\n",
       "      <td>89.0477</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.71</td>\n",
       "      <td>A026</td>\n",
       "      <td>iHILIC_pos_2020</td>\n",
       "      <td>36.31</td>\n",
       "      <td>1057</td>\n",
       "      <td>C00213</td>\n",
       "      <td>HMDB0000271</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A004_HILIC</td>\n",
       "      <td>L-(+)-Lactic acid</td>\n",
       "      <td>90.0317</td>\n",
       "      <td>4.48</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.21</td>\n",
       "      <td>-0.22</td>\n",
       "      <td>-0.28</td>\n",
       "      <td>-0.78</td>\n",
       "      <td>-0.49</td>\n",
       "      <td>...</td>\n",
       "      <td>90.0317</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>B002</td>\n",
       "      <td>iHILIC_neg_2020</td>\n",
       "      <td>93.20</td>\n",
       "      <td>96860</td>\n",
       "      <td>C00186</td>\n",
       "      <td>HMDB0000190</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A005_HILIC</td>\n",
       "      <td>Glycerin</td>\n",
       "      <td>92.0474</td>\n",
       "      <td>7.10</td>\n",
       "      <td>0.21</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-0.20</td>\n",
       "      <td>1.64</td>\n",
       "      <td>1.58</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>...</td>\n",
       "      <td>92.0473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.17</td>\n",
       "      <td>D034</td>\n",
       "      <td>iHILIC_neg_2020</td>\n",
       "      <td>None</td>\n",
       "      <td>733</td>\n",
       "      <td>C00116</td>\n",
       "      <td>HMDB0000131</td>\n",
       "      <td>473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>B180_HILIC</td>\n",
       "      <td>α,α-Trehalose</td>\n",
       "      <td>342.1162</td>\n",
       "      <td>15.63</td>\n",
       "      <td>1.90</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.07</td>\n",
       "      <td>-0.95</td>\n",
       "      <td>-3.74</td>\n",
       "      <td>-2.79</td>\n",
       "      <td>...</td>\n",
       "      <td>342.1162</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C01083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>B183_HILIC</td>\n",
       "      <td>Uridine 5'-diphosphate (UDP)</td>\n",
       "      <td>404.0022</td>\n",
       "      <td>18.51</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>-2.59</td>\n",
       "      <td>-5.55</td>\n",
       "      <td>-2.96</td>\n",
       "      <td>...</td>\n",
       "      <td>404.0022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>B184_HILIC</td>\n",
       "      <td>Adenosine diphosphate (ADP)</td>\n",
       "      <td>427.0296</td>\n",
       "      <td>17.32</td>\n",
       "      <td>-0.17</td>\n",
       "      <td>-0.19</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>-0.58</td>\n",
       "      <td>-2.30</td>\n",
       "      <td>-1.72</td>\n",
       "      <td>...</td>\n",
       "      <td>427.0294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>B186_HILIC</td>\n",
       "      <td>Uridine 5'-diphosphoglucuronic acid</td>\n",
       "      <td>580.0345</td>\n",
       "      <td>20.93</td>\n",
       "      <td>0.12</td>\n",
       "      <td>-0.19</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-2.12</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>...</td>\n",
       "      <td>580.0343</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00167</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>B187_HILIC</td>\n",
       "      <td>Guanosine 5'-diphospho-D-mannose</td>\n",
       "      <td>605.0774</td>\n",
       "      <td>19.93</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>1.54</td>\n",
       "      <td>1.46</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>...</td>\n",
       "      <td>605.0772</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00096</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>154 rows × 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID                                 Name  Calc. MW RT [min]  \\\n",
       "0    A001_HILIC                         Pyruvic acid   88.0160     3.93   \n",
       "1    A002_HILIC                        L-(+)-Alanine   89.0477    14.83   \n",
       "3    A003_HILIC                            Sarcosine   89.0477    13.63   \n",
       "5    A004_HILIC                    L-(+)-Lactic acid   90.0317     4.48   \n",
       "6    A005_HILIC                             Glycerin   92.0474     7.10   \n",
       "..          ...                                  ...       ...      ...   \n",
       "234  B180_HILIC                        α,α-Trehalose  342.1162    15.63   \n",
       "237  B183_HILIC         Uridine 5'-diphosphate (UDP)  404.0022    18.51   \n",
       "238  B184_HILIC          Adenosine diphosphate (ADP)  427.0296    17.32   \n",
       "240  B186_HILIC  Uridine 5'-diphosphoglucuronic acid  580.0345    20.93   \n",
       "241  B187_HILIC     Guanosine 5'-diphospho-D-mannose  605.0774    19.93   \n",
       "\n",
       "    (P40, KO) / (P40, WT) (P2, KO) / (P2, WT) (E14_5, KO) / (E14_5, WT)  \\\n",
       "0                   -1.32               -0.31                     -0.55   \n",
       "1                    0.00               -0.05                     -0.04   \n",
       "3                    0.31                0.04                      0.22   \n",
       "5                    1.08                0.21                     -0.22   \n",
       "6                    0.21               -0.06                     -0.20   \n",
       "..                    ...                 ...                       ...   \n",
       "234                  1.90                0.60                      0.07   \n",
       "237                 -0.03               -0.04                     -0.03   \n",
       "238                 -0.17               -0.19                     -0.04   \n",
       "240                  0.12               -0.19                      0.01   \n",
       "241                 -0.05               -0.01                     -0.03   \n",
       "\n",
       "    (P40, WT) / (P2, WT) (P40, WT) / (E14_5, WT) (P2, WT) / (E14_5, WT)  ...  \\\n",
       "0                   0.03                   -0.27                  -0.30  ...   \n",
       "1                   0.23                   -0.34                  -0.57  ...   \n",
       "3                  -4.07                   -4.17                  -0.09  ...   \n",
       "5                  -0.28                   -0.78                  -0.49  ...   \n",
       "6                   1.64                    1.58                  -0.06  ...   \n",
       "..                   ...                     ...                    ...  ...   \n",
       "234                -0.95                   -3.74                  -2.79  ...   \n",
       "237                -2.59                   -5.55                  -2.96  ...   \n",
       "238                -0.58                   -2.30                  -1.72  ...   \n",
       "240                -1.59                   -2.12                  -0.54  ...   \n",
       "241                 1.54                    1.46                  -0.08  ...   \n",
       "\n",
       "    Molecular Weight RT [min]2 DeltaMass [ppm]  DBID Reference List Name  \\\n",
       "0            88.0160       NaN           -0.11  B003     iHILIC_neg_2020   \n",
       "1            89.0477       NaN            0.34  A072     iHILIC_neg_2020   \n",
       "3            89.0477       NaN            0.71  A026     iHILIC_pos_2020   \n",
       "5            90.0317       NaN           -0.10  B002     iHILIC_neg_2020   \n",
       "6            92.0473       NaN            0.17  D034     iHILIC_neg_2020   \n",
       "..               ...       ...             ...   ...                 ...   \n",
       "234         342.1162       NaN           -0.19   NaN                 NaN   \n",
       "237         404.0022       NaN            0.03   NaN                 NaN   \n",
       "238         427.0294       NaN            0.31   NaN                 NaN   \n",
       "240         580.0343       NaN            0.28   NaN                 NaN   \n",
       "241         605.0772       NaN            0.38   NaN                 NaN   \n",
       "\n",
       "    mzLogic Score ChemSpider ID    KEGG         HMDB  \\\n",
       "0           68.35          1031  C00022  HMDB0000243   \n",
       "1           74.93          5735  C00041  HMDB0000161   \n",
       "3           36.31          1057  C00213  HMDB0000271   \n",
       "5           93.20         96860  C00186  HMDB0000190   \n",
       "6            None           733  C00116  HMDB0000131   \n",
       "..            ...           ...     ...          ...   \n",
       "234           NaN           NaN  C01083          NaN   \n",
       "237           NaN           NaN  C00015          NaN   \n",
       "238           NaN           NaN  C00008          NaN   \n",
       "240           NaN           NaN  C00167          NaN   \n",
       "241           NaN           NaN  C00096          NaN   \n",
       "\n",
       "    Mass List Search Results ID  \n",
       "0                           101  \n",
       "1                            57  \n",
       "3                            26  \n",
       "5                           109  \n",
       "6                           473  \n",
       "..                          ...  \n",
       "234                         NaN  \n",
       "237                         NaN  \n",
       "238                         NaN  \n",
       "240                         NaN  \n",
       "241                         NaN  \n",
       "\n",
       "[154 rows x 73 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ingest hilic results and print total shape for comparison with cleaned data (aka only compounds with KEGG ids remain)\n",
    "hilic = parse_metabolomics_results('../raw/Results_Untargeted_Metabolomics_E14-P40_LK_Nova_results_HILIC.csv')\n",
    "print(hilic.ID.unique().shape)\n",
    "hilic, no_kegg_hilic = clean_dataframe(hilic)\n",
    "hilic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "unsigned-showcase",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Calc. MW</th>\n",
       "      <th>RT [min]</th>\n",
       "      <th>(P40, KO) / (P40, WT)</th>\n",
       "      <th>(P2, KO) / (P2, WT)</th>\n",
       "      <th>(E14_5, KO) / (E14_5, WT)</th>\n",
       "      <th>(P40, WT) / (P2, WT)</th>\n",
       "      <th>(P40, WT) / (E14_5, WT)</th>\n",
       "      <th>(P2, WT) / (E14_5, WT)</th>\n",
       "      <th>...</th>\n",
       "      <th>Molecular Weight</th>\n",
       "      <th>RT [min]2</th>\n",
       "      <th>DeltaMass [ppm]</th>\n",
       "      <th>DBID</th>\n",
       "      <th>Reference List Name</th>\n",
       "      <th>mzLogic Score</th>\n",
       "      <th>ChemSpider ID</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>HMDB</th>\n",
       "      <th>Mass List Search Results ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>A033_HILIC</td>\n",
       "      <td>6-Oxo-pipecolinic acid</td>\n",
       "      <td>143.0583</td>\n",
       "      <td>3.85</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.77</td>\n",
       "      <td>-0.19</td>\n",
       "      <td>-1.59</td>\n",
       "      <td>-4.09</td>\n",
       "      <td>-2.50</td>\n",
       "      <td>...</td>\n",
       "      <td>143.0582</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.48</td>\n",
       "      <td>A111</td>\n",
       "      <td>iHILIC_neg_2020</td>\n",
       "      <td>39.46</td>\n",
       "      <td>2282737</td>\n",
       "      <td>None</td>\n",
       "      <td>HMDB0061705</td>\n",
       "      <td>578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>B123_HILIC</td>\n",
       "      <td>2-Hydroxycaproic acid</td>\n",
       "      <td>132.0786</td>\n",
       "      <td>2.50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>-0.25</td>\n",
       "      <td>-0.25</td>\n",
       "      <td>-1.65</td>\n",
       "      <td>-1.87</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>...</td>\n",
       "      <td>132.0786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>B127_HILIC</td>\n",
       "      <td>6-Aminonicotinic acid</td>\n",
       "      <td>138.0430</td>\n",
       "      <td>6.88</td>\n",
       "      <td>-0.68</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>0.48</td>\n",
       "      <td>-0.62</td>\n",
       "      <td>-1.10</td>\n",
       "      <td>...</td>\n",
       "      <td>138.0429</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.74</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>B130_HILIC</td>\n",
       "      <td>DL-Stachydrine</td>\n",
       "      <td>143.0947</td>\n",
       "      <td>9.20</td>\n",
       "      <td>0.18</td>\n",
       "      <td>-0.41</td>\n",
       "      <td>-0.31</td>\n",
       "      <td>1.01</td>\n",
       "      <td>-3.60</td>\n",
       "      <td>-4.61</td>\n",
       "      <td>...</td>\n",
       "      <td>143.0946</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>B139_HILIC</td>\n",
       "      <td>1-Methylguanine</td>\n",
       "      <td>165.0652</td>\n",
       "      <td>11.62</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.25</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.80</td>\n",
       "      <td>...</td>\n",
       "      <td>165.0651</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>c392_HILIC</td>\n",
       "      <td>PS(16:1(9Z)/18:1(9Z))</td>\n",
       "      <td>759.5050</td>\n",
       "      <td>2.13</td>\n",
       "      <td>0.63</td>\n",
       "      <td>0.14</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>-0.81</td>\n",
       "      <td>-3.47</td>\n",
       "      <td>-2.66</td>\n",
       "      <td>...</td>\n",
       "      <td>759.5050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>c393_HILIC</td>\n",
       "      <td>PS(18:2(9Z,12Z)/18:2(9Z,12Z))</td>\n",
       "      <td>783.5051</td>\n",
       "      <td>2.12</td>\n",
       "      <td>0.78</td>\n",
       "      <td>0.10</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>-2.19</td>\n",
       "      <td>-3.06</td>\n",
       "      <td>-0.87</td>\n",
       "      <td>...</td>\n",
       "      <td>783.5050</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td>c394_HILIC</td>\n",
       "      <td>Phosphatidylserine</td>\n",
       "      <td>791.5676</td>\n",
       "      <td>1.98</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.84</td>\n",
       "      <td>-0.63</td>\n",
       "      <td>1.50</td>\n",
       "      <td>-1.48</td>\n",
       "      <td>-2.98</td>\n",
       "      <td>...</td>\n",
       "      <td>791.5676</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td>c395_HILIC</td>\n",
       "      <td>1-oleoyl-2-arachidonoyl-sn-glycero-3-phospho-L...</td>\n",
       "      <td>809.5200</td>\n",
       "      <td>2.11</td>\n",
       "      <td>0.80</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>-0.69</td>\n",
       "      <td>-1.70</td>\n",
       "      <td>-1.02</td>\n",
       "      <td>...</td>\n",
       "      <td>809.5207</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.86</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>c396_HILIC</td>\n",
       "      <td>(15R,21S)-18,21,24,24-Tetrahydroxy-3-methyl-18...</td>\n",
       "      <td>844.5242</td>\n",
       "      <td>1.97</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.16</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>-1.89</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>0.12</td>\n",
       "      <td>...</td>\n",
       "      <td>844.5231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.29</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>227 rows × 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID                                               Name  Calc. MW  \\\n",
       "50   A033_HILIC                             6-Oxo-pipecolinic acid  143.0583   \n",
       "180  B123_HILIC                              2-Hydroxycaproic acid  132.0786   \n",
       "183  B127_HILIC                              6-Aminonicotinic acid  138.0430   \n",
       "186  B130_HILIC                                     DL-Stachydrine  143.0947   \n",
       "194  B139_HILIC                                    1-Methylguanine  165.0652   \n",
       "..          ...                                                ...       ...   \n",
       "443  c392_HILIC                              PS(16:1(9Z)/18:1(9Z))  759.5050   \n",
       "444  c393_HILIC                      PS(18:2(9Z,12Z)/18:2(9Z,12Z))  783.5051   \n",
       "445  c394_HILIC                                 Phosphatidylserine  791.5676   \n",
       "446  c395_HILIC  1-oleoyl-2-arachidonoyl-sn-glycero-3-phospho-L...  809.5200   \n",
       "447  c396_HILIC  (15R,21S)-18,21,24,24-Tetrahydroxy-3-methyl-18...  844.5242   \n",
       "\n",
       "    RT [min] (P40, KO) / (P40, WT) (P2, KO) / (P2, WT)  \\\n",
       "50      3.85                  0.16                0.77   \n",
       "180     2.50                  0.50               -0.25   \n",
       "183     6.88                 -0.68               -0.42   \n",
       "186     9.20                  0.18               -0.41   \n",
       "194    11.62                  0.32                0.24   \n",
       "..       ...                   ...                 ...   \n",
       "443     2.13                  0.63                0.14   \n",
       "444     2.12                  0.78                0.10   \n",
       "445     1.98                  0.95                0.84   \n",
       "446     2.11                  0.80                0.06   \n",
       "447     1.97                  0.74                0.16   \n",
       "\n",
       "    (E14_5, KO) / (E14_5, WT) (P40, WT) / (P2, WT) (P40, WT) / (E14_5, WT)  \\\n",
       "50                      -0.19                -1.59                   -4.09   \n",
       "180                     -0.25                -1.65                   -1.87   \n",
       "183                     -0.42                 0.48                   -0.62   \n",
       "186                     -0.31                 1.01                   -3.60   \n",
       "194                      0.39                 0.25                    1.05   \n",
       "..                        ...                  ...                     ...   \n",
       "443                     -0.18                -0.81                   -3.47   \n",
       "444                     -0.09                -2.19                   -3.06   \n",
       "445                     -0.63                 1.50                   -1.48   \n",
       "446                     -0.06                -0.69                   -1.70   \n",
       "447                     -0.18                -1.89                   -1.77   \n",
       "\n",
       "    (P2, WT) / (E14_5, WT)  ... Molecular Weight RT [min]2 DeltaMass [ppm]  \\\n",
       "50                   -2.50  ...         143.0582       NaN            0.48   \n",
       "180                  -0.21  ...         132.0786       NaN           -0.42   \n",
       "183                  -1.10  ...         138.0429       NaN            0.74   \n",
       "186                  -4.61  ...         143.0946       NaN            0.46   \n",
       "194                   0.80  ...         165.0651       NaN            0.72   \n",
       "..                     ...  ...              ...       ...             ...   \n",
       "443                  -2.66  ...         759.5050       NaN            0.01   \n",
       "444                  -0.87  ...         783.5050       NaN            0.02   \n",
       "445                  -2.98  ...         791.5676       NaN           -0.09   \n",
       "446                  -1.02  ...         809.5207       NaN           -0.86   \n",
       "447                   0.12  ...         844.5231       NaN            1.29   \n",
       "\n",
       "     DBID Reference List Name mzLogic Score ChemSpider ID  KEGG         HMDB  \\\n",
       "50   A111     iHILIC_neg_2020         39.46       2282737  None  HMDB0061705   \n",
       "180   NaN                 NaN           NaN           NaN  None          NaN   \n",
       "183   NaN                 NaN           NaN           NaN  None          NaN   \n",
       "186   NaN                 NaN           NaN           NaN  None          NaN   \n",
       "194   NaN                 NaN           NaN           NaN  None          NaN   \n",
       "..    ...                 ...           ...           ...   ...          ...   \n",
       "443   NaN                 NaN          None           NaN   NaN          NaN   \n",
       "444   NaN                 NaN          None           NaN   NaN          NaN   \n",
       "445   NaN                 NaN          None           NaN   NaN          NaN   \n",
       "446   NaN                 NaN          None           NaN   NaN          NaN   \n",
       "447   NaN                 NaN          None           NaN   NaN          NaN   \n",
       "\n",
       "    Mass List Search Results ID  \n",
       "50                          578  \n",
       "180                         NaN  \n",
       "183                         NaN  \n",
       "186                         NaN  \n",
       "194                         NaN  \n",
       "..                          ...  \n",
       "443                         NaN  \n",
       "444                         NaN  \n",
       "445                         NaN  \n",
       "446                         NaN  \n",
       "447                         NaN  \n",
       "\n",
       "[227 rows x 73 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_kegg_hilic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "robust-application",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the ingested data to files\n",
    "# as raw values\n",
    "import itertools as it\n",
    "conditions = ['WT', 'KO']\n",
    "timepoints = ['E14_5', 'P2', 'P40']\n",
    "def write_raw_values(df, timepoints, conditions, prefix):\n",
    "    for sample in it.product(timepoints, conditions):\n",
    "        sample_cols = hilic.columns[\n",
    "            df.columns.str.match(\n",
    "                '^S[0-9]{2}_' + '_'.join(sample)\n",
    "            )\n",
    "        ].to_list()\n",
    "\n",
    "        df.loc[:, ['KEGG'] + sample_cols].to_csv(\n",
    "            '_'.join([prefix, *sample]) + '.tsv',\n",
    "            sep = '\\t',\n",
    "            header = False,\n",
    "            index = False\n",
    "        )\n",
    "\n",
    "# in a format for ingestion in iPATH\n",
    "# this is obsolete because it only considers compounds which is not visually appealing in iPATH\n",
    "columns_wt = ['(P2, WT) / (E14_5, WT)', '(P40, WT) / (P2, WT)']\n",
    "columns_ko = ['(P2, KO) / (E14_5, KO)', '(P40, KO) / (P2, KO)']\n",
    "def write_foldchanges(df, columns, condition, prefix):\n",
    "    # writing base values for t0\n",
    "#     tmp = pd.DataFrame(\n",
    "#         get_ipath_selection(\n",
    "#             hilic, \n",
    "#             'KEGG', \n",
    "#             '#000000', \n",
    "#             20\n",
    "#         )\n",
    "#     )\n",
    "#     tmp.to_csv(\n",
    "#         '_'.join([prefix, 'base', condition]) + '.tsv',\n",
    "#         sep = '\\t',\n",
    "#         header = False,\n",
    "#         index = False\n",
    "#     )\n",
    "    \n",
    "    for column in columns:\n",
    "        suffix = column[1:-1] \\\n",
    "            .replace(', ', '_') \\\n",
    "            .replace(') / (', '_')\n",
    "        \n",
    "        df.loc[:, ['KEGG', column]].to_csv(\n",
    "            '_'.join([prefix, suffix]) + '.tsv',\n",
    "            sep = '\\t',\n",
    "            header = False,\n",
    "            index = False\n",
    "        )\n",
    "    \n",
    "# write_raw_values(\n",
    "#     hilic,\n",
    "#     timepoints,\n",
    "#     conditions,\n",
    "#     '../raw/metabolomics_hilic'\n",
    "# )\n",
    "for condition, columns in zip(\n",
    "    conditions, \n",
    "    [columns_wt, columns_ko]\n",
    "):\n",
    "    write_foldchanges(\n",
    "        hilic,\n",
    "        columns,\n",
    "        condition,\n",
    "        '../raw/metabolomics_hilic'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "official-coordination",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(206,)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Calc. MW</th>\n",
       "      <th>RT [min]</th>\n",
       "      <th>(P40, KO) / (P40, WT)</th>\n",
       "      <th>(P2, KO) / (P2, WT)</th>\n",
       "      <th>(E14_5, KO) / (E14_5, WT)</th>\n",
       "      <th>(P40, WT) / (P2, WT)</th>\n",
       "      <th>(P40, WT) / (E14_5, WT)</th>\n",
       "      <th>(P2, WT) / (E14_5, WT)</th>\n",
       "      <th>...</th>\n",
       "      <th>Molecular Weight</th>\n",
       "      <th>RT [min]2</th>\n",
       "      <th>DeltaMass [ppm]</th>\n",
       "      <th>DBID</th>\n",
       "      <th>Reference List Name</th>\n",
       "      <th>mzLogic Score</th>\n",
       "      <th>ChemSpider ID</th>\n",
       "      <th>KEGG</th>\n",
       "      <th>HMDB</th>\n",
       "      <th>Mass List Search Results ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A001_RP</td>\n",
       "      <td>L-(+)-Alanine</td>\n",
       "      <td>89.0478</td>\n",
       "      <td>3.86</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.15</td>\n",
       "      <td>-1.90</td>\n",
       "      <td>-2.05</td>\n",
       "      <td>...</td>\n",
       "      <td>89.0477</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.09</td>\n",
       "      <td>A072</td>\n",
       "      <td>RP_pos_2020</td>\n",
       "      <td>93.43</td>\n",
       "      <td>5735</td>\n",
       "      <td>C00041</td>\n",
       "      <td>HMDB0000161</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A002_RP</td>\n",
       "      <td>L-a-Amino-n-butyric acid</td>\n",
       "      <td>103.0633</td>\n",
       "      <td>4.05</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.46</td>\n",
       "      <td>-0.18</td>\n",
       "      <td>1.73</td>\n",
       "      <td>3.06</td>\n",
       "      <td>1.33</td>\n",
       "      <td>...</td>\n",
       "      <td>103.0633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.17</td>\n",
       "      <td>E19</td>\n",
       "      <td>RP_pos_2020</td>\n",
       "      <td>89.90</td>\n",
       "      <td>72524</td>\n",
       "      <td>C02356</td>\n",
       "      <td>HMDB0000452</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A003_RP</td>\n",
       "      <td>Choline</td>\n",
       "      <td>103.0997</td>\n",
       "      <td>3.82</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.55</td>\n",
       "      <td>0.17</td>\n",
       "      <td>-0.38</td>\n",
       "      <td>...</td>\n",
       "      <td>103.0997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.15</td>\n",
       "      <td>B018</td>\n",
       "      <td>RP_pos_2020</td>\n",
       "      <td>None</td>\n",
       "      <td>299</td>\n",
       "      <td>C00114</td>\n",
       "      <td>HMDB0000097</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A004_RP</td>\n",
       "      <td>L-Serine</td>\n",
       "      <td>105.0426</td>\n",
       "      <td>3.74</td>\n",
       "      <td>-0.08</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-0.53</td>\n",
       "      <td>...</td>\n",
       "      <td>105.0426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.35</td>\n",
       "      <td>A075</td>\n",
       "      <td>RP_neg_2020</td>\n",
       "      <td>91.03</td>\n",
       "      <td>5736</td>\n",
       "      <td>C00065</td>\n",
       "      <td>HMDB0000187</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A005_RP</td>\n",
       "      <td>Hypotaurine</td>\n",
       "      <td>109.0198</td>\n",
       "      <td>3.86</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.39</td>\n",
       "      <td>-2.54</td>\n",
       "      <td>-4.31</td>\n",
       "      <td>-1.77</td>\n",
       "      <td>...</td>\n",
       "      <td>109.0198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.68</td>\n",
       "      <td>B019</td>\n",
       "      <td>RP_pos_2020</td>\n",
       "      <td>70.96</td>\n",
       "      <td>96959</td>\n",
       "      <td>C00519</td>\n",
       "      <td>HMDB0000965</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>B103_RP</td>\n",
       "      <td>Uridine monophosphate (UMP)</td>\n",
       "      <td>324.0360</td>\n",
       "      <td>5.45</td>\n",
       "      <td>-0.13</td>\n",
       "      <td>0.17</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-1.49</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.50</td>\n",
       "      <td>...</td>\n",
       "      <td>324.0359</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>B104_RP</td>\n",
       "      <td>3'-Adenosine monophosphate (3'-AMP)</td>\n",
       "      <td>347.0631</td>\n",
       "      <td>6.04</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.75</td>\n",
       "      <td>1.74</td>\n",
       "      <td>0.99</td>\n",
       "      <td>...</td>\n",
       "      <td>347.0631</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C01367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>B105_RP</td>\n",
       "      <td>4-Hydroxytamoxifen</td>\n",
       "      <td>387.2198</td>\n",
       "      <td>26.07</td>\n",
       "      <td>0.07</td>\n",
       "      <td>1.94</td>\n",
       "      <td>-0.01</td>\n",
       "      <td>-5.14</td>\n",
       "      <td>-0.56</td>\n",
       "      <td>4.58</td>\n",
       "      <td>...</td>\n",
       "      <td>387.2198</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D06551</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>B107_RP</td>\n",
       "      <td>Adenosine diphosphate (ADP)</td>\n",
       "      <td>427.0298</td>\n",
       "      <td>4.74</td>\n",
       "      <td>-0.11</td>\n",
       "      <td>-0.16</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>-0.75</td>\n",
       "      <td>-3.17</td>\n",
       "      <td>-2.42</td>\n",
       "      <td>...</td>\n",
       "      <td>427.0294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>B110_RP</td>\n",
       "      <td>UDP-N-acetylglucosamine</td>\n",
       "      <td>607.0819</td>\n",
       "      <td>4.72</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.32</td>\n",
       "      <td>-0.73</td>\n",
       "      <td>-1.45</td>\n",
       "      <td>-0.72</td>\n",
       "      <td>...</td>\n",
       "      <td>607.0816</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C00043</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98 rows × 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          ID                                 Name  Calc. MW RT [min]  \\\n",
       "0    A001_RP                        L-(+)-Alanine   89.0478     3.86   \n",
       "1    A002_RP             L-a-Amino-n-butyric acid  103.0633     4.05   \n",
       "2    A003_RP                              Choline  103.0997     3.82   \n",
       "3    A004_RP                             L-Serine  105.0426     3.74   \n",
       "4    A005_RP                          Hypotaurine  109.0198     3.86   \n",
       "..       ...                                  ...       ...      ...   \n",
       "146  B103_RP          Uridine monophosphate (UMP)  324.0360     5.45   \n",
       "147  B104_RP  3'-Adenosine monophosphate (3'-AMP)  347.0631     6.04   \n",
       "148  B105_RP                   4-Hydroxytamoxifen  387.2198    26.07   \n",
       "151  B107_RP          Adenosine diphosphate (ADP)  427.0298     4.74   \n",
       "154  B110_RP              UDP-N-acetylglucosamine  607.0819     4.72   \n",
       "\n",
       "    (P40, KO) / (P40, WT) (P2, KO) / (P2, WT) (E14_5, KO) / (E14_5, WT)  \\\n",
       "0                    0.10                0.14                      0.21   \n",
       "1                    0.48                0.46                     -0.18   \n",
       "2                    0.08                0.27                      0.06   \n",
       "3                   -0.08                0.50                      0.18   \n",
       "4                    0.18                0.48                      0.39   \n",
       "..                    ...                 ...                       ...   \n",
       "146                 -0.13                0.17                     -0.01   \n",
       "147                 -0.07                0.09                      0.05   \n",
       "148                  0.07                1.94                     -0.01   \n",
       "151                 -0.11               -0.16                     -0.07   \n",
       "154                  0.12                0.04                     -0.32   \n",
       "\n",
       "    (P40, WT) / (P2, WT) (P40, WT) / (E14_5, WT) (P2, WT) / (E14_5, WT)  ...  \\\n",
       "0                   0.15                   -1.90                  -2.05  ...   \n",
       "1                   1.73                    3.06                   1.33  ...   \n",
       "2                   0.55                    0.17                  -0.38  ...   \n",
       "3                   0.58                    0.05                  -0.53  ...   \n",
       "4                  -2.54                   -4.31                  -1.77  ...   \n",
       "..                   ...                     ...                    ...  ...   \n",
       "146                -1.49                    0.01                   1.50  ...   \n",
       "147                 0.75                    1.74                   0.99  ...   \n",
       "148                -5.14                   -0.56                   4.58  ...   \n",
       "151                -0.75                   -3.17                  -2.42  ...   \n",
       "154                -0.73                   -1.45                  -0.72  ...   \n",
       "\n",
       "    Molecular Weight RT [min]2 DeltaMass [ppm]  DBID Reference List Name  \\\n",
       "0            89.0477       NaN            1.09  A072         RP_pos_2020   \n",
       "1           103.0633       NaN           -0.17   E19         RP_pos_2020   \n",
       "2           103.0997       NaN           -0.15  B018         RP_pos_2020   \n",
       "3           105.0426       NaN            0.35  A075         RP_neg_2020   \n",
       "4           109.0198       NaN            0.68  B019         RP_pos_2020   \n",
       "..               ...       ...             ...   ...                 ...   \n",
       "146         324.0359       NaN            0.35   NaN                 NaN   \n",
       "147         347.0631       NaN            0.10   NaN                 NaN   \n",
       "148         387.2198       NaN           -0.09   NaN                 NaN   \n",
       "151         427.0294       NaN            0.78   NaN                 NaN   \n",
       "154         607.0816       NaN            0.58   NaN                 NaN   \n",
       "\n",
       "    mzLogic Score ChemSpider ID    KEGG         HMDB  \\\n",
       "0           93.43          5735  C00041  HMDB0000161   \n",
       "1           89.90         72524  C02356  HMDB0000452   \n",
       "2            None           299  C00114  HMDB0000097   \n",
       "3           91.03          5736  C00065  HMDB0000187   \n",
       "4           70.96         96959  C00519  HMDB0000965   \n",
       "..            ...           ...     ...          ...   \n",
       "146           NaN           NaN  C00105          NaN   \n",
       "147           NaN           NaN  C01367          NaN   \n",
       "148           NaN           NaN  D06551          NaN   \n",
       "151           NaN           NaN  C00008          NaN   \n",
       "154           NaN           NaN  C00043          NaN   \n",
       "\n",
       "    Mass List Search Results ID  \n",
       "0                            71  \n",
       "1                           176  \n",
       "2                            72  \n",
       "3                            19  \n",
       "4                            75  \n",
       "..                          ...  \n",
       "146                         NaN  \n",
       "147                         NaN  \n",
       "148                         NaN  \n",
       "151                         NaN  \n",
       "154                         NaN  \n",
       "\n",
       "[98 rows x 73 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ingestion for rp results which are not yet used in the workflow\n",
    "# since I didn't find a way to combine with hilic\n",
    "rp = parse_metabolomics_results('../raw/Results_Untargeted_Metabolomics_E14-P40_LK_Nova_results_RP.csv')\n",
    "print(rp.ID.unique().shape)\n",
    "rp, no_kegg_rp = clean_dataframe(rp)\n",
    "rp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "spatial-desire",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KEGG</th>\n",
       "      <th>EC</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C00022</td>\n",
       "      <td>4.1.2.55</td>\n",
       "      <td>Pyruvic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C00022</td>\n",
       "      <td>4.1.2.58</td>\n",
       "      <td>Pyruvic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C00022</td>\n",
       "      <td>4.1.2.14</td>\n",
       "      <td>Pyruvic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00022</td>\n",
       "      <td>4.1.2.18</td>\n",
       "      <td>Pyruvic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C00022</td>\n",
       "      <td>4.1.2.20</td>\n",
       "      <td>Pyruvic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2977</th>\n",
       "      <td>C00002</td>\n",
       "      <td>2.7.1.186</td>\n",
       "      <td>Adenosine 5'-triphosphate (5'-ATP)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2978</th>\n",
       "      <td>C00002</td>\n",
       "      <td>2.7.1.187</td>\n",
       "      <td>Adenosine 5'-triphosphate (5'-ATP)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2979</th>\n",
       "      <td>C00002</td>\n",
       "      <td>2.7.1.188</td>\n",
       "      <td>Adenosine 5'-triphosphate (5'-ATP)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2980</th>\n",
       "      <td>C05122</td>\n",
       "      <td>2.3.1.65</td>\n",
       "      <td>Taurocholic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2981</th>\n",
       "      <td>C05122</td>\n",
       "      <td>3.5.1.24</td>\n",
       "      <td>Taurocholic acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2982 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        KEGG         EC                                Name\n",
       "0     C00022   4.1.2.55                        Pyruvic acid\n",
       "1     C00022   4.1.2.58                        Pyruvic acid\n",
       "2     C00022   4.1.2.14                        Pyruvic acid\n",
       "3     C00022   4.1.2.18                        Pyruvic acid\n",
       "4     C00022   4.1.2.20                        Pyruvic acid\n",
       "...      ...        ...                                 ...\n",
       "2977  C00002  2.7.1.186  Adenosine 5'-triphosphate (5'-ATP)\n",
       "2978  C00002  2.7.1.187  Adenosine 5'-triphosphate (5'-ATP)\n",
       "2979  C00002  2.7.1.188  Adenosine 5'-triphosphate (5'-ATP)\n",
       "2980  C05122   2.3.1.65                    Taurocholic acid\n",
       "2981  C05122   3.5.1.24                    Taurocholic acid\n",
       "\n",
       "[2982 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preliminary attempt to integrate with COMPASS data via EC numbers\n",
    "# worked but didn't continue as some puzzle pieces are missing\n",
    "# like how to compute a combined score for reaction and compounds etc.\n",
    "import requests\n",
    "def map_compounds_to_ec(compounds):\n",
    "    response = requests.get(\n",
    "        'http://rest.kegg.jp/link/enzyme/' + \n",
    "        '+'.join(compounds)\n",
    "    )\n",
    "    mapping = []\n",
    "    for string in response.text.split('\\n'):\n",
    "        if string:\n",
    "            mapping.append(\n",
    "                [s.split(':')[1] for s in string.split('\\t')]\n",
    "            )\n",
    "    \n",
    "    df = pd.DataFrame(\n",
    "        mapping,\n",
    "        columns = ['KEGG', 'EC']\n",
    "    )\n",
    "    return df\n",
    "\n",
    "kegg2ec = map_compounds_to_ec(\n",
    "    hilic.KEGG.to_list()\n",
    ")\n",
    "kegg2ec = kegg2ec.merge(\n",
    "    hilic.loc[:, ['Name', 'KEGG']],\n",
    "    on = 'KEGG',\n",
    "    how = 'left'\n",
    ")\n",
    "kegg2ec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "taken-emergency",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>logFC</th>\n",
       "      <th>metadata_r_id</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13DAMPPOX_pos</th>\n",
       "      <td>1.016497</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1.4.3.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2DR1PP_pos</th>\n",
       "      <td>1.011031</td>\n",
       "      <td>2DR1PP</td>\n",
       "      <td>3.1.3.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2HBO_neg</th>\n",
       "      <td>1.009523</td>\n",
       "      <td>2HBO</td>\n",
       "      <td>1.1.1.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2HBO_pos</th>\n",
       "      <td>0.999164</td>\n",
       "      <td>2HBO</td>\n",
       "      <td>1.1.1.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2OXOADOXm_pos</th>\n",
       "      <td>0.997916</td>\n",
       "      <td>2OXOADOXm</td>\n",
       "      <td>2.3.1.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1487_pos</th>\n",
       "      <td>1.030900</td>\n",
       "      <td>r1487</td>\n",
       "      <td>6.2.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1488_pos</th>\n",
       "      <td>1.017051</td>\n",
       "      <td>r1488</td>\n",
       "      <td>6.2.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r1492_pos</th>\n",
       "      <td>1.030697</td>\n",
       "      <td>r1492</td>\n",
       "      <td>6.2.1.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_citr(c)_pos</th>\n",
       "      <td>1.019971</td>\n",
       "      <td>sink_citr(c)</td>\n",
       "      <td>2.4.99.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sink_pre_prot(r)_neg</th>\n",
       "      <td>0.999260</td>\n",
       "      <td>sink_pre_prot(r)</td>\n",
       "      <td>2.4.99.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2146 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         logFC     metadata_r_id ec_number\n",
       "13DAMPPOX_pos         1.016497         13DAMPPOX   1.4.3.6\n",
       "2DR1PP_pos            1.011031            2DR1PP  3.1.3.10\n",
       "2HBO_neg              1.009523              2HBO  1.1.1.27\n",
       "2HBO_pos              0.999164              2HBO  1.1.1.27\n",
       "2OXOADOXm_pos         0.997916         2OXOADOXm  2.3.1.61\n",
       "...                        ...               ...       ...\n",
       "r1487_pos             1.030900             r1487   6.2.1.3\n",
       "r1488_pos             1.017051             r1488   6.2.1.3\n",
       "r1492_pos             1.030697             r1492   6.2.1.3\n",
       "sink_citr(c)_pos      1.019971      sink_citr(c)  2.4.99.9\n",
       "sink_pre_prot(r)_neg  0.999260  sink_pre_prot(r)  2.4.99.9\n",
       "\n",
       "[2146 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compass = pd.read_csv(\n",
    "    '../raw/compass_results_neurons.tsv',\n",
    "    sep = '\\t',\n",
    "    index_col = 0\n",
    ")\n",
    "compass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "signal-picking",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>KEGG</th>\n",
       "      <th>EC</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1542</th>\n",
       "      <td>C00020</td>\n",
       "      <td>6.2.1.3</td>\n",
       "      <td>Adenosine 5'-monophosphate (5'-AMP)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2659</th>\n",
       "      <td>C00002</td>\n",
       "      <td>6.2.1.3</td>\n",
       "      <td>Adenosine 5'-triphosphate (5'-ATP)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        KEGG       EC                                 Name\n",
       "1542  C00020  6.2.1.3  Adenosine 5'-monophosphate (5'-AMP)\n",
       "2659  C00002  6.2.1.3   Adenosine 5'-triphosphate (5'-ATP)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg2ec[kegg2ec.EC == '6.2.1.3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "extreme-america",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get KEGG pathways via KEGG API and pickle the results\n",
    "import pickle\n",
    "def parse_gene_record(gene_record):\n",
    "    # [:-1] to get rid of the trailing bracket\n",
    "    try:\n",
    "        idsym, namekeggec = gene_record[:-1].split(\n",
    "            '; ',\n",
    "            maxsplit = 1\n",
    "        )\n",
    "        geneid, symbol = idsym.split()\n",
    "    \n",
    "    except ValueError:\n",
    "        idsym, namekeggec = gene_record[:-1].split(maxsplit = 1)\n",
    "        geneid, symbol = idsym, None\n",
    "        \n",
    "    record = {\n",
    "        'ncbigeneid': geneid,\n",
    "        'genesymbol': symbol,\n",
    "    }\n",
    "    \n",
    "    try:\n",
    "        name, keggec = namekeggec.split(\n",
    "            ' [',\n",
    "            maxsplit = 1\n",
    "        )\n",
    "        for key, val in [s.split(':') for s in keggec.split('] [')]:\n",
    "            record[key] = val\n",
    "    \n",
    "    except ValueError:\n",
    "        name = namekeggec.strip()\n",
    "    \n",
    "    record['name'] = name\n",
    "        \n",
    "    return geneid, record\n",
    "\n",
    "\n",
    "def parse_kegg_pathway(responsetxt):\n",
    "    keywords = {\n",
    "        'ENTRY',\n",
    "        'NAME',\n",
    "        'DESCRIPTION',\n",
    "        'CLASS',\n",
    "        'MODULE',\n",
    "        'GENE',\n",
    "        'COMPOUND'\n",
    "    }\n",
    "    \n",
    "    record_parsers = {\n",
    "        'COMPOUND': lambda x: x.split(maxsplit = 1) if len(x.split(maxsplit = 1)) > 1 else (x.strip(), None),\n",
    "        'GENE': parse_gene_record\n",
    "    }\n",
    "    \n",
    "    current_key, entries = None, None\n",
    "    parse_results = {}\n",
    "    for line in responsetxt.split('\\n'):\n",
    "        if line:\n",
    "            key = line[:12].strip()\n",
    "            \n",
    "        else:\n",
    "            continue\n",
    "            \n",
    "        if key:\n",
    "            if current_key:\n",
    "                parse_results[current_key] = entries\n",
    "                \n",
    "            current_key = key\n",
    "            entry = line.split(maxsplit = 1)\n",
    "            \n",
    "            if current_key in record_parsers.keys():\n",
    "                key, val = record_parsers[current_key](entry[1])\n",
    "                entries = {\n",
    "                    key: val\n",
    "                }\n",
    "            \n",
    "            else:\n",
    "                entries = [\n",
    "                    entry[1] if len(entry) > 1 else ''\n",
    "                ]\n",
    "        \n",
    "        else:\n",
    "            if current_key in record_parsers.keys():\n",
    "                key, val = record_parsers[current_key](line.strip()) \n",
    "                entries[key] = val\n",
    "            \n",
    "            else:\n",
    "                entries.append(\n",
    "                    line.strip()\n",
    "                )\n",
    "     \n",
    "    # also save the last bits\n",
    "    parse_results[current_key] = entries  \n",
    "    \n",
    "    # remove anything that is not in keywords\n",
    "    keys = set(parse_results.keys())\n",
    "    for key in keys.difference(keywords):\n",
    "        parse_results.pop(key)\n",
    "    \n",
    "    return parse_results\n",
    "\n",
    "\n",
    "def get_kegg_pathway_maps(taxid = 'mmu'):\n",
    "    r = requests.get(\n",
    "        f'http://rest.kegg.jp/list/pathway/{taxid}'\n",
    "    )\n",
    "    response_lines = [line.split('\\t') for line in r.text.split('\\n') if line]\n",
    "    pathway_map_ids = {\n",
    "        key.split(':')[1]: val for key, val in response_lines\n",
    "    }\n",
    "    \n",
    "    pathway_maps = {}\n",
    "    for mapid in pathway_map_ids.keys():\n",
    "        r = requests.get(\n",
    "            f'http://rest.kegg.jp/get/{mapid}'\n",
    "        )\n",
    "        pathway_maps[mapid] = parse_kegg_pathway(r.text)\n",
    "        \n",
    "    return pathway_maps\n",
    "\n",
    "\n",
    "pathwaymaps = get_kegg_pathway_maps()\n",
    "with open('../raw/kegg_pathway_maps.pickle', 'wb') as handle:\n",
    "    pickle.dump(\n",
    "        pathwaymaps,\n",
    "        handle\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "dynamic-stone",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load pickled results from above cell\n",
    "import pickle\n",
    "with open('../raw/kegg_pathway_maps.pickle', 'rb') as handle:\n",
    "    pathwaymaps = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "governing-combining",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select pathways that either are a metabolism or process information\n",
    "def select_maps(pathwaydict, class_contains):\n",
    "    selection = {}\n",
    "    for k, v in pathwaydict.items():\n",
    "        if 'CLASS' in v and any([pattern in v['CLASS'][0].lower() for pattern in class_contains]):\n",
    "            selection[k] = v\n",
    "            \n",
    "    return selection\n",
    "\n",
    "selected_pathways = select_maps(\n",
    "    pathwaymaps,\n",
    "    ['metabolism', 'information processing']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "spread-methodology",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapid</th>\n",
       "      <th>name</th>\n",
       "      <th>ncommon</th>\n",
       "      <th>ncompounds</th>\n",
       "      <th>compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mmu00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis - Mus musculus (m...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>[C00022, C00186, C00103, C00111, C00074]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mmu00020</td>\n",
       "      <td>Citrate cycle (TCA cycle) - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>[C00417, C00074, C00158, C00022]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mmu00030</td>\n",
       "      <td>Pentose phosphate pathway - Mus musculus (mouse)</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>[C00022, C00257, C05382, C00620, C00199, C00345]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mmu00040</td>\n",
       "      <td>Pentose and glucuronate interconversions - Mus...</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>[C00167, C00199, C00022, C00103, C00532, C0011...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mmu00051</td>\n",
       "      <td>Fructose and mannose metabolism - Mus musculus...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>[C00111, C00186, C00096]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>mmu04310</td>\n",
       "      <td>Wnt signaling pathway - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>mmu04340</td>\n",
       "      <td>Hedgehog signaling pathway - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>mmu04370</td>\n",
       "      <td>VEGF signaling pathway - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>mmu04371</td>\n",
       "      <td>Apelin signaling pathway - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>mmu04512</td>\n",
       "      <td>ECM-receptor interaction - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mapid                                               name  ncommon  \\\n",
       "0    mmu00010  Glycolysis / Gluconeogenesis - Mus musculus (m...        0   \n",
       "1    mmu00020   Citrate cycle (TCA cycle) - Mus musculus (mouse)        0   \n",
       "2    mmu00030   Pentose phosphate pathway - Mus musculus (mouse)        2   \n",
       "3    mmu00040  Pentose and glucuronate interconversions - Mus...        2   \n",
       "4    mmu00051  Fructose and mannose metabolism - Mus musculus...        1   \n",
       "..        ...                                                ...      ...   \n",
       "102  mmu04310       Wnt signaling pathway - Mus musculus (mouse)        0   \n",
       "103  mmu04340  Hedgehog signaling pathway - Mus musculus (mouse)        0   \n",
       "104  mmu04370      VEGF signaling pathway - Mus musculus (mouse)        0   \n",
       "105  mmu04371    Apelin signaling pathway - Mus musculus (mouse)        0   \n",
       "106  mmu04512    ECM-receptor interaction - Mus musculus (mouse)        0   \n",
       "\n",
       "     ncompounds                                          compounds  \n",
       "0             5           [C00022, C00186, C00103, C00111, C00074]  \n",
       "1             4                   [C00417, C00074, C00158, C00022]  \n",
       "2             6   [C00022, C00257, C05382, C00620, C00199, C00345]  \n",
       "3             8  [C00167, C00199, C00022, C00103, C00532, C0011...  \n",
       "4             3                           [C00111, C00186, C00096]  \n",
       "..          ...                                                ...  \n",
       "102           0                                                 []  \n",
       "103           0                                                 []  \n",
       "104           0                                                 []  \n",
       "105           0                                                 []  \n",
       "106           0                                                 []  \n",
       "\n",
       "[107 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the following cells are a simple implementation of a pathway enrichment analysis\n",
    "# based on differential abundance of compounds using Fisher's exact test\n",
    "def compute_overlaps(diff_compounds, pathways, metabolites = None):\n",
    "    '''\n",
    "    computes overlaps between differentially regulated compounds and \n",
    "    each set of pathway compounds. Pathway compounds are limited to\n",
    "    the set of compounds given by metabolites (i.e. adjustment to background\n",
    "    aka only quantified metabolites are considered)\n",
    "    '''\n",
    "    diff_compounds = set(diff_compounds)\n",
    "    overlaps = []\n",
    "    for key, pathway in pathways.items():\n",
    "        if 'COMPOUND' in pathway:\n",
    "            pathway_metabolites = set(pathway['COMPOUND'].keys())\n",
    "            if metabolites:\n",
    "                pathway_metabolites = pathway_metabolites.intersection(\n",
    "                    set(metabolites)\n",
    "                )\n",
    "            overlaps.append(\n",
    "                [\n",
    "                    key, \n",
    "                    pathway['NAME'][0],\n",
    "                    len(\n",
    "                        diff_compounds.intersection(\n",
    "                            pathway_metabolites\n",
    "                        )\n",
    "                    ),\n",
    "                    len(pathway_metabolites),\n",
    "                    list(pathway_metabolites)\n",
    "                ]\n",
    "            )\n",
    "    \n",
    "    return pd.DataFrame(overlaps, columns = ['mapid', 'name', 'ncommon', 'ncompounds', 'compounds'])\n",
    "\n",
    "pvalcol = '(P40, WT) / (P2, WT).1'\n",
    "def get_diff_reg_compounds(df, pvalcol, logfoldcol, alpha = 0.05):\n",
    "    upregulated = df.loc[\n",
    "        (df[pvalcol].astype(float) <= alpha) & (df[logfoldcol].astype(float) > 0),\n",
    "        ['KEGG', 'Name', pvalcol, logfoldcol]\n",
    "    ]\n",
    "    downregulated = df.loc[\n",
    "        (df[pvalcol].astype(float) <= alpha) & (df[logfoldcol].astype(float) < 0),\n",
    "        ['KEGG', 'Name', pvalcol, logfoldcol]\n",
    "    ]\n",
    "    return upregulated, downregulated\n",
    "\n",
    "upreg, downreg = get_diff_reg_compounds(\n",
    "    hilic,\n",
    "    pvalcol,\n",
    "    pvalcol[:-2]\n",
    ")\n",
    "overlaps = compute_overlaps(upreg.KEGG.to_list(), selected_pathways, hilic.KEGG.to_list())\n",
    "overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "pretty-outline",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pvalue computation based on hypergeometric distribution\n",
    "from scipy.stats import hypergeom\n",
    "import numpy as np\n",
    "def pvalue(k, M, n, N):\n",
    "    \"\"\"\n",
    "    computes the probability to find k or more overlapping genes between two gene sets\n",
    "    of n and N genes, where n is the number of genes in the gene set to which we overlap\n",
    "    (i.e. number of genes in a pathway or number of differentially expressed genes in\n",
    "    another patient) and N is the number of genes in the gene set which we are interested\n",
    "    in (i.e. number of differentially expressed genes) which are drawn from M total genes\n",
    "    (i.e. number of genes considered during differential expression analysis). This is also\n",
    "    known as computing the pvalue for fisher's exact test.\n",
    "    See http://www.pathwaycommons.org/guide/primers/statistics/fishers_exact_test/ for\n",
    "    more information on how this is computed\n",
    "    :param k:   number of common genes between two gene sets\n",
    "    :param M:   total number of genes considered during DEA\n",
    "    :param n:   number of genes in gene set with which we overlap\n",
    "    :param N:   number of differentially expressed genes\n",
    "    :return:    probability to find k or more common genes between the two gene sets\n",
    "    \"\"\"\n",
    "    # k - 1 because we are computing P(x >= k) which includes k\n",
    "    pval = hypergeom.sf(\n",
    "        k - 1, M, n, N\n",
    "    )\n",
    "    return pval\n",
    "\n",
    "def fdrcorrection(df, fdr = 0.05):\n",
    "    # benjamini-hochberg correction\n",
    "    # taken from statsmodels.stats.multitest.fdrcorrection\n",
    "    # https://github.com/statsmodels/statsmodels/blob/main/statsmodels/stats/multitest.py\n",
    "    df.sort_values(\n",
    "        by = 'pvalue',\n",
    "        inplace = True\n",
    "    )\n",
    "    df['padj'] = np.arange(1, len(df) + 1) / len(df) * fdr\n",
    "    reject = (df.pvalue <= df.padj).values\n",
    "    if reject.any():\n",
    "        rejectmax = np.nonzero(reject)[0][-1]\n",
    "        reject[:rejectmax] = True\n",
    "    \n",
    "    df['reject'] = reject\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "conservative-magnitude",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapid</th>\n",
       "      <th>name</th>\n",
       "      <th>ncommon</th>\n",
       "      <th>ncompounds</th>\n",
       "      <th>compounds</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>reject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>mmu00250</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism - ...</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>[C00334, C00022, C00025, C03406, C00064, C0004...</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.000467</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>mmu00230</td>\n",
       "      <td>Purine metabolism - Mus musculus (mouse)</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>[C00044, C00212, C00144, C00262, C00002, C0038...</td>\n",
       "      <td>0.001302</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>mmu02010</td>\n",
       "      <td>ABC transporters - Mus musculus (mouse)</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "      <td>[C00114, C00378, C00387, C00065, C00188, C0115...</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.001402</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>mmu00410</td>\n",
       "      <td>beta-Alanine metabolism - Mus musculus (mouse)</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>[C00106, C00334, C00386, C00049]</td>\n",
       "      <td>0.002393</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>mmu00564</td>\n",
       "      <td>Glycerophospholipid metabolism - Mus musculus ...</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>[C00093, C00114, C00346, C00065, C00570, C0067...</td>\n",
       "      <td>0.002797</td>\n",
       "      <td>0.002336</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>mmu00531</td>\n",
       "      <td>Glycosaminoglycan degradation - Mus musculus (...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.048131</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>mmu00515</td>\n",
       "      <td>Mannose type O-glycan biosynthesis - Mus muscu...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>[C00063]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.048598</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>mmu00513</td>\n",
       "      <td>Various types of N-glycan biosynthesis - Mus m...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049065</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>mmu00780</td>\n",
       "      <td>Biotin metabolism - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>[C00047]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.049533</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>mmu04512</td>\n",
       "      <td>ECM-receptor interaction - Mus musculus (mouse)</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mapid                                               name  ncommon  \\\n",
       "19   mmu00250  Alanine, aspartate and glutamate metabolism - ...        5   \n",
       "16   mmu00230           Purine metabolism - Mus musculus (mouse)        6   \n",
       "82   mmu02010            ABC transporters - Mus musculus (mouse)        8   \n",
       "31   mmu00410     beta-Alanine metabolism - Mus musculus (mouse)        3   \n",
       "50   mmu00564  Glycerophospholipid metabolism - Mus musculus ...        4   \n",
       "..        ...                                                ...      ...   \n",
       "44   mmu00531  Glycosaminoglycan degradation - Mus musculus (...        0   \n",
       "41   mmu00515  Mannose type O-glycan biosynthesis - Mus muscu...        0   \n",
       "40   mmu00513  Various types of N-glycan biosynthesis - Mus m...        0   \n",
       "69   mmu00780           Biotin metabolism - Mus musculus (mouse)        0   \n",
       "106  mmu04512    ECM-receptor interaction - Mus musculus (mouse)        0   \n",
       "\n",
       "     ncompounds                                          compounds    pvalue  \\\n",
       "19           11  [C00334, C00022, C00025, C03406, C00064, C0004...  0.001275   \n",
       "16           16  [C00044, C00212, C00144, C00262, C00002, C0038...  0.001302   \n",
       "82           29  [C00114, C00378, C00387, C00065, C00188, C0115...  0.001855   \n",
       "31            4                   [C00106, C00334, C00386, C00049]  0.002393   \n",
       "50            8  [C00093, C00114, C00346, C00065, C00570, C0067...  0.002797   \n",
       "..          ...                                                ...       ...   \n",
       "44            0                                                 []  1.000000   \n",
       "41            1                                           [C00063]  1.000000   \n",
       "40            0                                                 []  1.000000   \n",
       "69            1                                           [C00047]  1.000000   \n",
       "106           0                                                 []  1.000000   \n",
       "\n",
       "         padj  reject  \n",
       "19   0.000467   False  \n",
       "16   0.000935   False  \n",
       "82   0.001402   False  \n",
       "31   0.001869   False  \n",
       "50   0.002336   False  \n",
       "..        ...     ...  \n",
       "44   0.048131   False  \n",
       "41   0.048598   False  \n",
       "40   0.049065   False  \n",
       "69   0.049533   False  \n",
       "106  0.050000   False  \n",
       "\n",
       "[107 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# M is the number of background genes and is taken from the raw excel file\n",
    "# as this is the true background for the compounds\n",
    "M = 396\n",
    "overlaps['pvalue'] = overlaps.apply(\n",
    "    lambda x: pvalue(x.ncommon, M, x.ncompounds, len(upreg)),\n",
    "    axis = 1\n",
    ")\n",
    "overlaps = fdrcorrection(overlaps)\n",
    "overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "funded-democrat",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapid</th>\n",
       "      <th>name</th>\n",
       "      <th>ncommon</th>\n",
       "      <th>ncompounds</th>\n",
       "      <th>compounds</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>reject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>mmu00250</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism - ...</td>\n",
       "      <td>5</td>\n",
       "      <td>11</td>\n",
       "      <td>[C00334, C00022, C00025, C03406, C00064, C0004...</td>\n",
       "      <td>0.001275</td>\n",
       "      <td>0.000467</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>mmu00230</td>\n",
       "      <td>Purine metabolism - Mus musculus (mouse)</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>[C00044, C00212, C00144, C00262, C00002, C0038...</td>\n",
       "      <td>0.001302</td>\n",
       "      <td>0.000935</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>mmu02010</td>\n",
       "      <td>ABC transporters - Mus musculus (mouse)</td>\n",
       "      <td>8</td>\n",
       "      <td>29</td>\n",
       "      <td>[C00114, C00378, C00387, C00065, C00188, C0115...</td>\n",
       "      <td>0.001855</td>\n",
       "      <td>0.001402</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>mmu00410</td>\n",
       "      <td>beta-Alanine metabolism - Mus musculus (mouse)</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>[C00106, C00334, C00386, C00049]</td>\n",
       "      <td>0.002393</td>\n",
       "      <td>0.001869</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>mmu00564</td>\n",
       "      <td>Glycerophospholipid metabolism - Mus musculus ...</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>[C00093, C00114, C00346, C00065, C00570, C0067...</td>\n",
       "      <td>0.002797</td>\n",
       "      <td>0.002336</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>mmu04080</td>\n",
       "      <td>Neuroactive ligand-receptor interaction - Mus ...</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>[C00212, C00334, C00002, C00025, C12270, C0001...</td>\n",
       "      <td>0.004718</td>\n",
       "      <td>0.002804</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>mmu00910</td>\n",
       "      <td>Nitrogen metabolism - Mus musculus (mouse)</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>[C00025, C00064]</td>\n",
       "      <td>0.007608</td>\n",
       "      <td>0.003271</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>mmu00340</td>\n",
       "      <td>Histidine metabolism - Mus musculus (mouse)</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>[C00386, C00025, C02835, C05828, C00049, C01152]</td>\n",
       "      <td>0.010554</td>\n",
       "      <td>0.003738</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>mmu00480</td>\n",
       "      <td>Glutathione metabolism - Mus musculus (mouse)</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>[C00051, C01879, C00025, C00072, C00005, C0012...</td>\n",
       "      <td>0.017349</td>\n",
       "      <td>0.004206</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>mmu00220</td>\n",
       "      <td>Arginine biosynthesis - Mus musculus (mouse)</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>[C00327, C00025, C03406, C00062, C00064, C0004...</td>\n",
       "      <td>0.017349</td>\n",
       "      <td>0.004673</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>mmu01040</td>\n",
       "      <td>Biosynthesis of unsaturated fatty acids - Mus ...</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>[C06429, C16513, C01595, C16525, C16527, C0324...</td>\n",
       "      <td>0.026081</td>\n",
       "      <td>0.005140</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>mmu00770</td>\n",
       "      <td>Pantothenate and CoA biosynthesis - Mus muscul...</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>[C00049, C00106, C00010, C00022]</td>\n",
       "      <td>0.040704</td>\n",
       "      <td>0.005607</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>mmu00740</td>\n",
       "      <td>Riboflavin metabolism - Mus musculus (mouse)</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>[C00044, C00016, C00255, C00199]</td>\n",
       "      <td>0.040704</td>\n",
       "      <td>0.006075</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>mmu00650</td>\n",
       "      <td>Butanoate metabolism - Mus musculus (mouse)</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>[C00025, C00334, C01089, C00022]</td>\n",
       "      <td>0.040704</td>\n",
       "      <td>0.006542</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       mapid                                               name  ncommon  \\\n",
       "19  mmu00250  Alanine, aspartate and glutamate metabolism - ...        5   \n",
       "16  mmu00230           Purine metabolism - Mus musculus (mouse)        6   \n",
       "82  mmu02010            ABC transporters - Mus musculus (mouse)        8   \n",
       "31  mmu00410     beta-Alanine metabolism - Mus musculus (mouse)        3   \n",
       "50  mmu00564  Glycerophospholipid metabolism - Mus musculus ...        4   \n",
       "96  mmu04080  Neuroactive ligand-receptor interaction - Mus ...        4   \n",
       "75  mmu00910         Nitrogen metabolism - Mus musculus (mouse)        2   \n",
       "26  mmu00340        Histidine metabolism - Mus musculus (mouse)        3   \n",
       "36  mmu00480      Glutathione metabolism - Mus musculus (mouse)        3   \n",
       "15  mmu00220       Arginine biosynthesis - Mus musculus (mouse)        3   \n",
       "81  mmu01040  Biosynthesis of unsaturated fatty acids - Mus ...        3   \n",
       "68  mmu00770  Pantothenate and CoA biosynthesis - Mus muscul...        2   \n",
       "65  mmu00740       Riboflavin metabolism - Mus musculus (mouse)        2   \n",
       "62  mmu00650        Butanoate metabolism - Mus musculus (mouse)        2   \n",
       "\n",
       "    ncompounds                                          compounds    pvalue  \\\n",
       "19          11  [C00334, C00022, C00025, C03406, C00064, C0004...  0.001275   \n",
       "16          16  [C00044, C00212, C00144, C00262, C00002, C0038...  0.001302   \n",
       "82          29  [C00114, C00378, C00387, C00065, C00188, C0115...  0.001855   \n",
       "31           4                   [C00106, C00334, C00386, C00049]  0.002393   \n",
       "50           8  [C00093, C00114, C00346, C00065, C00570, C0067...  0.002797   \n",
       "96           9  [C00212, C00334, C00002, C00025, C12270, C0001...  0.004718   \n",
       "75           2                                   [C00025, C00064]  0.007608   \n",
       "26           6   [C00386, C00025, C02835, C05828, C00049, C01152]  0.010554   \n",
       "36           7  [C00051, C01879, C00025, C00072, C00005, C0012...  0.017349   \n",
       "15           7  [C00327, C00025, C03406, C00062, C00064, C0004...  0.017349   \n",
       "81           8  [C06429, C16513, C01595, C16525, C16527, C0324...  0.026081   \n",
       "68           4                   [C00049, C00106, C00010, C00022]  0.040704   \n",
       "65           4                   [C00044, C00016, C00255, C00199]  0.040704   \n",
       "62           4                   [C00025, C00334, C01089, C00022]  0.040704   \n",
       "\n",
       "        padj  reject  \n",
       "19  0.000467   False  \n",
       "16  0.000935   False  \n",
       "82  0.001402   False  \n",
       "31  0.001869   False  \n",
       "50  0.002336   False  \n",
       "96  0.002804   False  \n",
       "75  0.003271   False  \n",
       "26  0.003738   False  \n",
       "36  0.004206   False  \n",
       "15  0.004673   False  \n",
       "81  0.005140   False  \n",
       "68  0.005607   False  \n",
       "65  0.006075   False  \n",
       "62  0.006542   False  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps[overlaps.pvalue <= 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "appointed-patrick",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'(P2, WT) / (E14_5, WT)':         mapid                                               name foldchange\n",
       " 0    mmu00230           Purine metabolism - Mus musculus (mouse)       0.86\n",
       " 1    mmu01040  Biosynthesis of unsaturated fatty acids - Mus ...      1.275\n",
       " 2    mmu04022  cGMP-PKG signaling pathway - Mus musculus (mouse)       1.99\n",
       " 3    mmu00030   Pentose phosphate pathway - Mus musculus (mouse)       1.64\n",
       " 4    mmu00240       Pyrimidine metabolism - Mus musculus (mouse)      -0.08\n",
       " ..        ...                                                ...        ...\n",
       " 120    C00072                                      Ascorbic acid       0.55\n",
       " 121    C00127                             L-Glutathione oxidized      -0.05\n",
       " 122    C00005                                              NADPH      -3.00\n",
       " 123    C00519                                        Hypotaurine      -1.73\n",
       " 124    C05122                                   Taurocholic acid       3.80\n",
       " \n",
       " [125 rows x 3 columns],\n",
       " '(P40, WT) / (P2, WT)':         mapid                                               name foldchange\n",
       " 0    mmu00250  Alanine, aspartate and glutamate metabolism - ...       0.69\n",
       " 1    mmu00230           Purine metabolism - Mus musculus (mouse)      0.715\n",
       " 2    mmu02010            ABC transporters - Mus musculus (mouse)       0.25\n",
       " 3    mmu00410     beta-Alanine metabolism - Mus musculus (mouse)       0.88\n",
       " 4    mmu00564  Glycerophospholipid metabolism - Mus musculus ...      0.655\n",
       " ..        ...                                                ...        ...\n",
       " 110    C00363                  Thymidine 5'-diphosphate (5'-TDP)      -1.25\n",
       " 111    C00112                   Cytidine 5'-diphosphate (5'-CDP)      -0.93\n",
       " 112    C00063                  Cytidine 5'-triphosphate (5'-CTP)      -2.70\n",
       " 113    C00029                        Uridine 5'-diphosphoglucose      -0.26\n",
       " 114    C02067                                      Pseudouridine      -2.05\n",
       " \n",
       " [115 rows x 3 columns],\n",
       " '(P2, KO) / (E14_5, KO)':         mapid                                               name foldchange\n",
       " 0    mmu01040  Biosynthesis of unsaturated fatty acids - Mus ...      1.555\n",
       " 1    mmu00230           Purine metabolism - Mus musculus (mouse)      0.915\n",
       " 2    mmu00030   Pentose phosphate pathway - Mus musculus (mouse)      1.915\n",
       " 3    mmu00240       Pyrimidine metabolism - Mus musculus (mouse)      0.295\n",
       " 4    mmu00562  Inositol phosphate metabolism - Mus musculus (...       2.58\n",
       " ..        ...                                                ...        ...\n",
       " 103    C02989                            DL-Methionine sulfoxide      -1.02\n",
       " 104    C02291                                L-(+)-Cystathionine       3.53\n",
       " 105    C00021                          S-Adenosyl-L-homocysteine       0.06\n",
       " 106    C00019                 S-(5'-Adenosyl)-L-methionine (SAM)      -0.08\n",
       " 107    C00170                       5'-S-Methyl-5'-thioadenosine      -0.69\n",
       " \n",
       " [108 rows x 3 columns],\n",
       " '(P40, KO) / (P2, KO)':         mapid                                               name foldchange\n",
       " 0    mmu02010            ABC transporters - Mus musculus (mouse)       0.29\n",
       " 1    mmu00230           Purine metabolism - Mus musculus (mouse)       0.46\n",
       " 2    mmu00410     beta-Alanine metabolism - Mus musculus (mouse)      0.815\n",
       " 3    mmu04080  Neuroactive ligand-receptor interaction - Mus ...      -0.37\n",
       " 4    mmu00760  Nicotinate and nicotinamide metabolism - Mus m...        0.6\n",
       " ..        ...                                                ...        ...\n",
       " 116    C00073                                   L-(-)-Methionine      -0.41\n",
       " 117    C02989                            DL-Methionine sulfoxide      -0.18\n",
       " 118    C00021                          S-Adenosyl-L-homocysteine       0.59\n",
       " 119    C00606                            L-Cysteinesulfinic acid      -0.44\n",
       " 120    C00170                       5'-S-Methyl-5'-thioadenosine      -1.01\n",
       " \n",
       " [121 rows x 3 columns]}"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# computes the pathway enrichments and average log fold changes for enriched pathways\n",
    "# based on differential abundance of quantified compounds\n",
    "def compute_selections(df, pathways, keggidcol, foldchange_columns, pvalue_columns, alpha = 0.05, fdr = 0.1):\n",
    "    selections = {}\n",
    "    for pvalcol, fccol in zip(pvalue_columns, foldchange_columns):\n",
    "        selection = pd.DataFrame(columns = ['mapid', 'name', 'foldchange'])\n",
    "        for diff_reg_compounds in get_diff_reg_compounds(df, pvalcol, fccol):\n",
    "            enrichment = compute_overlaps(\n",
    "                diff_reg_compounds[keggidcol].to_list(),\n",
    "                pathways,\n",
    "                df[keggidcol].to_list()\n",
    "            )\n",
    "            enrichment['pvalue'] = enrichment.apply(\n",
    "                lambda x: pvalue(\n",
    "                    x.ncommon, M, x.ncompounds, len(diff_reg_compounds)\n",
    "                ),\n",
    "                axis = 1\n",
    "            )\n",
    "            enrichment = fdrcorrection(enrichment, fdr = fdr)\n",
    "            enrichment = enrichment.loc[enrichment.pvalue <= alpha, :].reset_index(drop = True)\n",
    "            enrichment['foldchange'] = .0\n",
    "            compound_dfs = []\n",
    "            for i, row in enrichment.iterrows():\n",
    "                compounds = enrichment.at[i, 'compounds']\n",
    "                enrichment.loc[i, 'foldchange'] = df.loc[df[keggidcol].isin(compounds), fccol].median()\n",
    "                compound_dfs.append(\n",
    "                    df.loc[df[keggidcol].isin(compounds), [keggidcol, 'Name', fccol]].rename(\n",
    "                        columns = {\n",
    "                            keggidcol: 'mapid',\n",
    "                            'Name': 'name',\n",
    "                            fccol: 'foldchange'\n",
    "                        }\n",
    "                    )\n",
    "                ) \n",
    "            \n",
    "            \n",
    "            selection = pd.concat(\n",
    "                [\n",
    "                    selection, \n",
    "                    enrichment.loc[:, ['mapid', 'name', 'foldchange']],\n",
    "                    *compound_dfs\n",
    "                ]\n",
    "            )\n",
    "        \n",
    "        selections[fccol] = selection \\\n",
    "            .drop_duplicates() \\\n",
    "            .reset_index(drop = True)\n",
    "    \n",
    "    return selections\n",
    "\n",
    "foldchangecols = [\n",
    "    '(P2, WT) / (E14_5, WT)', \n",
    "    '(P40, WT) / (P2, WT)', \n",
    "    '(P2, KO) / (E14_5, KO)', \n",
    "    '(P40, KO) / (P2, KO)'\n",
    "]\n",
    "selections = compute_selections(\n",
    "    hilic,\n",
    "    selected_pathways,\n",
    "    'KEGG',\n",
    "    foldchangecols,\n",
    "    [col + '.1' for col in foldchangecols]\n",
    ")\n",
    "selections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "existing-valentine",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformats enrichment results from above to conform to iPATH specifications\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "coolwarm = plt.get_cmap('coolwarm')\n",
    "def get_ipath_selection(df, fccol, idcol, widths, cmap = None, color = None, replace_idprefix = None):\n",
    "    df = df.copy()\n",
    "    df.drop_duplicates(\n",
    "        inplace = True\n",
    "    )\n",
    "    df.loc[:, fccol] = df[fccol].astype(float)\n",
    "    \n",
    "    if cmap:\n",
    "        norm = mpl.colors.Normalize(\n",
    "            vmin = df[fccol].quantile(0.05),\n",
    "            vmax = df[fccol].quantile(0.95),\n",
    "            clip = True\n",
    "        )\n",
    "        df['color'] = df[fccol].apply(\n",
    "            lambda x: mpl.colors.to_hex(cmap(norm(x)))\n",
    "        )\n",
    "    \n",
    "    elif color:\n",
    "        df['color'] = color\n",
    "\n",
    "    df['idprefix'] = df[idcol].str.replace(\n",
    "        '\\d+|[.-]', '',\n",
    "        regex = True\n",
    "    )\n",
    "    df['width'] = df['idprefix'].apply(\n",
    "        lambda x: f'W{widths[x]}'\n",
    "    )\n",
    "    if replace_idprefix:\n",
    "        for oldprefix, newprefix in replace_idprefix.items():\n",
    "            df.loc[:, idcol] = df[idcol].str.replace(\n",
    "                oldprefix,\n",
    "                newprefix\n",
    "            )\n",
    "            \n",
    "    return df.loc[:, ['mapid', 'color', 'width']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "robust-london",
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing selection files\n",
    "prefix = '../raw/ipath_selection_hilic'\n",
    "for key, selection in selections.items():\n",
    "    suffix = key[1:-1] \\\n",
    "        .replace(', ', '_') \\\n",
    "        .replace(') / (', '_')\n",
    "    \n",
    "    ipath_selection = get_ipath_selection(\n",
    "        selection,\n",
    "        'foldchange',\n",
    "        'mapid',\n",
    "        {'C': 15, 'mmu': 10},\n",
    "        cmap = coolwarm,\n",
    "        replace_idprefix = {'mmu': 'map'}\n",
    "    )\n",
    "    ipath_selection.to_csv(\n",
    "        '_'.join([prefix, suffix]) + '.tsv',\n",
    "        sep = '\\t',\n",
    "        header = False,\n",
    "        index = False\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "comparative-familiar",
   "metadata": {},
   "outputs": [],
   "source": [
    "# writing base selection files\n",
    "wtbase = pd.concat(\n",
    "    [selection for key, selection in selections.items() if 'WT' in key]\n",
    ").drop_duplicates().reset_index(drop = True)\n",
    "kdbase = pd.concat(\n",
    "    [selection for key, selection in selections.items() if 'KO' in key]\n",
    ").drop_duplicates().reset_index(drop = True)\n",
    "\n",
    "prefix = '../raw/ipath_selection_hilic_base'\n",
    "for suffix, selection in zip(['WT', 'KO'], [wtbase, kdbase]):\n",
    "    ipath_selection = get_ipath_selection(\n",
    "        selection,\n",
    "        'foldchange',\n",
    "        'mapid',\n",
    "        {'C': 15, 'mmu': 10},\n",
    "        color = '#dddddd',\n",
    "        replace_idprefix = {'mmu': 'map'}\n",
    "    )\n",
    "    ipath_selection.to_csv(\n",
    "        '_'.join([prefix, suffix]) + '.tsv',\n",
    "        sep = '\\t',\n",
    "        header = False,\n",
    "        index = False\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "based-voltage",
   "metadata": {},
   "source": [
    "# Integration with scRNAseq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "infinite-consumption",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>seq_folder</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>log10GenesPerUMI</th>\n",
       "      <th>mitoRatio</th>\n",
       "      <th>cells</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_mouseID</th>\n",
       "      <th>sample_litter</th>\n",
       "      <th>sample_treatment</th>\n",
       "      <th>sample_sex</th>\n",
       "      <th>RNA_snn_res.0.9</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162325</th>\n",
       "      <td>wt</td>\n",
       "      <td>136321</td>\n",
       "      <td>3141</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>136624</td>\n",
       "      <td>3172</td>\n",
       "      <td>0.681787</td>\n",
       "      <td>0.037987</td>\n",
       "      <td>wt_162325</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162326</th>\n",
       "      <td>wt</td>\n",
       "      <td>109397</td>\n",
       "      <td>3513</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>109533</td>\n",
       "      <td>3537</td>\n",
       "      <td>0.704158</td>\n",
       "      <td>0.073549</td>\n",
       "      <td>wt_162326</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162328</th>\n",
       "      <td>wt</td>\n",
       "      <td>132297</td>\n",
       "      <td>4008</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>132761</td>\n",
       "      <td>4084</td>\n",
       "      <td>0.704867</td>\n",
       "      <td>0.078540</td>\n",
       "      <td>wt_162328</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162329</th>\n",
       "      <td>wt</td>\n",
       "      <td>114295</td>\n",
       "      <td>4265</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>114493</td>\n",
       "      <td>4321</td>\n",
       "      <td>0.718668</td>\n",
       "      <td>0.108740</td>\n",
       "      <td>wt_162329</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162338</th>\n",
       "      <td>wt</td>\n",
       "      <td>79354</td>\n",
       "      <td>3931</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>79467</td>\n",
       "      <td>3981</td>\n",
       "      <td>0.734664</td>\n",
       "      <td>0.058326</td>\n",
       "      <td>wt_162338</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162341</th>\n",
       "      <td>wt</td>\n",
       "      <td>116909</td>\n",
       "      <td>3415</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>117073</td>\n",
       "      <td>3447</td>\n",
       "      <td>0.697933</td>\n",
       "      <td>0.060953</td>\n",
       "      <td>wt_162341</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162343</th>\n",
       "      <td>wt</td>\n",
       "      <td>129070</td>\n",
       "      <td>3882</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>129241</td>\n",
       "      <td>3915</td>\n",
       "      <td>0.702886</td>\n",
       "      <td>0.064608</td>\n",
       "      <td>wt_162343</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162423</th>\n",
       "      <td>wt</td>\n",
       "      <td>277949</td>\n",
       "      <td>4922</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>279631</td>\n",
       "      <td>5079</td>\n",
       "      <td>0.680386</td>\n",
       "      <td>0.065522</td>\n",
       "      <td>wt_162423</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162438</th>\n",
       "      <td>wt</td>\n",
       "      <td>551099</td>\n",
       "      <td>5216</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>551782</td>\n",
       "      <td>5290</td>\n",
       "      <td>0.648486</td>\n",
       "      <td>0.030061</td>\n",
       "      <td>wt_162438</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162455</th>\n",
       "      <td>wt</td>\n",
       "      <td>513383</td>\n",
       "      <td>5764</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>513903</td>\n",
       "      <td>5853</td>\n",
       "      <td>0.659684</td>\n",
       "      <td>0.035662</td>\n",
       "      <td>wt_162455</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162469</th>\n",
       "      <td>wt</td>\n",
       "      <td>732113</td>\n",
       "      <td>6561</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>732929</td>\n",
       "      <td>6685</td>\n",
       "      <td>0.652184</td>\n",
       "      <td>0.038466</td>\n",
       "      <td>wt_162469</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162500</th>\n",
       "      <td>wt</td>\n",
       "      <td>316925</td>\n",
       "      <td>6631</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>319202</td>\n",
       "      <td>6945</td>\n",
       "      <td>0.697970</td>\n",
       "      <td>0.103502</td>\n",
       "      <td>wt_162500</td>\n",
       "      <td>wt</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162526</th>\n",
       "      <td>wt</td>\n",
       "      <td>307751</td>\n",
       "      <td>4451</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>308066</td>\n",
       "      <td>4501</td>\n",
       "      <td>0.665612</td>\n",
       "      <td>0.043319</td>\n",
       "      <td>wt_162526</td>\n",
       "      <td>wt</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162561</th>\n",
       "      <td>wt</td>\n",
       "      <td>560051</td>\n",
       "      <td>5862</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>560353</td>\n",
       "      <td>5925</td>\n",
       "      <td>0.656295</td>\n",
       "      <td>0.025834</td>\n",
       "      <td>wt_162561</td>\n",
       "      <td>wt</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162563</th>\n",
       "      <td>wt</td>\n",
       "      <td>441536</td>\n",
       "      <td>5890</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>442133</td>\n",
       "      <td>5972</td>\n",
       "      <td>0.668866</td>\n",
       "      <td>0.048723</td>\n",
       "      <td>wt_162563</td>\n",
       "      <td>wt</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162330</th>\n",
       "      <td>ko</td>\n",
       "      <td>106049</td>\n",
       "      <td>4215</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>106261</td>\n",
       "      <td>4253</td>\n",
       "      <td>0.721931</td>\n",
       "      <td>0.057415</td>\n",
       "      <td>ko_162330</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162332</th>\n",
       "      <td>ko</td>\n",
       "      <td>227793</td>\n",
       "      <td>5296</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>228111</td>\n",
       "      <td>5363</td>\n",
       "      <td>0.696026</td>\n",
       "      <td>0.078146</td>\n",
       "      <td>ko_162332</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162337</th>\n",
       "      <td>ko</td>\n",
       "      <td>145405</td>\n",
       "      <td>5157</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>145614</td>\n",
       "      <td>5218</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.075364</td>\n",
       "      <td>ko_162337</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162348</th>\n",
       "      <td>ko</td>\n",
       "      <td>185919</td>\n",
       "      <td>4490</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>186012</td>\n",
       "      <td>4520</td>\n",
       "      <td>0.693635</td>\n",
       "      <td>0.062147</td>\n",
       "      <td>ko_162348</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162350</th>\n",
       "      <td>ko</td>\n",
       "      <td>206738</td>\n",
       "      <td>5469</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>207038</td>\n",
       "      <td>5536</td>\n",
       "      <td>0.704131</td>\n",
       "      <td>0.056685</td>\n",
       "      <td>ko_162350</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162352</th>\n",
       "      <td>ko</td>\n",
       "      <td>114618</td>\n",
       "      <td>4218</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>114759</td>\n",
       "      <td>4257</td>\n",
       "      <td>0.717244</td>\n",
       "      <td>0.115189</td>\n",
       "      <td>ko_162352</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/464</td>\n",
       "      <td>G1/59 A</td>\n",
       "      <td>prep 1</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162392</th>\n",
       "      <td>ko</td>\n",
       "      <td>286917</td>\n",
       "      <td>6140</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>289793</td>\n",
       "      <td>6493</td>\n",
       "      <td>0.697983</td>\n",
       "      <td>0.073801</td>\n",
       "      <td>ko_162392</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162394</th>\n",
       "      <td>ko</td>\n",
       "      <td>565183</td>\n",
       "      <td>6422</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>566466</td>\n",
       "      <td>6576</td>\n",
       "      <td>0.663627</td>\n",
       "      <td>0.037257</td>\n",
       "      <td>ko_162394</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162395</th>\n",
       "      <td>ko</td>\n",
       "      <td>371514</td>\n",
       "      <td>5330</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>372944</td>\n",
       "      <td>5452</td>\n",
       "      <td>0.670638</td>\n",
       "      <td>0.038971</td>\n",
       "      <td>ko_162395</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162396</th>\n",
       "      <td>ko</td>\n",
       "      <td>233805</td>\n",
       "      <td>4474</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>235431</td>\n",
       "      <td>4618</td>\n",
       "      <td>0.682157</td>\n",
       "      <td>0.065709</td>\n",
       "      <td>ko_162396</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162397</th>\n",
       "      <td>ko</td>\n",
       "      <td>209352</td>\n",
       "      <td>4786</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>211819</td>\n",
       "      <td>4996</td>\n",
       "      <td>0.694451</td>\n",
       "      <td>0.062629</td>\n",
       "      <td>ko_162397</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162398</th>\n",
       "      <td>ko</td>\n",
       "      <td>398571</td>\n",
       "      <td>5436</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>400306</td>\n",
       "      <td>5601</td>\n",
       "      <td>0.669047</td>\n",
       "      <td>0.053042</td>\n",
       "      <td>ko_162398</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162399</th>\n",
       "      <td>ko</td>\n",
       "      <td>379639</td>\n",
       "      <td>5561</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>382834</td>\n",
       "      <td>5838</td>\n",
       "      <td>0.674594</td>\n",
       "      <td>0.065107</td>\n",
       "      <td>ko_162399</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162414</th>\n",
       "      <td>ko</td>\n",
       "      <td>336479</td>\n",
       "      <td>5288</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>337729</td>\n",
       "      <td>5422</td>\n",
       "      <td>0.675430</td>\n",
       "      <td>0.031330</td>\n",
       "      <td>ko_162414</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162426</th>\n",
       "      <td>ko</td>\n",
       "      <td>571643</td>\n",
       "      <td>5647</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>573952</td>\n",
       "      <td>5850</td>\n",
       "      <td>0.654148</td>\n",
       "      <td>0.045211</td>\n",
       "      <td>ko_162426</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162429</th>\n",
       "      <td>ko</td>\n",
       "      <td>361997</td>\n",
       "      <td>5021</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>363634</td>\n",
       "      <td>5159</td>\n",
       "      <td>0.667648</td>\n",
       "      <td>0.099993</td>\n",
       "      <td>ko_162429</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/469</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162456</th>\n",
       "      <td>ko</td>\n",
       "      <td>433931</td>\n",
       "      <td>5490</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>434645</td>\n",
       "      <td>5590</td>\n",
       "      <td>0.664655</td>\n",
       "      <td>0.044142</td>\n",
       "      <td>ko_162456</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162462</th>\n",
       "      <td>ko</td>\n",
       "      <td>589058</td>\n",
       "      <td>6295</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>590642</td>\n",
       "      <td>6457</td>\n",
       "      <td>0.660166</td>\n",
       "      <td>0.056674</td>\n",
       "      <td>ko_162462</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162463</th>\n",
       "      <td>ko</td>\n",
       "      <td>383863</td>\n",
       "      <td>4845</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>384243</td>\n",
       "      <td>4918</td>\n",
       "      <td>0.661065</td>\n",
       "      <td>0.034551</td>\n",
       "      <td>ko_162463</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162474</th>\n",
       "      <td>ko</td>\n",
       "      <td>711025</td>\n",
       "      <td>5671</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>711965</td>\n",
       "      <td>5780</td>\n",
       "      <td>0.642794</td>\n",
       "      <td>0.066801</td>\n",
       "      <td>ko_162474</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162476</th>\n",
       "      <td>ko</td>\n",
       "      <td>352966</td>\n",
       "      <td>4826</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>355477</td>\n",
       "      <td>5025</td>\n",
       "      <td>0.666774</td>\n",
       "      <td>0.146381</td>\n",
       "      <td>ko_162476</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162490</th>\n",
       "      <td>ko</td>\n",
       "      <td>371498</td>\n",
       "      <td>4661</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>371923</td>\n",
       "      <td>4725</td>\n",
       "      <td>0.659624</td>\n",
       "      <td>0.038484</td>\n",
       "      <td>ko_162490</td>\n",
       "      <td>ko</td>\n",
       "      <td>GNF1/473</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162503</th>\n",
       "      <td>ko</td>\n",
       "      <td>354930</td>\n",
       "      <td>4245</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>355147</td>\n",
       "      <td>4281</td>\n",
       "      <td>0.654284</td>\n",
       "      <td>0.038466</td>\n",
       "      <td>ko_162503</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>m</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162519</th>\n",
       "      <td>ko</td>\n",
       "      <td>412552</td>\n",
       "      <td>4358</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>412888</td>\n",
       "      <td>4407</td>\n",
       "      <td>0.648905</td>\n",
       "      <td>0.038451</td>\n",
       "      <td>ko_162519</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162520</th>\n",
       "      <td>ko</td>\n",
       "      <td>633954</td>\n",
       "      <td>4889</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>634300</td>\n",
       "      <td>4946</td>\n",
       "      <td>0.636688</td>\n",
       "      <td>0.031236</td>\n",
       "      <td>ko_162520</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162521</th>\n",
       "      <td>ko</td>\n",
       "      <td>491593</td>\n",
       "      <td>4639</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>492384</td>\n",
       "      <td>4719</td>\n",
       "      <td>0.645407</td>\n",
       "      <td>0.039370</td>\n",
       "      <td>ko_162521</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162523</th>\n",
       "      <td>ko</td>\n",
       "      <td>58177</td>\n",
       "      <td>4583</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>62993</td>\n",
       "      <td>5193</td>\n",
       "      <td>0.774160</td>\n",
       "      <td>0.058800</td>\n",
       "      <td>ko_162523</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162534</th>\n",
       "      <td>ko</td>\n",
       "      <td>359250</td>\n",
       "      <td>4387</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>359573</td>\n",
       "      <td>4429</td>\n",
       "      <td>0.656308</td>\n",
       "      <td>0.033084</td>\n",
       "      <td>ko_162534</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162551</th>\n",
       "      <td>ko</td>\n",
       "      <td>288246</td>\n",
       "      <td>3967</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>288691</td>\n",
       "      <td>4005</td>\n",
       "      <td>0.659765</td>\n",
       "      <td>0.031612</td>\n",
       "      <td>ko_162551</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162553</th>\n",
       "      <td>ko</td>\n",
       "      <td>331737</td>\n",
       "      <td>4999</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>332022</td>\n",
       "      <td>5055</td>\n",
       "      <td>0.670822</td>\n",
       "      <td>0.071040</td>\n",
       "      <td>ko_162553</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162554</th>\n",
       "      <td>ko</td>\n",
       "      <td>435996</td>\n",
       "      <td>5670</td>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>436357</td>\n",
       "      <td>5742</td>\n",
       "      <td>0.666519</td>\n",
       "      <td>0.042850</td>\n",
       "      <td>ko_162554</td>\n",
       "      <td>ko</td>\n",
       "      <td>GN21/475</td>\n",
       "      <td>G1/59 B</td>\n",
       "      <td>prep 2</td>\n",
       "      <td>f</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>neurons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       orig.ident  nCount_RNA  nFeature_RNA     seq_folder    nUMI  nGene  \\\n",
       "162325         wt      136321          3141  SeuratProject  136624   3172   \n",
       "162326         wt      109397          3513  SeuratProject  109533   3537   \n",
       "162328         wt      132297          4008  SeuratProject  132761   4084   \n",
       "162329         wt      114295          4265  SeuratProject  114493   4321   \n",
       "162338         wt       79354          3931  SeuratProject   79467   3981   \n",
       "162341         wt      116909          3415  SeuratProject  117073   3447   \n",
       "162343         wt      129070          3882  SeuratProject  129241   3915   \n",
       "162423         wt      277949          4922  SeuratProject  279631   5079   \n",
       "162438         wt      551099          5216  SeuratProject  551782   5290   \n",
       "162455         wt      513383          5764  SeuratProject  513903   5853   \n",
       "162469         wt      732113          6561  SeuratProject  732929   6685   \n",
       "162500         wt      316925          6631  SeuratProject  319202   6945   \n",
       "162526         wt      307751          4451  SeuratProject  308066   4501   \n",
       "162561         wt      560051          5862  SeuratProject  560353   5925   \n",
       "162563         wt      441536          5890  SeuratProject  442133   5972   \n",
       "162330         ko      106049          4215  SeuratProject  106261   4253   \n",
       "162332         ko      227793          5296  SeuratProject  228111   5363   \n",
       "162337         ko      145405          5157  SeuratProject  145614   5218   \n",
       "162348         ko      185919          4490  SeuratProject  186012   4520   \n",
       "162350         ko      206738          5469  SeuratProject  207038   5536   \n",
       "162352         ko      114618          4218  SeuratProject  114759   4257   \n",
       "162392         ko      286917          6140  SeuratProject  289793   6493   \n",
       "162394         ko      565183          6422  SeuratProject  566466   6576   \n",
       "162395         ko      371514          5330  SeuratProject  372944   5452   \n",
       "162396         ko      233805          4474  SeuratProject  235431   4618   \n",
       "162397         ko      209352          4786  SeuratProject  211819   4996   \n",
       "162398         ko      398571          5436  SeuratProject  400306   5601   \n",
       "162399         ko      379639          5561  SeuratProject  382834   5838   \n",
       "162414         ko      336479          5288  SeuratProject  337729   5422   \n",
       "162426         ko      571643          5647  SeuratProject  573952   5850   \n",
       "162429         ko      361997          5021  SeuratProject  363634   5159   \n",
       "162456         ko      433931          5490  SeuratProject  434645   5590   \n",
       "162462         ko      589058          6295  SeuratProject  590642   6457   \n",
       "162463         ko      383863          4845  SeuratProject  384243   4918   \n",
       "162474         ko      711025          5671  SeuratProject  711965   5780   \n",
       "162476         ko      352966          4826  SeuratProject  355477   5025   \n",
       "162490         ko      371498          4661  SeuratProject  371923   4725   \n",
       "162503         ko      354930          4245  SeuratProject  355147   4281   \n",
       "162519         ko      412552          4358  SeuratProject  412888   4407   \n",
       "162520         ko      633954          4889  SeuratProject  634300   4946   \n",
       "162521         ko      491593          4639  SeuratProject  492384   4719   \n",
       "162523         ko       58177          4583  SeuratProject   62993   5193   \n",
       "162534         ko      359250          4387  SeuratProject  359573   4429   \n",
       "162551         ko      288246          3967  SeuratProject  288691   4005   \n",
       "162553         ko      331737          4999  SeuratProject  332022   5055   \n",
       "162554         ko      435996          5670  SeuratProject  436357   5742   \n",
       "\n",
       "        log10GenesPerUMI  mitoRatio      cells sample sample_mouseID  \\\n",
       "162325          0.681787   0.037987  wt_162325     wt       GNF1/464   \n",
       "162326          0.704158   0.073549  wt_162326     wt       GNF1/464   \n",
       "162328          0.704867   0.078540  wt_162328     wt       GNF1/464   \n",
       "162329          0.718668   0.108740  wt_162329     wt       GNF1/464   \n",
       "162338          0.734664   0.058326  wt_162338     wt       GNF1/464   \n",
       "162341          0.697933   0.060953  wt_162341     wt       GNF1/464   \n",
       "162343          0.702886   0.064608  wt_162343     wt       GNF1/464   \n",
       "162423          0.680386   0.065522  wt_162423     wt       GNF1/469   \n",
       "162438          0.648486   0.030061  wt_162438     wt       GNF1/473   \n",
       "162455          0.659684   0.035662  wt_162455     wt       GNF1/473   \n",
       "162469          0.652184   0.038466  wt_162469     wt       GNF1/473   \n",
       "162500          0.697970   0.103502  wt_162500     wt       GNF1/473   \n",
       "162526          0.665612   0.043319  wt_162526     wt       GN21/475   \n",
       "162561          0.656295   0.025834  wt_162561     wt       GN21/475   \n",
       "162563          0.668866   0.048723  wt_162563     wt       GN21/475   \n",
       "162330          0.721931   0.057415  ko_162330     ko       GNF1/464   \n",
       "162332          0.696026   0.078146  ko_162332     ko       GNF1/464   \n",
       "162337          0.720000   0.075364  ko_162337     ko       GNF1/464   \n",
       "162348          0.693635   0.062147  ko_162348     ko       GNF1/464   \n",
       "162350          0.704131   0.056685  ko_162350     ko       GNF1/464   \n",
       "162352          0.717244   0.115189  ko_162352     ko       GNF1/464   \n",
       "162392          0.697983   0.073801  ko_162392     ko       GNF1/469   \n",
       "162394          0.663627   0.037257  ko_162394     ko       GNF1/469   \n",
       "162395          0.670638   0.038971  ko_162395     ko       GNF1/469   \n",
       "162396          0.682157   0.065709  ko_162396     ko       GNF1/469   \n",
       "162397          0.694451   0.062629  ko_162397     ko       GNF1/469   \n",
       "162398          0.669047   0.053042  ko_162398     ko       GNF1/469   \n",
       "162399          0.674594   0.065107  ko_162399     ko       GNF1/469   \n",
       "162414          0.675430   0.031330  ko_162414     ko       GNF1/469   \n",
       "162426          0.654148   0.045211  ko_162426     ko       GNF1/469   \n",
       "162429          0.667648   0.099993  ko_162429     ko       GNF1/469   \n",
       "162456          0.664655   0.044142  ko_162456     ko       GNF1/473   \n",
       "162462          0.660166   0.056674  ko_162462     ko       GNF1/473   \n",
       "162463          0.661065   0.034551  ko_162463     ko       GNF1/473   \n",
       "162474          0.642794   0.066801  ko_162474     ko       GNF1/473   \n",
       "162476          0.666774   0.146381  ko_162476     ko       GNF1/473   \n",
       "162490          0.659624   0.038484  ko_162490     ko       GNF1/473   \n",
       "162503          0.654284   0.038466  ko_162503     ko       GN21/475   \n",
       "162519          0.648905   0.038451  ko_162519     ko       GN21/475   \n",
       "162520          0.636688   0.031236  ko_162520     ko       GN21/475   \n",
       "162521          0.645407   0.039370  ko_162521     ko       GN21/475   \n",
       "162523          0.774160   0.058800  ko_162523     ko       GN21/475   \n",
       "162534          0.656308   0.033084  ko_162534     ko       GN21/475   \n",
       "162551          0.659765   0.031612  ko_162551     ko       GN21/475   \n",
       "162553          0.670822   0.071040  ko_162553     ko       GN21/475   \n",
       "162554          0.666519   0.042850  ko_162554     ko       GN21/475   \n",
       "\n",
       "       sample_litter sample_treatment sample_sex  RNA_snn_res.0.9  \\\n",
       "162325       G1/59 A           prep 1          f                3   \n",
       "162326       G1/59 A           prep 1          f                3   \n",
       "162328       G1/59 A           prep 1          f                3   \n",
       "162329       G1/59 A           prep 1          f                3   \n",
       "162338       G1/59 A           prep 1          f                3   \n",
       "162341       G1/59 A           prep 1          f                3   \n",
       "162343       G1/59 A           prep 1          f                3   \n",
       "162423       G1/59 B           prep 2          m                3   \n",
       "162438       G1/59 B           prep 2          m                3   \n",
       "162455       G1/59 B           prep 2          m                3   \n",
       "162469       G1/59 B           prep 2          m                3   \n",
       "162500       G1/59 B           prep 2          m                3   \n",
       "162526       G1/59 B           prep 2          f                3   \n",
       "162561       G1/59 B           prep 2          f                3   \n",
       "162563       G1/59 B           prep 2          f                3   \n",
       "162330       G1/59 A           prep 1          f                3   \n",
       "162332       G1/59 A           prep 1          f                3   \n",
       "162337       G1/59 A           prep 1          f                3   \n",
       "162348       G1/59 A           prep 1          f                3   \n",
       "162350       G1/59 A           prep 1          f                3   \n",
       "162352       G1/59 A           prep 1          f                3   \n",
       "162392       G1/59 B           prep 2          m                3   \n",
       "162394       G1/59 B           prep 2          m                3   \n",
       "162395       G1/59 B           prep 2          m                3   \n",
       "162396       G1/59 B           prep 2          m                3   \n",
       "162397       G1/59 B           prep 2          m                3   \n",
       "162398       G1/59 B           prep 2          m                3   \n",
       "162399       G1/59 B           prep 2          m                3   \n",
       "162414       G1/59 B           prep 2          m                3   \n",
       "162426       G1/59 B           prep 2          m                3   \n",
       "162429       G1/59 B           prep 2          m                3   \n",
       "162456       G1/59 B           prep 2          m                3   \n",
       "162462       G1/59 B           prep 2          m                3   \n",
       "162463       G1/59 B           prep 2          m                3   \n",
       "162474       G1/59 B           prep 2          m                3   \n",
       "162476       G1/59 B           prep 2          m                3   \n",
       "162490       G1/59 B           prep 2          m                3   \n",
       "162503       G1/59 B           prep 2          m                3   \n",
       "162519       G1/59 B           prep 2          f                3   \n",
       "162520       G1/59 B           prep 2          f                3   \n",
       "162521       G1/59 B           prep 2          f                3   \n",
       "162523       G1/59 B           prep 2          f                3   \n",
       "162534       G1/59 B           prep 2          f                3   \n",
       "162551       G1/59 B           prep 2          f                3   \n",
       "162553       G1/59 B           prep 2          f                3   \n",
       "162554       G1/59 B           prep 2          f                3   \n",
       "\n",
       "        seurat_clusters cell_type  \n",
       "162325                3   neurons  \n",
       "162326                3   neurons  \n",
       "162328                3   neurons  \n",
       "162329                3   neurons  \n",
       "162338                3   neurons  \n",
       "162341                3   neurons  \n",
       "162343                3   neurons  \n",
       "162423                3   neurons  \n",
       "162438                3   neurons  \n",
       "162455                3   neurons  \n",
       "162469                3   neurons  \n",
       "162500                3   neurons  \n",
       "162526                3   neurons  \n",
       "162561                3   neurons  \n",
       "162563                3   neurons  \n",
       "162330                3   neurons  \n",
       "162332                3   neurons  \n",
       "162337                3   neurons  \n",
       "162348                3   neurons  \n",
       "162350                3   neurons  \n",
       "162352                3   neurons  \n",
       "162392                3   neurons  \n",
       "162394                3   neurons  \n",
       "162395                3   neurons  \n",
       "162396                3   neurons  \n",
       "162397                3   neurons  \n",
       "162398                3   neurons  \n",
       "162399                3   neurons  \n",
       "162414                3   neurons  \n",
       "162426                3   neurons  \n",
       "162429                3   neurons  \n",
       "162456                3   neurons  \n",
       "162462                3   neurons  \n",
       "162463                3   neurons  \n",
       "162474                3   neurons  \n",
       "162476                3   neurons  \n",
       "162490                3   neurons  \n",
       "162503                3   neurons  \n",
       "162519                3   neurons  \n",
       "162520                3   neurons  \n",
       "162521                3   neurons  \n",
       "162523                3   neurons  \n",
       "162534                3   neurons  \n",
       "162551                3   neurons  \n",
       "162553                3   neurons  \n",
       "162554                3   neurons  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "from scipy.io import mmread\n",
    "\n",
    "def data2h5ad(prefix):\n",
    "    X = mmread(prefix + '.mtx').tocsr()\n",
    "    obs = pd.read_csv(\n",
    "        prefix + '.metadata.tsv',\n",
    "        sep = '\\t'\n",
    "    )\n",
    "    with open(prefix + f'.rownames.txt', 'r') as f:\n",
    "        genes = [r.rstrip() for r in f]\n",
    "    \n",
    "    var = pd.DataFrame(\n",
    "        index = genes\n",
    "    )\n",
    "    \n",
    "    return ad.AnnData(\n",
    "        X = X.T,\n",
    "        obs = obs,\n",
    "        var = var\n",
    "    )\n",
    "columns = [\n",
    "    'sample_treatment', \n",
    "    'sample_sex', \n",
    "    'sample_litter', \n",
    "    'sample_mouseID', \n",
    "    'cell_type',\n",
    "    'quantification'\n",
    "]\n",
    "annotation = {\n",
    "    0: 'olfactory_cells',\n",
    "    1: 'oligodendrytes',\n",
    "    2: 'astrocytes',\n",
    "    3: 'neurons'\n",
    "}\n",
    "adata = data2h5ad('../processed/novarino_scRNA')\n",
    "adata.obs.index = [x.split('_')[1] for x in adata.obs.index]\n",
    "adata.obs['cell_type'] = adata.obs.seurat_clusters.apply(lambda x: annotation[x])\n",
    "adata = adata[adata.obs.cell_type == 'neurons', :].copy()\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "broke-conflict",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 46 × 26249\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'seq_folder', 'nUMI', 'nGene', 'log10GenesPerUMI', 'mitoRatio', 'cells', 'sample', 'sample_mouseID', 'sample_litter', 'sample_treatment', 'sample_sex', 'RNA_snn_res.0.9', 'seurat_clusters', 'cell_type'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers['counts'] = adata.X.astype(int).copy()\n",
    "sc.pp.normalize_total(adata)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "running-simple",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_38312/2006032247.py:6: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  'lfc': np.log2(ko/wt)\n",
      "/tmp/ipykernel_38312/2006032247.py:6: RuntimeWarning: invalid value encountered in true_divide\n",
      "  'lfc': np.log2(ko/wt)\n",
      "/tmp/ipykernel_38312/2006032247.py:6: RuntimeWarning: divide by zero encountered in log2\n",
      "  'lfc': np.log2(ko/wt)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Gm38148</td>\n",
       "      <td>-0.619308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Gm6085</td>\n",
       "      <td>-0.695657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Gm6123</td>\n",
       "      <td>0.947214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>-0.626928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Lypla1</td>\n",
       "      <td>0.115159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26244</th>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>-0.103994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26245</th>\n",
       "      <td>mt-Te</td>\n",
       "      <td>-0.155918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26246</th>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>0.194500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26247</th>\n",
       "      <td>mt-Tt</td>\n",
       "      <td>-1.862060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26248</th>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>0.077618</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14221 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genesymbol       lfc\n",
       "5        Gm38148 -0.619308\n",
       "14        Gm6085 -0.695657\n",
       "16        Gm6123  0.947214\n",
       "17        Mrpl15 -0.626928\n",
       "18        Lypla1  0.115159\n",
       "...          ...       ...\n",
       "26244     mt-Nd6 -0.103994\n",
       "26245      mt-Te -0.155918\n",
       "26246    mt-Cytb  0.194500\n",
       "26247      mt-Tt -1.862060\n",
       "26248      mt-Tp  0.077618\n",
       "\n",
       "[14221 rows x 2 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt = np.array(adata[adata.obs['sample'] == 'wt', :].X.mean(axis = 0))[0]\n",
    "ko = np.array(adata[adata.obs['sample'] == 'ko', :].X.mean(axis = 0))[0]\n",
    "lfc = pd.DataFrame(\n",
    "    {\n",
    "        'genesymbol': adata.var.index,\n",
    "        'lfc': np.log2(ko/wt)\n",
    "    }\n",
    ")\n",
    "lfc = lfc[~lfc.lfc.isna() & ~np.isinf(lfc.lfc)]\n",
    "lfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "sensitive-cooper",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hk2</td>\n",
       "      <td>2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Hk3</td>\n",
       "      <td>2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hk1</td>\n",
       "      <td>2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hkdc1</td>\n",
       "      <td>2.7.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gck</td>\n",
       "      <td>2.7.1.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5497</th>\n",
       "      <td>Casp8</td>\n",
       "      <td>3.4.22.61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5498</th>\n",
       "      <td>Casp7</td>\n",
       "      <td>3.4.22.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5501</th>\n",
       "      <td>Mmp3</td>\n",
       "      <td>3.4.24.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5502</th>\n",
       "      <td>Mmp9</td>\n",
       "      <td>3.4.24.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5503</th>\n",
       "      <td>Mmp14</td>\n",
       "      <td>3.4.24.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2292 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol  ec_number\n",
       "0           Hk2    2.7.1.1\n",
       "1           Hk3    2.7.1.1\n",
       "2           Hk1    2.7.1.1\n",
       "3         Hkdc1    2.7.1.1\n",
       "4           Gck    2.7.1.2\n",
       "...         ...        ...\n",
       "5497      Casp8  3.4.22.61\n",
       "5498      Casp7  3.4.22.60\n",
       "5501       Mmp3  3.4.24.17\n",
       "5502       Mmp9  3.4.24.35\n",
       "5503      Mmp14  3.4.24.80\n",
       "\n",
       "[2292 rows x 2 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_gene_ec_numbers(pathways):\n",
    "    sym_to_ec = []\n",
    "    for name, pathway in pathways.items():\n",
    "        if 'GENE' in pathway:\n",
    "            for ncbid, gene in pathway['GENE'].items():\n",
    "                if 'EC' in gene and 'genesymbol' in gene:\n",
    "                    for ec in gene['EC'].split():\n",
    "                        sym_to_ec.append(\n",
    "                            [gene['genesymbol'], ec]\n",
    "                        )\n",
    "    \n",
    "    return pd.DataFrame(sym_to_ec, columns = ['genesymbol', 'ec_number'])\n",
    "\n",
    "sym_to_ec = get_gene_ec_numbers(selected_pathways)\n",
    "sym_to_ec.drop_duplicates(\n",
    "    subset = ['genesymbol'],\n",
    "    inplace = True\n",
    ")\n",
    "sym_to_ec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "noble-novel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>lfc</th>\n",
       "      <th>ec_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lypla1</td>\n",
       "      <td>0.115159</td>\n",
       "      <td>3.1.1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sgk3</td>\n",
       "      <td>1.379994</td>\n",
       "      <td>2.7.11.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rdh10</td>\n",
       "      <td>0.542524</td>\n",
       "      <td>1.1.1.-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ube2w</td>\n",
       "      <td>-1.229854</td>\n",
       "      <td>2.3.2.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B3gat2</td>\n",
       "      <td>-0.659934</td>\n",
       "      <td>2.4.1.135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1465</th>\n",
       "      <td>Ctps2</td>\n",
       "      <td>0.396960</td>\n",
       "      <td>6.3.4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466</th>\n",
       "      <td>Siah1b</td>\n",
       "      <td>-0.553851</td>\n",
       "      <td>2.3.2.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1467</th>\n",
       "      <td>Prps2</td>\n",
       "      <td>-2.201418</td>\n",
       "      <td>2.7.6.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468</th>\n",
       "      <td>Hccs</td>\n",
       "      <td>-0.290481</td>\n",
       "      <td>4.4.1.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>Mid1</td>\n",
       "      <td>-0.303377</td>\n",
       "      <td>2.3.2.27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1470 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol       lfc  ec_number\n",
       "0        Lypla1  0.115159    3.1.1.5\n",
       "1          Sgk3  1.379994   2.7.11.1\n",
       "2         Rdh10  0.542524    1.1.1.-\n",
       "3         Ube2w -1.229854   2.3.2.25\n",
       "4        B3gat2 -0.659934  2.4.1.135\n",
       "...         ...       ...        ...\n",
       "1465      Ctps2  0.396960    6.3.4.2\n",
       "1466     Siah1b -0.553851   2.3.2.27\n",
       "1467      Prps2 -2.201418    2.7.6.1\n",
       "1468       Hccs -0.290481   4.4.1.17\n",
       "1469       Mid1 -0.303377   2.3.2.27\n",
       "\n",
       "[1470 rows x 3 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lfc = lfc.merge(\n",
    "    sym_to_ec,\n",
    "    how = 'inner',\n",
    "    on = 'genesymbol'\n",
    ")\n",
    "lfc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "registered-valley",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'(P2, WT) / (E14_5, WT)':         mapid                                               name foldchange\n",
       " 0    mmu00230           Purine metabolism - Mus musculus (mouse)       0.86\n",
       " 1    mmu01040  Biosynthesis of unsaturated fatty acids - Mus ...      1.275\n",
       " 2    mmu04022  cGMP-PKG signaling pathway - Mus musculus (mouse)       1.99\n",
       " 3    mmu00030   Pentose phosphate pathway - Mus musculus (mouse)       1.64\n",
       " 4    mmu00240       Pyrimidine metabolism - Mus musculus (mouse)      -0.08\n",
       " ..        ...                                                ...        ...\n",
       " 120    C00072                                      Ascorbic acid       0.55\n",
       " 121    C00127                             L-Glutathione oxidized      -0.05\n",
       " 122    C00005                                              NADPH      -3.00\n",
       " 123    C00519                                        Hypotaurine      -1.73\n",
       " 124    C05122                                   Taurocholic acid       3.80\n",
       " \n",
       " [125 rows x 3 columns],\n",
       " '(P40, WT) / (P2, WT)':         mapid                                               name foldchange\n",
       " 0    mmu00250  Alanine, aspartate and glutamate metabolism - ...       0.69\n",
       " 1    mmu00230           Purine metabolism - Mus musculus (mouse)      0.715\n",
       " 2    mmu02010            ABC transporters - Mus musculus (mouse)       0.25\n",
       " 3    mmu00410     beta-Alanine metabolism - Mus musculus (mouse)       0.88\n",
       " 4    mmu00564  Glycerophospholipid metabolism - Mus musculus ...      0.655\n",
       " ..        ...                                                ...        ...\n",
       " 110    C00363                  Thymidine 5'-diphosphate (5'-TDP)      -1.25\n",
       " 111    C00112                   Cytidine 5'-diphosphate (5'-CDP)      -0.93\n",
       " 112    C00063                  Cytidine 5'-triphosphate (5'-CTP)      -2.70\n",
       " 113    C00029                        Uridine 5'-diphosphoglucose      -0.26\n",
       " 114    C02067                                      Pseudouridine      -2.05\n",
       " \n",
       " [115 rows x 3 columns],\n",
       " '(P2, KO) / (E14_5, KO)':         mapid                                               name foldchange\n",
       " 0    mmu01040  Biosynthesis of unsaturated fatty acids - Mus ...      1.555\n",
       " 1    mmu00230           Purine metabolism - Mus musculus (mouse)      0.915\n",
       " 2    mmu00030   Pentose phosphate pathway - Mus musculus (mouse)      1.915\n",
       " 3    mmu00240       Pyrimidine metabolism - Mus musculus (mouse)      0.295\n",
       " 4    mmu00562  Inositol phosphate metabolism - Mus musculus (...       2.58\n",
       " ..        ...                                                ...        ...\n",
       " 103    C02989                            DL-Methionine sulfoxide      -1.02\n",
       " 104    C02291                                L-(+)-Cystathionine       3.53\n",
       " 105    C00021                          S-Adenosyl-L-homocysteine       0.06\n",
       " 106    C00019                 S-(5'-Adenosyl)-L-methionine (SAM)      -0.08\n",
       " 107    C00170                       5'-S-Methyl-5'-thioadenosine      -0.69\n",
       " \n",
       " [108 rows x 3 columns],\n",
       " '(P40, KO) / (P2, KO)':         mapid                                               name foldchange\n",
       " 0    mmu02010            ABC transporters - Mus musculus (mouse)       0.29\n",
       " 1    mmu00230           Purine metabolism - Mus musculus (mouse)       0.46\n",
       " 2    mmu00410     beta-Alanine metabolism - Mus musculus (mouse)      0.815\n",
       " 3    mmu04080  Neuroactive ligand-receptor interaction - Mus ...      -0.37\n",
       " 4    mmu00760  Nicotinate and nicotinamide metabolism - Mus m...        0.6\n",
       " ..        ...                                                ...        ...\n",
       " 116    C00073                                   L-(-)-Methionine      -0.41\n",
       " 117    C02989                            DL-Methionine sulfoxide      -0.18\n",
       " 118    C00021                          S-Adenosyl-L-homocysteine       0.59\n",
       " 119    C00606                            L-Cysteinesulfinic acid      -0.44\n",
       " 120    C00170                       5'-S-Methyl-5'-thioadenosine      -1.01\n",
       " \n",
       " [121 rows x 3 columns]}"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "patient-watson",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "def make_union(a, b):\n",
    "    return a.union(b)\n",
    "\n",
    "selection_union = reduce(\n",
    "    make_union,\n",
    "    [set(d.mapid) for d in selections.values()]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "extended-final",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "523"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ec_numbers = set()\n",
    "for pathwayid in [s for s in selection_union if s.startswith('mmu')]:\n",
    "    pathway = selected_pathways[pathwayid]\n",
    "    if 'GENE' in pathway:\n",
    "        for ncbid, gene in pathway['GENE'].items():\n",
    "            if 'EC' in gene:\n",
    "                for ec in gene['EC'].split():\n",
    "                    ec_numbers.add(ec)\n",
    "len(ec_numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "choice-ontario",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapid</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EC3.1.1.5</td>\n",
       "      <td>0.115159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EC2.7.11.1</td>\n",
       "      <td>1.379994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EC1.1.1.-</td>\n",
       "      <td>0.542524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EC3.1.3.66</td>\n",
       "      <td>-0.045725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EC2.7.11.1</td>\n",
       "      <td>0.588497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1465</th>\n",
       "      <td>EC6.3.4.2</td>\n",
       "      <td>0.396960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1466</th>\n",
       "      <td>EC2.3.2.27</td>\n",
       "      <td>-0.553851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1467</th>\n",
       "      <td>EC2.7.6.1</td>\n",
       "      <td>-2.201418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1468</th>\n",
       "      <td>EC4.4.1.17</td>\n",
       "      <td>-0.290481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1469</th>\n",
       "      <td>EC2.3.2.27</td>\n",
       "      <td>-0.303377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>777 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           mapid       lfc\n",
       "0      EC3.1.1.5  0.115159\n",
       "1     EC2.7.11.1  1.379994\n",
       "2      EC1.1.1.-  0.542524\n",
       "7     EC3.1.3.66 -0.045725\n",
       "11    EC2.7.11.1  0.588497\n",
       "...          ...       ...\n",
       "1465   EC6.3.4.2  0.396960\n",
       "1466  EC2.3.2.27 -0.553851\n",
       "1467   EC2.7.6.1 -2.201418\n",
       "1468  EC4.4.1.17 -0.290481\n",
       "1469  EC2.3.2.27 -0.303377\n",
       "\n",
       "[777 rows x 2 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enzymes = lfc.loc[lfc.ec_number.isin(ec_numbers), ['ec_number', 'lfc']]\n",
    "enzymes.rename(\n",
    "    columns = {'ec_number': 'mapid'},\n",
    "    inplace = True\n",
    ")\n",
    "# prepend prefix required by iPATH\n",
    "enzymes.loc[:, 'mapid'] = enzymes.mapid.apply(\n",
    "    lambda x: 'EC' + x\n",
    ")\n",
    "enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "painted-gasoline",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapid</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C00022</td>\n",
       "      <td>-0.31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C00041</td>\n",
       "      <td>-0.05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00213</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C00186</td>\n",
       "      <td>0.21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>C00116</td>\n",
       "      <td>-0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>C16527</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>C01083</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>C00015</td>\n",
       "      <td>-0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>C00008</td>\n",
       "      <td>-0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>C00167</td>\n",
       "      <td>-0.19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      mapid    lfc\n",
       "0    C00022  -0.31\n",
       "1    C00041  -0.05\n",
       "3    C00213   0.04\n",
       "5    C00186   0.21\n",
       "6    C00116  -0.06\n",
       "..      ...    ...\n",
       "232  C16527   0.24\n",
       "234  C01083   0.60\n",
       "237  C00015  -0.04\n",
       "238  C00008  -0.19\n",
       "240  C00167  -0.19\n",
       "\n",
       "[119 rows x 2 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compounds = hilic.loc[\n",
    "    hilic.KEGG.isin([s for s in selection_union if s.startswith('C')]),\n",
    "    ['KEGG', '(P2, KO) / (P2, WT)']\n",
    "]\n",
    "compounds.rename(\n",
    "    columns = {\n",
    "        'KEGG': 'mapid',\n",
    "        '(P2, KO) / (P2, WT)': 'lfc'\n",
    "    },\n",
    "    inplace = True\n",
    ")\n",
    "compounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "continental-suite",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapid</th>\n",
       "      <th>lfc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EC3.1.1.5</td>\n",
       "      <td>0.115159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EC2.7.11.1</td>\n",
       "      <td>1.379994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EC1.1.1.-</td>\n",
       "      <td>0.542524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EC3.1.3.66</td>\n",
       "      <td>-0.045725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EC2.7.11.1</td>\n",
       "      <td>0.588497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>C16527</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>C01083</td>\n",
       "      <td>0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>C00015</td>\n",
       "      <td>-0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>C00008</td>\n",
       "      <td>-0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>C00167</td>\n",
       "      <td>-0.19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>896 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          mapid       lfc\n",
       "0     EC3.1.1.5  0.115159\n",
       "1    EC2.7.11.1  1.379994\n",
       "2     EC1.1.1.-  0.542524\n",
       "7    EC3.1.3.66 -0.045725\n",
       "11   EC2.7.11.1  0.588497\n",
       "..          ...       ...\n",
       "232      C16527      0.24\n",
       "234      C01083      0.60\n",
       "237      C00015     -0.04\n",
       "238      C00008     -0.19\n",
       "240      C00167     -0.19\n",
       "\n",
       "[896 rows x 2 columns]"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "integrated_selection = pd.concat(\n",
    "    [enzymes, compounds]\n",
    ")\n",
    "integrated_selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "peaceful-respect",
   "metadata": {},
   "outputs": [],
   "source": [
    "prefix = '../raw/ipath_selection_hilic'\n",
    "for key, selection in selections.items():\n",
    "    suffix = 'P2_KO_WT_integrated'\n",
    "    ipath_selection = get_ipath_selection(\n",
    "        integrated_selection,\n",
    "        'lfc',\n",
    "        'mapid',\n",
    "        {'C': 15, 'EC': 10},\n",
    "        cmap = coolwarm\n",
    "    )\n",
    "    ipath_selection.to_csv(\n",
    "        '_'.join([prefix, suffix]) + '.tsv',\n",
    "        sep = '\\t',\n",
    "        header = False,\n",
    "        index = False\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-nova]",
   "language": "python",
   "name": "conda-env-.conda-nova-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
