{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "hawaiian-cargo",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "from scipy.sparse import coo_matrix, hstack, vstack, issparse\n",
    "\n",
    "def read_brain_atlas_data(filename):\n",
    "    h5 = h5py.File(filename, 'r')\n",
    "    print('reading chunks')\n",
    "    countdata = h5['data']['counts']\n",
    "    nrows, ncols = countdata.shape\n",
    "    sparsechunks, sparsecolchunks = [], []\n",
    "    for rowslice, colslice in countdata.iter_chunks():\n",
    "        sparsechunk = coo_matrix(\n",
    "            countdata[rowslice, colslice].T.astype(np.int32),\n",
    "        )\n",
    "        sparsechunk.eliminate_zeros()\n",
    "        sparsechunks.append(\n",
    "            sparsechunk\n",
    "        )\n",
    "        \n",
    "        if colslice.stop == ncols:\n",
    "            sparsecolchunks.append(\n",
    "                vstack(sparsechunks)\n",
    "            )\n",
    "            \n",
    "            del sparsechunks\n",
    "            sparsechunks = []\n",
    "    \n",
    "    print('stack column chunks')\n",
    "    X = hstack(sparsecolchunks)\n",
    "    del sparsecolchunks\n",
    "    \n",
    "    print('convert to csr matrix')\n",
    "    X = X.tocsr()\n",
    "    \n",
    "    row_col = dict()\n",
    "    for h5key, dictkey in zip(['gene', 'samples'], ['var', 'obs']):\n",
    "        print(f'reading {h5key}')\n",
    "        row_col[dictkey] = pd.DataFrame(\n",
    "            index = np.char.decode(\n",
    "                h5['data'][h5key][()],\n",
    "                encoding = 'utf-8'\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    print('converting to AnnData')\n",
    "    return ad.AnnData(X, **row_col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "needed-alberta",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this takes approximately 200GB of RAM due to the count matrix being saved as dense array\n",
    "# adata = read_brain_atlas_data('../data/mouse_brain_adult_counts.h5')\n",
    "# adata.obs.index = adata.obs.index.str.replace('10X_cells.', '', regex = True)\n",
    "# metadata = pd.read_csv(\n",
    "#     '../data/mouse_brain_adult_anno.csv',\n",
    "#     index_col = 0\n",
    "# )\n",
    "# metadata.index = metadata.sample_name.str.replace('10X_cells.', '', regex = True).to_list()\n",
    "# metadata.drop(\n",
    "#     columns = [\n",
    "#         'sample_name',\n",
    "#         'exp_component_name'\n",
    "#     ]\n",
    "# )\n",
    "# adata.obs = adata.obs.merge(\n",
    "#     metadata,\n",
    "#     left_index = True,\n",
    "#     right_index = True,\n",
    "#     how = 'left'\n",
    "# )\n",
    "# adata.write_h5ad('../data/mouse_brain_adult.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "yellow-right",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only using primary somatosensory cortex (SSp)\n",
    "adata = sc.read_h5ad(\n",
    "    '../data/mouse_brain_adult.h5ad'\n",
    ")\n",
    "adult = adata[adata.obs.region_label == 'SSp', :].copy()\n",
    "del adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "second-effort",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>n_hkgene</th>\n",
       "      <th>S_Score</th>\n",
       "      <th>G2M_Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>CC_Difference</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>...</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>species</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease__ontology_label</th>\n",
       "      <th>organ</th>\n",
       "      <th>organ__ontology_label</th>\n",
       "      <th>library_preparation_protocol</th>\n",
       "      <th>library_preparation_protocol__ontology_label</th>\n",
       "      <th>sex</th>\n",
       "      <th>species__ontology_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E10_v1_AAACCTGAGGGTCTCC-1</th>\n",
       "      <td>E10</td>\n",
       "      <td>1544</td>\n",
       "      <td>1022</td>\n",
       "      <td>0.020078</td>\n",
       "      <td>51</td>\n",
       "      <td>0.356987</td>\n",
       "      <td>0.330795</td>\n",
       "      <td>S</td>\n",
       "      <td>0.026192</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_E10</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E10_v1_AAACCTGCACAACGCC-1</th>\n",
       "      <td>E10</td>\n",
       "      <td>1157</td>\n",
       "      <td>783</td>\n",
       "      <td>0.014693</td>\n",
       "      <td>39</td>\n",
       "      <td>0.453854</td>\n",
       "      <td>0.260560</td>\n",
       "      <td>S</td>\n",
       "      <td>0.193294</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_E10</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E10_v1_AAACCTGCACGACGAA-1</th>\n",
       "      <td>E10</td>\n",
       "      <td>2081</td>\n",
       "      <td>1200</td>\n",
       "      <td>0.016338</td>\n",
       "      <td>67</td>\n",
       "      <td>0.447598</td>\n",
       "      <td>0.218746</td>\n",
       "      <td>S</td>\n",
       "      <td>0.228852</td>\n",
       "      <td>34</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_E10</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E10_v1_AAACCTGCAGACAAGC-1</th>\n",
       "      <td>E10</td>\n",
       "      <td>2490</td>\n",
       "      <td>1430</td>\n",
       "      <td>0.021285</td>\n",
       "      <td>71</td>\n",
       "      <td>0.227723</td>\n",
       "      <td>0.307102</td>\n",
       "      <td>G2M</td>\n",
       "      <td>-0.079379</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_E10</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E10_v1_AAACCTGCAGAGCCAA-1</th>\n",
       "      <td>E10</td>\n",
       "      <td>2514</td>\n",
       "      <td>1416</td>\n",
       "      <td>0.021082</td>\n",
       "      <td>70</td>\n",
       "      <td>0.738406</td>\n",
       "      <td>0.136556</td>\n",
       "      <td>S</td>\n",
       "      <td>0.601850</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_E10</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_TTTGTCAGTTGATTGC-1</th>\n",
       "      <td>P4</td>\n",
       "      <td>8028</td>\n",
       "      <td>3130</td>\n",
       "      <td>0.064275</td>\n",
       "      <td>74</td>\n",
       "      <td>-0.225020</td>\n",
       "      <td>-0.244369</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.019349</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_P4</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_TTTGTCATCAACGGCC-1</th>\n",
       "      <td>P4</td>\n",
       "      <td>15768</td>\n",
       "      <td>4425</td>\n",
       "      <td>0.041540</td>\n",
       "      <td>79</td>\n",
       "      <td>-0.219746</td>\n",
       "      <td>-0.236435</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.016689</td>\n",
       "      <td>24</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_P4</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_TTTGTCATCCGTTGTC-1</th>\n",
       "      <td>P4</td>\n",
       "      <td>21692</td>\n",
       "      <td>4931</td>\n",
       "      <td>0.041398</td>\n",
       "      <td>76</td>\n",
       "      <td>-0.207196</td>\n",
       "      <td>-0.244800</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.037604</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_P4</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_TTTGTCATCGATGAGG-1</th>\n",
       "      <td>P4</td>\n",
       "      <td>3680</td>\n",
       "      <td>1897</td>\n",
       "      <td>0.057880</td>\n",
       "      <td>67</td>\n",
       "      <td>-0.185044</td>\n",
       "      <td>-0.185335</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.000291</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_P4</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4_TTTGTCATCTTTCCTC-1</th>\n",
       "      <td>P4</td>\n",
       "      <td>6641</td>\n",
       "      <td>2667</td>\n",
       "      <td>0.024394</td>\n",
       "      <td>75</td>\n",
       "      <td>-0.141394</td>\n",
       "      <td>-0.225909</td>\n",
       "      <td>G1</td>\n",
       "      <td>0.084515</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>mouse_P4</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0008930</td>\n",
       "      <td>somatosensory cortex</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2 sequencing</td>\n",
       "      <td>mixed</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>98047 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          orig_ident  nCount_RNA  nFeature_RNA  percent_mito  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1        E10        1544          1022      0.020078   \n",
       "E10_v1_AAACCTGCACAACGCC-1        E10        1157           783      0.014693   \n",
       "E10_v1_AAACCTGCACGACGAA-1        E10        2081          1200      0.016338   \n",
       "E10_v1_AAACCTGCAGACAAGC-1        E10        2490          1430      0.021285   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1        E10        2514          1416      0.021082   \n",
       "...                              ...         ...           ...           ...   \n",
       "P4_TTTGTCAGTTGATTGC-1             P4        8028          3130      0.064275   \n",
       "P4_TTTGTCATCAACGGCC-1             P4       15768          4425      0.041540   \n",
       "P4_TTTGTCATCCGTTGTC-1             P4       21692          4931      0.041398   \n",
       "P4_TTTGTCATCGATGAGG-1             P4        3680          1897      0.057880   \n",
       "P4_TTTGTCATCTTTCCTC-1             P4        6641          2667      0.024394   \n",
       "\n",
       "                           n_hkgene   S_Score  G2M_Score Phase  CC_Difference  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1        51  0.356987   0.330795     S       0.026192   \n",
       "E10_v1_AAACCTGCACAACGCC-1        39  0.453854   0.260560     S       0.193294   \n",
       "E10_v1_AAACCTGCACGACGAA-1        67  0.447598   0.218746     S       0.228852   \n",
       "E10_v1_AAACCTGCAGACAAGC-1        71  0.227723   0.307102   G2M      -0.079379   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1        70  0.738406   0.136556     S       0.601850   \n",
       "...                             ...       ...        ...   ...            ...   \n",
       "P4_TTTGTCAGTTGATTGC-1            74 -0.225020  -0.244369    G1       0.019349   \n",
       "P4_TTTGTCATCAACGGCC-1            79 -0.219746  -0.236435    G1       0.016689   \n",
       "P4_TTTGTCATCCGTTGTC-1            76 -0.207196  -0.244800    G1       0.037604   \n",
       "P4_TTTGTCATCGATGAGG-1            67 -0.185044  -0.185335    G1       0.000291   \n",
       "P4_TTTGTCATCTTTCCTC-1            75 -0.141394  -0.225909    G1       0.084515   \n",
       "\n",
       "                           seurat_clusters  ...   donor_id          species  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1               34  ...  mouse_E10  NCBITaxon_10090   \n",
       "E10_v1_AAACCTGCACAACGCC-1               34  ...  mouse_E10  NCBITaxon_10090   \n",
       "E10_v1_AAACCTGCACGACGAA-1               34  ...  mouse_E10  NCBITaxon_10090   \n",
       "E10_v1_AAACCTGCAGACAAGC-1               37  ...  mouse_E10  NCBITaxon_10090   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1               37  ...  mouse_E10  NCBITaxon_10090   \n",
       "...                                    ...  ...        ...              ...   \n",
       "P4_TTTGTCAGTTGATTGC-1                   16  ...   mouse_P4  NCBITaxon_10090   \n",
       "P4_TTTGTCATCAACGGCC-1                   24  ...   mouse_P4  NCBITaxon_10090   \n",
       "P4_TTTGTCATCCGTTGTC-1                   16  ...   mouse_P4  NCBITaxon_10090   \n",
       "P4_TTTGTCATCGATGAGG-1                   16  ...   mouse_P4  NCBITaxon_10090   \n",
       "P4_TTTGTCATCTTTCCTC-1                   12  ...   mouse_P4  NCBITaxon_10090   \n",
       "\n",
       "                                disease disease__ontology_label  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1  PATO_0000461                  normal   \n",
       "E10_v1_AAACCTGCACAACGCC-1  PATO_0000461                  normal   \n",
       "E10_v1_AAACCTGCACGACGAA-1  PATO_0000461                  normal   \n",
       "E10_v1_AAACCTGCAGACAAGC-1  PATO_0000461                  normal   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1  PATO_0000461                  normal   \n",
       "...                                 ...                     ...   \n",
       "P4_TTTGTCAGTTGATTGC-1      PATO_0000461                  normal   \n",
       "P4_TTTGTCATCAACGGCC-1      PATO_0000461                  normal   \n",
       "P4_TTTGTCATCCGTTGTC-1      PATO_0000461                  normal   \n",
       "P4_TTTGTCATCGATGAGG-1      PATO_0000461                  normal   \n",
       "P4_TTTGTCATCTTTCCTC-1      PATO_0000461                  normal   \n",
       "\n",
       "                                    organ organ__ontology_label  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1  UBERON_0008930  somatosensory cortex   \n",
       "E10_v1_AAACCTGCACAACGCC-1  UBERON_0008930  somatosensory cortex   \n",
       "E10_v1_AAACCTGCACGACGAA-1  UBERON_0008930  somatosensory cortex   \n",
       "E10_v1_AAACCTGCAGACAAGC-1  UBERON_0008930  somatosensory cortex   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1  UBERON_0008930  somatosensory cortex   \n",
       "...                                   ...                   ...   \n",
       "P4_TTTGTCAGTTGATTGC-1      UBERON_0008930  somatosensory cortex   \n",
       "P4_TTTGTCATCAACGGCC-1      UBERON_0008930  somatosensory cortex   \n",
       "P4_TTTGTCATCCGTTGTC-1      UBERON_0008930  somatosensory cortex   \n",
       "P4_TTTGTCATCGATGAGG-1      UBERON_0008930  somatosensory cortex   \n",
       "P4_TTTGTCATCTTTCCTC-1      UBERON_0008930  somatosensory cortex   \n",
       "\n",
       "                          library_preparation_protocol  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1                  EFO_0009899   \n",
       "E10_v1_AAACCTGCACAACGCC-1                  EFO_0009899   \n",
       "E10_v1_AAACCTGCACGACGAA-1                  EFO_0009899   \n",
       "E10_v1_AAACCTGCAGACAAGC-1                  EFO_0009899   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1                  EFO_0009899   \n",
       "...                                                ...   \n",
       "P4_TTTGTCAGTTGATTGC-1                      EFO_0009899   \n",
       "P4_TTTGTCATCAACGGCC-1                      EFO_0009899   \n",
       "P4_TTTGTCATCCGTTGTC-1                      EFO_0009899   \n",
       "P4_TTTGTCATCGATGAGG-1                      EFO_0009899   \n",
       "P4_TTTGTCATCTTTCCTC-1                      EFO_0009899   \n",
       "\n",
       "                          library_preparation_protocol__ontology_label    sex  \\\n",
       "E10_v1_AAACCTGAGGGTCTCC-1                         10X 3' v2 sequencing  mixed   \n",
       "E10_v1_AAACCTGCACAACGCC-1                         10X 3' v2 sequencing  mixed   \n",
       "E10_v1_AAACCTGCACGACGAA-1                         10X 3' v2 sequencing  mixed   \n",
       "E10_v1_AAACCTGCAGACAAGC-1                         10X 3' v2 sequencing  mixed   \n",
       "E10_v1_AAACCTGCAGAGCCAA-1                         10X 3' v2 sequencing  mixed   \n",
       "...                                                                ...    ...   \n",
       "P4_TTTGTCAGTTGATTGC-1                             10X 3' v2 sequencing  mixed   \n",
       "P4_TTTGTCATCAACGGCC-1                             10X 3' v2 sequencing  mixed   \n",
       "P4_TTTGTCATCCGTTGTC-1                             10X 3' v2 sequencing  mixed   \n",
       "P4_TTTGTCATCGATGAGG-1                             10X 3' v2 sequencing  mixed   \n",
       "P4_TTTGTCATCTTTCCTC-1                             10X 3' v2 sequencing  mixed   \n",
       "\n",
       "                          species__ontology_label  \n",
       "E10_v1_AAACCTGAGGGTCTCC-1            Mus musculus  \n",
       "E10_v1_AAACCTGCACAACGCC-1            Mus musculus  \n",
       "E10_v1_AAACCTGCACGACGAA-1            Mus musculus  \n",
       "E10_v1_AAACCTGCAGACAAGC-1            Mus musculus  \n",
       "E10_v1_AAACCTGCAGAGCCAA-1            Mus musculus  \n",
       "...                                           ...  \n",
       "P4_TTTGTCAGTTGATTGC-1                Mus musculus  \n",
       "P4_TTTGTCATCAACGGCC-1                Mus musculus  \n",
       "P4_TTTGTCATCCGTTGTC-1                Mus musculus  \n",
       "P4_TTTGTCATCGATGAGG-1                Mus musculus  \n",
       "P4_TTTGTCATCTTTCCTC-1                Mus musculus  \n",
       "\n",
       "[98047 rows x 27 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(\n",
    "    '../data/metaData_scDevSC.txt',\n",
    "    sep = '\\t',\n",
    "    skiprows = [1]\n",
    ")\n",
    "metadata.index = metadata.NAME.to_list()\n",
    "metadata.drop(\n",
    "    columns = ['NAME'],\n",
    "    inplace = True\n",
    ")\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "legitimate-estonia",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 3485 × 27998\n",
      "    var: 'gene_ids'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3339 × 27998\n",
       "    obs: 'orig_ident', 'nCount_RNA', 'nFeature_RNA', 'percent_mito', 'n_hkgene', 'S_Score', 'G2M_Score', 'Phase', 'CC_Difference', 'seurat_clusters', 'RNA_snn_res_1', 'scrublet_doublet', 'RNA_snn_res_2', 'Doublet_intersect', 'Gral_cellType', 'New_cellType', 'biosample_id', 'donor_id', 'species', 'disease', 'disease__ontology_label', 'organ', 'organ__ontology_label', 'library_preparation_protocol', 'library_preparation_protocol__ontology_label', 'sex', 'species__ontology_label'\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e14_5 = sc.read_10x_h5('../data/mouse_brain_E14_5.h5')\n",
    "e14_5.var_names_make_unique()\n",
    "print(e14_5)\n",
    "e14_5_metadata = metadata[metadata.index.str.startswith('E14')]\n",
    "e14_5_metadata.index = e14_5_metadata.index.str.replace('E14_5_', '')\n",
    "e14_5_metadata.index = [i + '-1' for i in e14_5_metadata.index]\n",
    "e14_5.obs = e14_5.obs.merge(\n",
    "    e14_5_metadata,\n",
    "    left_index = True,\n",
    "    right_index = True,\n",
    "    how = 'left'\n",
    ")\n",
    "e14_5 = e14_5[~e14_5.obs.isna().all(axis = 1), :].copy()\n",
    "e14_5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "marine-cherry",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 6480 × 27998\n",
      "    var: 'gene_ids'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6367 × 27998\n",
       "    obs: 'orig_ident', 'nCount_RNA', 'nFeature_RNA', 'percent_mito', 'n_hkgene', 'S_Score', 'G2M_Score', 'Phase', 'CC_Difference', 'seurat_clusters', 'RNA_snn_res_1', 'scrublet_doublet', 'RNA_snn_res_2', 'Doublet_intersect', 'Gral_cellType', 'New_cellType', 'biosample_id', 'donor_id', 'species', 'disease', 'disease__ontology_label', 'organ', 'organ__ontology_label', 'library_preparation_protocol', 'library_preparation_protocol__ontology_label', 'sex', 'species__ontology_label'\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p4 = sc.read_10x_h5('../data/mouse_brain_P4.h5')\n",
    "p4.var_names_make_unique()\n",
    "print(p4)\n",
    "p4_metadata = metadata[metadata.index.str.startswith('P4')]\n",
    "p4_metadata.index = p4_metadata.index.str.replace('P4_', '')\n",
    "p4_metadata\n",
    "p4.obs = p4.obs.merge(\n",
    "    p4_metadata,\n",
    "    left_index = True,\n",
    "    right_index = True,\n",
    "    how = 'left'\n",
    ")\n",
    "p4 = p4[~p4.obs.isna().all(axis = 1), :].copy()\n",
    "p4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "reduced-target",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26589"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_genes = set(adult.var.index).intersection(\n",
    "    set(p4.var.index)\n",
    ")\n",
    "len(common_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "banned-singapore",
   "metadata": {},
   "outputs": [],
   "source": [
    "e14_5 = e14_5[:, e14_5.var.index.isin(common_genes)].copy()\n",
    "p4 = p4[:, p4.var.index.isin(common_genes)].copy()\n",
    "adult = adult[:, adult.var.index.isin(common_genes)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "contrary-collins",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Allen_Brain_Atlas_IN':       class_label subclass_label  subclass_id\n",
       " 0       GABAergic           Sncg            4\n",
       " 1       GABAergic          Lamp5            3\n",
       " 2       GABAergic          Lamp5            3\n",
       " 3       GABAergic          Lamp5            3\n",
       " 4       GABAergic          Lamp5            3\n",
       " ..            ...            ...          ...\n",
       " 370  Non-Neuronal          Oligo           37\n",
       " 371  Non-Neuronal          Oligo           37\n",
       " 372  Non-Neuronal          Astro           38\n",
       " 373  Non-Neuronal          Astro           38\n",
       " 374  Non-Neuronal          Astro           38\n",
       " \n",
       " [375 rows x 3 columns],\n",
       " 'Allen_Brain_Atlas_OUT':       class_label subclass_label  subclass_id\n",
       " 0   Glutamatergic             CR            1\n",
       " 1       GABAergic          Meis2            2\n",
       " 2       GABAergic          Meis2            2\n",
       " 3    Non-Neuronal           Endo           39\n",
       " 4    Non-Neuronal       SMC-Peri           40\n",
       " 5    Non-Neuronal       SMC-Peri           40\n",
       " 6    Non-Neuronal       SMC-Peri           40\n",
       " 7    Non-Neuronal           VLMC           41\n",
       " 8    Non-Neuronal           VLMC           41\n",
       " 9    Non-Neuronal           VLMC           41\n",
       " 10   Non-Neuronal      Micro-PVM           42\n",
       " 11   Non-Neuronal      Micro-PVM           42\n",
       " 12   Non-Neuronal      Micro-PVM           42\n",
       " 13    Low Quality    Low Quality           43\n",
       " 14    Low Quality    Low Quality           43\n",
       " 15    Low Quality    Low Quality           43,\n",
       " 'Di_Bella_IN':                 class_label             subclass_label\n",
       " 0              Progenitors          apical progenitors\n",
       " 1              Progenitors   intermediate progenitors \n",
       " 2                 GABAergic               interneurons\n",
       " 3   Immature Glutamatergic           migrating neurons\n",
       " 4   Immature Glutamatergic           immature neurons \n",
       " 5     Mature Glutamatergic                 UL_CPN_L2&3\n",
       " 6     Mature Glutamatergic                 DL_CPN_L5&6\n",
       " 7     Mature Glutamatergic             Stellate_Layer4\n",
       " 8     Mature Glutamatergic                        SCPN\n",
       " 9     Mature Glutamatergic                          NP\n",
       " 10    Mature Glutamatergic                      CthPN \n",
       " 11             Non-Neuronal                      Oligo\n",
       " 12             Non-Neuronal                      Astro,\n",
       " 'Di_Bella_OUT':      class_label     subclass_label\n",
       " 0  Glutamatergic                 CR\n",
       " 1   Non-Neuronal          microglia\n",
       " 2   Non-Neuronal      ependymocytes\n",
       " 3   Non-Neuronal  endothelial cells\n",
       " 4   Non-Neuronal               VLMC\n",
       " 5   Non-Neuronal          pericytes\n",
       " 6   Non-Neuronal   red blood cells }"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_grouping = pd.read_excel(\n",
    "    '../meta/scRNASeq_datasets_wt_groups.xlsx', \n",
    "    sheet_name = None\n",
    ")\n",
    "cell_grouping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "worthy-alcohol",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove cells\n",
    "rem_cellids = cell_grouping['Di_Bella_OUT'].subclass_label.str.strip().str.lower().to_list() + ['low quality cells', 'doublet', 'cajal retzius cells']\n",
    "\n",
    "e14_5 = e14_5[~e14_5.obs.New_cellType.str.lower().isin(rem_cellids)].copy()\n",
    "p4 = p4[~p4.obs.New_cellType.str.lower().isin(rem_cellids)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "cordless-greek",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate class labels\n",
    "group_assignment = {\n",
    "    row['subclass_label'].strip().lower(): row['class_label'].strip() for i, row in cell_grouping['Di_Bella_IN'].iterrows()\n",
    "}\n",
    "group_assignment['astrocytes'] = 'Non-Neuronal'\n",
    "group_assignment['dl_cpn_1'] = 'Mature Glutamatergic'\n",
    "group_assignment['dl_cpn_2'] = 'Mature Glutamatergic'\n",
    "group_assignment['oligodendrocytes'] = 'Non-Neuronal'\n",
    "group_assignment['layer 4'] = 'Mature Glutamatergic'\n",
    "group_assignment['ul cpn'] = 'Mature Glutamatergic'\n",
    "group_assignment['cycling glial cells'] = 'Non-Neuronal'\n",
    "group_assignment['layer 6b'] = 'Mature Glutamatergic'\n",
    "\n",
    "e14_5.obs['class_label'] = e14_5.obs.New_cellType.str.lower().apply(\n",
    "    lambda x: group_assignment[x]\n",
    ")\n",
    "p4.obs['class_label'] = p4.obs.New_cellType.str.lower().apply(\n",
    "    lambda x: group_assignment[x]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "specific-reference",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove cells; adult does not need to be relabelled since it already contains this labelling so only removal is done here\n",
    "rem_cellids = cell_grouping['Allen_Brain_Atlas_OUT'].subclass_label.str.strip().str.lower().unique()\n",
    "adult = adult[~adult.obs.subclass_label.str.lower().isin(rem_cellids), :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "radical-permission",
   "metadata": {},
   "outputs": [],
   "source": [
    "keep_adult = ['class_label', 'subclass_label']\n",
    "adult.obs = adult.obs.loc[:, keep_adult]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "protecting-messaging",
   "metadata": {},
   "outputs": [],
   "source": [
    "for adata in [e14_5, p4]:\n",
    "    adata.obs = adata.obs.loc[:, ['class_label', 'New_cellType']]\n",
    "    adata.obs.rename(\n",
    "        columns = {\n",
    "            'New_cellType': 'subclass_label'\n",
    "        },\n",
    "        inplace = True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "wicked-prison",
   "metadata": {},
   "outputs": [],
   "source": [
    "# relabel Glutamateric to Mature Glutamatergic to comply to labelling of dev data\n",
    "adult.obs.loc[:, 'class_label'] = adult.obs.class_label.apply(\n",
    "    lambda x: 'Mature ' + x if x == 'Glutamatergic' else x\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "orange-religious",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'class_label' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'subclass_label' as categorical\n",
      "/users/daniel.malzl/.conda/envs/nova/lib/python3.9/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'developmental_stage' as categorical\n"
     ]
    }
   ],
   "source": [
    "# combine and write to file\n",
    "for adata, ident in zip([e14_5, p4, adult], ['E14.5', 'P4', 'Adult']):\n",
    "    adata.obs.index = [ident + '_' + i for i in adata.obs.index]\n",
    "    adata.obs['developmental_stage'] = ident\n",
    "    \n",
    "adata = ad.concat(\n",
    "    [e14_5, p4, adult]\n",
    ")\n",
    "adata.write_h5ad('../data/mouse_brain_development.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "foreign-constitutional",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>subclass_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>class_label</th>\n",
       "      <th>developmental_stage</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">GABAergic</th>\n",
       "      <th>Adult</th>\n",
       "      <td>4053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E14.5</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Immature Glutamatergic</th>\n",
       "      <th>Adult</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E14.5</th>\n",
       "      <td>1284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Mature Glutamatergic</th>\n",
       "      <th>Adult</th>\n",
       "      <td>74263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E14.5</th>\n",
       "      <td>632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>3340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Non-Neuronal</th>\n",
       "      <th>Adult</th>\n",
       "      <td>2199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E14.5</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>1089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Progenitors</th>\n",
       "      <th>Adult</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E14.5</th>\n",
       "      <td>1058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            subclass_label\n",
       "class_label            developmental_stage                \n",
       "GABAergic              Adult                          4053\n",
       "                       E14.5                           119\n",
       "                       P4                              882\n",
       "Immature Glutamatergic Adult                             0\n",
       "                       E14.5                          1284\n",
       "                       P4                              912\n",
       "Mature Glutamatergic   Adult                         74263\n",
       "                       E14.5                           632\n",
       "                       P4                             3340\n",
       "Non-Neuronal           Adult                          2199\n",
       "                       E14.5                             0\n",
       "                       P4                             1089\n",
       "Progenitors            Adult                             0\n",
       "                       E14.5                          1058\n",
       "                       P4                                0"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.groupby(['class_label', 'developmental_stage']).count()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-nova]",
   "language": "python",
   "name": "conda-env-.conda-nova-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
